<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (Version 7, May 2011) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_7-05_2011" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_v9_2012.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_06_29_2010.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_7-05_2011.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (Version 7, May 2011)</span></h1><p class="small">Created: May 1, 2011; Last Update: May 9, 2014.</p></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="#release_notes_7-05_.Download">Download</a></p></li><li><p><a href="#release_notes_7-05_.Third_Party_Pack">Third Party Packages</a></p></li><li><p><a href="#release_notes_7-05_.Build">Build</a></p></li><li><p><a href="#release_notes_7-05_.New_Developments">New Developments</a></p><ul><li><p><a href="#release_notes_7-05_.HIGHLIGHTS">HIGHLIGHTS</a></p></li><li><p><a href="#release_notes_7-05_.CORELIB">CORELIB</a></p></li><li><p><a href="#release_notes_7-05_.SERIAL">SERIAL</a></p></li><li><p><a href="#release_notes_7-05_.DATATOOL">DATATOOL</a></p></li><li><p><a href="#release_notes_7-05_.CONNECT">CONNECT</a></p></li><li><p><a href="#release_notes_7-05_.UTILITES">UTILITIES</a></p></li><li><p><a href="#release_notes_7-05_.BIOOBJECTS">BIO-OBJECTS</a></p></li><li><p><a href="#release_notes_7-05_.BIOTOOLS">BIO-TOOLS</a></p></li><li><p><a href="#release_notes_7-05_.COBALT">COBALT</a></p></li><li><p><a href="#release_notes_7-05_.Phylogenetic_tre">PHYLOGENETIC TREES</a></p></li><li><p><a href="#release_notes_7-05_.BIOOBJECT_LIBRAR">BIO-OBJECT LIBRARIES</a></p></li><li><p><a href="#release_notes_7-05_.BIOOBJECT_MANAGE">BIO-OBJECT MANAGER</a></p><ul><li><p><a href="#release_notes_7-05_.New_functional_1">New functionality</a></p></li><li><p><a href="#release_notes_7-05_.Improvements">Improvements</a></p></li><li><p><a href="#release_notes_7-05_.Bug_fixes">Bug fixes</a></p></li></ul></li><li><p><a href="#release_notes_7-05_.GENBANK_DATA_LOA">GENBANK DATA LOADER</a></p><ul><li><p><a href="#release_notes_7-05_.New_functional_2">New functionality</a></p></li><li><p><a href="#release_notes_7-05_.Improvement">Improvements</a></p></li></ul></li><li><p><a href="#release_notes_7-05_.OBJECT_MANAGER_T">OBJECT MANAGER TEST AND DEMO APPLICATIONS</a></p></li><li><p><a href="#release_notes_7-05_.FASTA">FASTA</a></p></li><li><p><a href="#release_notes_7-05_.asn2flat_utility">asn2flat utility</a></p></li><li><p><a href="#release_notes_7-05_.XMLWRAPP">XMLWRAPP</a></p></li><li><p><a href="#release_notes_7-05_.WindowMasker">WindowMasker</a></p></li><li><p><a href="#release_notes_7-05_.BUILD_FRAMEWORK">BUILD FRAMEWORK (Unix)</a></p></li><li><p><a href="#release_notes_7-05_.BLAST">BLAST</a></p></li><li><p><a href="#release_notes_7-05_.APPLICATIONS">APPLICATIONS</a></p><ul><li><p><a href="#release_notes_7-05_.NETCACHE">NETCACHE</a></p></li><li><p><a href="#release_notes_7-05_.GRID_APPLICATION">GRID APPLICATIONS</a></p></li></ul></li></ul></li><li><p><a href="#release_notes_7-05_.Documentation">Documentation</a></p><ul><li><p><a href="#release_notes_7-05_.Location">Location</a></p></li><li><p><a href="#release_notes_7-05_.Content">Content</a></p></li></ul></li><li><p><a href="#release_notes_7-05_.Supported_Platfo">Supported Platforms (OS's and Compilers)</a></p><ul><li><p><a href="#release_notes_7-05_.HYPERLINK__boo_1">Unix</a></p></li><li><p><a href="#release_notes_7-05_.HYPERLINK__boo_2">MS Windows</a></p></li><li><p><a href="#release_notes_7-05_.HYPERLINK__boo_3">Mac OS X</a></p></li><li><p><a href="#release_notes_7-05_.Added_Platforms">Added Platforms</a></p></li><li><p><a href="#release_notes_7-05_.Discontinued_Pla">Discontinued Platforms</a></p></li></ul></li><li><p><a href="#release_notes_7-05_.Caveats_and_Hint">Caveats and Hints</a></p><ul><li><p><a href="#release_notes_7-05_.GCC_34x">GCC 3.4.x</a></p></li></ul></li></ul><div class="sec" id="release_notes_7-05_.Download"><h2><span class="title">Download</span></h2><p>Download the source code archives at:</p><p><a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/</a></p><ul><li><p> <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ncbi_cxx--7_0_0.tar.gz">ncbi_cxx--7_0_0.tar.gz</a> — for Unix'es (see the list of Unix flavors below) and MacOSX</p></li><li><p><a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ncbi_cxx--7_0_0.exe">ncbi_cxx--7_0_0.exe</a> — for MS-Windows (32- and 64-bit) / MSVC++ 9.0 — self-extracting</p></li><li><p><a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ncbi_cxx--7_0_0.zip">ncbi_cxx--7_0_0.zip</a> — for MS-Windows (32- and 64-bit) / MSVC++ 9.0 </p></li></ul><p>The sources correspond to the NCBI production tree <a href="ch_getcode_svn.html#ch_getcode_svn.chkout_production_tree">sources</a>, which are originally based on the development tree source snapshot from October 25, 2010 but also include many hundreds of important and safe code updates made since then and through March 28, 2011 (and then some).</p><p>There are also two sub-directories, containing easily buildable source distributives of the NCBI C Toolkit (for MS Windows and Unix) and selected 3rd-party packages (for MS Windows only). These are the versions that the NCBI C++ Toolkit should build with. For build instructions, see README files there:</p><ul><li><p> <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/NCBI_C_Toolkit/README">NCBI_C_Toolkit </a> </p></li><li><p> <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ThirdParty/README">ThirdParty</a></p></li></ul></div><div class="sec" id="release_notes_7-05_.Third_Party_Pack"><h2><span class="title">Third Party Packages</span></h2><p>Some parts of the C++ Toolkit just cannot be built without 3<sup>rd</sup> party libraries, and other parts of the Toolkit will work more efficiently or provide more functionality if some 3rd-party packages (such as BerkeleyDB which is used for local data cache and for local data storage) are available.</p><p>For more information, see the <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ThirdParty/README">FTP README</a>.</p><p>The following table shows the versions of 3<sup>rd</sup> party packages that are believed to be compatible with the C++ Toolkit.</p><div class="table" id="release_notes_7-05_.T.compatible_version"><div class="caption"><p>Table 1. Compatible Versions of Third Party Packages</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="middle" rowspan="1" colspan="1">Package</th><th align="left" valign="top" rowspan="1" colspan="1">FreeBSD<br />32</th><th align="left" valign="top" rowspan="1" colspan="1">Linux<br />32</th><th align="left" valign="top" rowspan="1" colspan="1">Linux<br />64</th><th align="left" valign="top" rowspan="1" colspan="1">Mac<br />OS X</th><th align="left" valign="top" rowspan="1" colspan="1">SunOS<br />x86</th><th align="left" valign="top" rowspan="1" colspan="1">SunOS<br />SPARC</th><th align="left" valign="top" rowspan="1" colspan="1">Windows <sup>a</sup> (32, 64 bit)</th></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"> <a href="http://www.oracle.com/us/products/database/berkeley-db/index.html">BerkeleyDB</a> </td><td align="left" valign="top" rowspan="1" colspan="1">4.4.20, 4.6.21.1</td><td align="left" valign="top" rowspan="1" colspan="1">4.6.21.1</td><td align="left" valign="top" rowspan="1" colspan="1">4.6.21.1</td><td align="left" valign="top" rowspan="1" colspan="1">4.5.20</td><td align="left" valign="top" rowspan="1" colspan="1">4.6.21.1</td><td align="left" valign="top" rowspan="1" colspan="1">4.5.20</td><td align="left" valign="top" rowspan="1" colspan="1">4.6.21.NC</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.boost.org">Boost Test</a> </td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0</td><td align="left" valign="top" rowspan="1" colspan="1">1.41.0</td><td align="left" valign="top" rowspan="1" colspan="1">1.41.0</td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0</td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0</td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0</td><td align="left" valign="top" rowspan="1" colspan="1">1.42.0</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.fastcgi.com">FastCGI</a> </td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">2.4.0</td><td align="left" valign="top" rowspan="1" colspan="1">2.4.0</td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">2.1</td><td align="left" valign="top" rowspan="1" colspan="1">2.4.0</td><td align="left" valign="top" rowspan="1" colspan="1">2.4.0</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.bzip.org">libbzip2</a> </td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">1.0.2</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://freshmeat.net/projects/libjpeg">libjpeg</a> </td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">6b</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.libpng.org/pub/png/libpng.html">libpng</a> </td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.7</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.libtiff.org">libtiff</a> </td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.1</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://sourceforge.net/projects/giflib/files/libungif-4.x/libungif-4.1.4/">libungif</a> <sup>b</sup></td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">4.1.3</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.oberhumer.com/opensource/lzo"><span class="small-caps">LZO</span></a> </td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">2.02</td><td align="left" valign="top" rowspan="1" colspan="1">2.02</td><td align="left" valign="top" rowspan="1" colspan="1">2.02</td><td align="left" valign="top" rowspan="1" colspan="1">2.02</td><td align="left" valign="top" rowspan="1" colspan="1">2.02</td><td align="left" valign="top" rowspan="1" colspan="1">2.02</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.pcre.org">PCRE</a> </td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">7.9</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.sqlite.org">SQLite3</a> </td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.sybase.com">Sybase</a> </td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">12.5-32bit</td><td align="left" valign="top" rowspan="1" colspan="1">12.5-64bit</td><td align="left" valign="top" rowspan="1" colspan="1">-</td><td align="left" valign="top" rowspan="1" colspan="1">12.5.1</td><td align="left" valign="top" rowspan="1" colspan="1">12.0-EBF9209</td><td align="left" valign="top" rowspan="1" colspan="1">12.5</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.zlib.org">zlib</a> </td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">current</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.3</td></tr></tbody></table></div></div><p> <sup>a</sup> Applies to MSVC 9.0</p><p><sup>b</sup> Only 4.1.3 version is supported currently.</p><div class="table" id="release_notes_7-05_.T.currently_supporte"><div class="caption"><p>Table 2. Currently Supported/Tested Versions of Third Party Packages</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Package</th><th align="left" valign="top" rowspan="1" colspan="1">Versions expected to work (obtained by build-environment inspection in some cases)</th><th align="left" valign="top" rowspan="1" colspan="1">Versions known to work (used in-house on any platform)</th></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"> <a href="http://www.oracle.com/us/products/database/berkeley-db/index.html">BerkeleyDB</a> </td><td align="left" valign="top" rowspan="1" colspan="1">4.3.0 or newer</td><td align="left" valign="top" rowspan="1" colspan="1">4.5.20, 4.6.21, 4.6.21.1, 4.7.25</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.boost.org">Boost Test</a> </td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0 or newer</td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0, 1.40.0, 1.41.0, 1.42.0, 1.45.0</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.fastcgi.com">FastCGI</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">2.1, 2.4.0</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.bzip.org">libbzip2</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.0.2, 1.0.3, 1.0.5, 1.0.6</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://freshmeat.net/projects/libjpeg">libjpeg</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">6b</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.libpng.org/pub/png/libpng.html">libpng</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.5, 1.2.7, 1.2.8, 1.2.10, 1.2.42, 1.2.44, 1.4.3</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.libtiff.org">libtiff</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.1, 3.8.2, 3.9.2, 3.9.4</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://sourceforge.net/projects/giflib/files/libungif-4.x/libungif-4.1.4/">libungif</a></td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">4.1.0b1 (libungif),<br />4.1.3, 4.1.6 (giflib)</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://xmlsoft.org/">libxml2</a></td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">2.7.3, 2.7.6</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://xmlsoft.org/">libxslt</a></td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.1.24, 1.1.26</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.oberhumer.com/opensource/lzo"><span class="small-caps">LZO</span></a> </td><td align="left" valign="top" rowspan="1" colspan="1">2.x</td><td align="left" valign="top" rowspan="1" colspan="1">2.02, 2.03</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.pcre.org">PCRE</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">4.4, 6.6, 7.8, 7.9</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.sqlite.org">SQLite3</a> </td><td align="left" valign="top" rowspan="1" colspan="1">3.6.6 or newer</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.sybase.com">Sybase</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">12.0-EBF9209, 12.5.0.6-ESD13, 12.5.1,12.5.1.10-ESD26</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.zlib.org">zlib</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.1, 1.2.3, 1.2.3.3</td></tr></tbody></table></div></div><p>For Mac OS X and Unix OS’s, the user is expected to download and build the 3<sup>rd</sup> party packages themselves.  The release’s package list includes links to download sites. However, the user still needs a list of the 3<sup>rd</sup> party packages and which versions of them are compatible with the release.</p><p>To facilitate the building of these 3rd-party libraries on Windows, there is an <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ThirdParty/README">archive</a> that bundles together source code of the 3rd-party packages, plus MSVC "solutions" to build all (or any combination) of them. </p><div class="table" id="release_notes_7-05_.T.versions_of_third_"><div class="caption"><p>Table 3. Versions of Third Party Packages Included in the FTP Archive_</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Package</th><th align="left" valign="top" rowspan="1" colspan="1">Depends On</th><th align="left" valign="top" rowspan="1" colspan="1">Included Version <sup>a</sup> </th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">BerkeleyDB</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">4.6.21.NC </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Boost Test</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">1.42.0</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libbzip2</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">1.0.2</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libjpeg</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">6b</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libpng</td><td align="left" valign="top" rowspan="1" colspan="1">zlib 1.2.3</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.7</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libtiff</td><td align="left" valign="top" rowspan="1" colspan="1">libjpeg 6b, zlib 1.2.3</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.1</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libungif</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">4.1.3</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">LZO</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">2.02 </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">PCRE</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">7.9</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">SQLite3</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">zlib</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">1.2.3</td></tr></tbody></table></div></div><p> <sup>a</sup> Applies to MSVC 9.0 </p></div><div class="sec" id="release_notes_7-05_.Build"><h2><span class="title">Build</span></h2><p>For guidelines to configure, build and install the Toolkit see <a href="ch_config.html">here</a>.</p></div><div class="sec" id="release_notes_7-05_.New_Developments"><h2><span class="title">New Developments</span></h2><div class="sec" id="release_notes_7-05_.HIGHLIGHTS"><h3><span class="title">HIGHLIGHTS</span></h3><ul><li><p>Added LDS2 (Local Data Storage v.2) which is based on SQLite3, has new features and better performance. Also implemented LDS2 data loader to use LDS2 from the Object Manager.</p></li><li><p>XmlWrapp –this convenient XML handling API has been mostly finished (and even polished).</p></li><li><p>Implemented tunneling and authorization of HTTP connections and tunneling of secure sockets, through HTTP proxies.</p></li><li><p><span class="nctnt ncbi-class">CFormatGuess</span> now allows distinguishing between GTF, GFF3, and GFF2. It's a possibly breaking change. For more details see below.</p></li><li><p>Implemented major parts of <span class="nctnt ncbi-class">CFeatTree</span>, the class to organize features defined on a biological sequence into a hierarchy that reflects their parent-child relationships (based on the feature subtypes).</p></li></ul></div><div class="sec" id="release_notes_7-05_.CORELIB"><h3><span class="title">CORELIB</span></h3><ul><li><p>Implemented locale-independent conversion of string to double and back; changed core libraries to use it.</p></li><li><p><span class="nctnt ncbi-func">NStr::Justify()</span> -- for formatting of paragraphs of text.</p></li><li><p><span class="nctnt ncbi-class">CNcbiApplication</span> -- make FindProgramExecutablePath static, and more robust; add a static higher-level <span class="nctnt ncbi-func">GetAppName</span> method.  Look for global configuration files in more cases.</p></li><li><p><span class="nctnt ncbi-func">CMetaRegistry::FindRegistry</span> -- new method exposing the logic determining which file (if any) to load.</p></li><li><p><span class="nctnt ncbi-func">CEnvironmentCleaner</span> -- new class to discard unwanted environment variables.</p></li><li><p><span class="nctnt ncbi-class">CFileIO</span> -- back to original behavior: do not close the file handle if it's assigned via <span class="nctnt ncbi-func">SetFileHandle()</span>.</p></li></ul></div><div class="sec" id="release_notes_7-05_.SERIAL"><h3><span class="title">SERIAL</span></h3><ul><li><p>Serialization of <span class="nctnt ncbi-monospace">AnyContent</span> data objects -- fixed to recognize and properly process attributes in their values.</p></li><li><p>Corrected the reading of XML data to assign to an element default value when it has no content.</p></li><li><p>Added support for sequences of elements, where the element has a default value.</p></li></ul></div><div class="sec" id="release_notes_7-05_.DATATOOL"><h3><span class="title">DATATOOL</span></h3><ul><li><p>Corrected code generation of:</p><ul><li><p>CHOICE data objects;</p></li><li><p>binary data types with attributes.</p></li></ul></li><li><p>Corrected conversion of double type values to preserve more significant digits.</p></li></ul></div><div class="sec" id="release_notes_7-05_.CONNECT"><h3><span class="title">CONNECT</span></h3><ul><li><p>Added keepalive socket option (fSOCK_KeepAlive).</p></li><li><p>Added NCBI connectivity test (<span class="nctnt ncbi-class">CConnTest</span>).</p></li></ul></div><div class="sec" id="release_notes_7-05_.UTILITES"><h3><span class="title">UTILITES</span></h3><ul><li><p><span class="nctnt ncbi-func">g_FindDataFile</span> -- New function for locating data files in (configurable) standard locations.</p></li><li><p><span class="nctnt ncbi-class">CChecksumStreamWriter</span> - new class to compute checksum of the data written to a stream.</p></li><li><p><span class="nctnt ncbi-func">g_GZip_ScanForChunks()</span> - new API, to query compressed stream positions. Added implementation for getting positions for separate gzip-files inside concatenated gzip file.</p></li><li><p>Added compression/decompression stream manipulators (<span class="nctnt ncbi-path">include/util/compress/stream_util.hpp</span>).</p></li><li><p><span class="nctnt ncbi-class">CFormatGuess</span> (<span class="nctnt ncbi-path">util/format_guess.{h/c}pp</span>) updated, with a possibly breaking change. The purpose of this is to allow <span class="nctnt ncbi-class">CFormatGuess</span> to distinguish between GTF, GFF3, and GFF2. Currently it lumps all of those formats into a one 'eGtf' value. The old 'eGtf' value (3) is being replaced with 'eGtf_POISONED', and will not be returned again. The new value for 'eGtf' (21) will mean a file that should be read with <span class="nctnt ncbi-class">CGtfReader</span> (<span class="nctnt ncbi-path">objtools/readers/gtf_reader.hpp</span>). The new value 'eGff3' (22) is for files meant to be read with <span class="nctnt ncbi-class">CGff3Reader</span> (<span class="nctnt ncbi-path">objtools/readers/gff3_reader.hpp</span>), and 'eGff2' (24) is for files meant to be read with <span class="nctnt ncbi-class">CGff2Reader</span> (<span class="nctnt ncbi-path">include/objtools/readers/gff2_reader.hpp</span>)</p></li></ul></div><div class="sec" id="release_notes_7-05_.BIOOBJECTS"><h3><span class="title">BIO-OBJECTS</span></h3><ul><li><p><span class="nctnt ncbi-func">CBioseq::GetNonLocalId</span> -- New method to help place sequences imported from FASTA files with range specifications in more context; wrapped by <span class="nctnt ncbi-func">CBioseq_Handle::GetNonLocalIdOrNull</span> (likewise new).</p></li><li><p><span class="nctnt ncbi-func">CSeq_id::IdentifyAccession</span> -- Implement or improve recognition for more prefixes (GA, HH, HI, HO-HU, JA-JO, EAAA-EZZZ, and IAA-IZZ, some of which correspond to the new possibility of DDBJ TPA WGS data) and mixed-in TPA protein accessions (mostly from EMBL, but some from GenBank too).</p></li><li><p>Distinguish WGS master accessions by a new flag bit.  Relax over-strict PDB recognition logic.</p></li><li><p><span class="nctnt ncbi-func">CSeq_id::IsValidLocalID</span>, <span class="nctnt ncbi-func">CSeq_id::ParseIDs</span> -- New functionality for working with plain-text sequence identifiers, factored out of <span class="nctnt ncbi-class">CFastaReader</span> and generalized somewhat.</p></li><li><p><span class="nctnt ncbi-type">SSeqIdRange</span> -- New type (complete with parser and on-the-fly "iterator") for working with Seq-id ranges, as present in some FASTA defline source modifiers.</p></li></ul></div><div class="sec" id="release_notes_7-05_.BIOTOOLS"><h3><span class="title">BIO-TOOLS</span></h3><ul><li><p><span class="nctnt ncbi-class">CFastaOstream</span> -- Optionally accept custom titles for single sequences.  Tag negative-strand ranges with leading 'c's.</p></li><li><p><span class="nctnt ncbi-class">CFastaReader</span> -- Support negative-strand ranges and Sequin's compact defline-style gap syntax ("&gt;?N" where N is a number; or "&gt;?unk100").</p></li></ul></div><div class="sec" id="release_notes_7-05_.COBALT"><h3><span class="title">COBALT</span></h3><ul><li><p>Added command-line option <span class="nctnt ncbi-cmd">-num_domain_hits</span> that limits number of conserved domains per sequence used in computing alignment constraints.</p></li></ul></div><div class="sec" id="release_notes_7-05_.Phylogenetic_tre"><h3><span class="title">Phylogenetic trees</span></h3><ul><li><p>Added higher level interface for computing phylogenetic tree from sequence alignments (for example BLAST and COBALT results). Class <span class="nctnt ncbi-class">CPhyTreeCalc</span> computes phylogenetic tree, and <span class="nctnt ncbi-class">CPhyTreeFormater</span> prints the tree in Newick and Nexus format.</p></li></ul></div><div class="sec" id="release_notes_7-05_.BIOOBJECT_LIBRAR"><h3><span class="title">BIO-OBJECT LIBRARIES</span></h3><ul><li><p>Implemented <span class="nctnt ncbi-func">CheckNumRows()</span> and other methods for sparse alignments.</p></li><li><p>To reduce memory footprint: added read-hooks to reduce memory used by alignments after deserialization; <span class="nctnt ncbi-monospace">Na-strand</span> now uses one byte of memory where possible; <span class="nctnt ncbi-monospace">Score.value</span> choice is now embedded in <span class="nctnt ncbi-class">CScore</span>.</p></li><li><p>Capitalize accession in <span class="nctnt ncbi-func">CSeq_id::GetLabel()</span>.</p></li></ul></div><div class="sec" id="release_notes_7-05_.BIOOBJECT_MANAGE"><h3><span class="title">BIO-OBJECT MANAGER</span></h3><div class="sec" id="release_notes_7-05_.New_functional_1"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added getter methods for boolean fields in <span class="nctnt ncbi-class">CTableFieldHandle</span>.</p></li><li><p>Added <span class="nctnt ncbi-func">GetBestGeneForFeat()</span> based on <span class="nctnt ncbi-class">CFeatTree</span>.</p></li><li><p>Implemented <span class="nctnt ncbi-func">GetBestOverlappingFeat()</span> on <span class="nctnt ncbi-class">CFeatTree</span>.</p></li><li><p>Added fast <span class="nctnt ncbi-func">CScope::GetTaxid()</span>.</p></li><li><p>Implemented bulk loading for acc/ver, gi, label, and taxid.</p></li><li><p>Added zero-length gaps check to <span class="nctnt ncbi-class">CSeqMap</span> and <span class="nctnt ncbi-class">CSeqVector</span>.</p></li><li><p>Implemented <span class="nctnt ncbi-func">GetLength()</span> and <span class="nctnt ncbi-func">GetCoverage()</span> for bond locations.</p></li></ul></div><div class="sec" id="release_notes_7-05_.Improvements"><h4><span class="title">Improvements:</span></h4><ul><li><p>Added helper method to fill <span class="nctnt ncbi-class">CFeatTree</span> on location.</p></li><li><p>Sped up mapping of simple <span class="nctnt ncbi-class">CSeq_loc_mix</span> locations in <span class="nctnt ncbi-class">CFeat_CI</span>.</p></li><li><p>Stricter sorting of features in <span class="nctnt ncbi-class">CFeat_CI</span> to avoid ambiguities.</p></li><li><p><span class="nctnt ncbi-class">CSeq_feat_Handle</span> getters now work with Seq-table features too.</p></li><li><p><span class="nctnt ncbi-monospace">Seq-table</span> features now support multi-level user fields.</p></li><li><p>Non <span class="nctnt ncbi-monospace">Seq-feat Seq-tables</span> are now recognized even if located in split chunk.</p></li><li><p>Sped up <span class="nctnt ncbi-func">CBioseq_Handle::AddId()</span>.</p></li><li><p>Optimized <span class="nctnt ncbi-func">CScope::AttachXxx()</span>.</p></li><li><p>Support split of named annotation.</p></li><li><p><span class="nctnt ncbi-class">CSeqVector</span> and <span class="nctnt ncbi-class">CSeqVector_CI</span>'s <span class="nctnt ncbi-func">CanGetRange()</span> now return false instead of throwing an exception.</p></li><li><p>Allow to specify how to deal with existing handles in <span class="nctnt ncbi-func">ResetHistory()</span>.</p></li><li><p>Optimized re-parenting if more features are added to <span class="nctnt ncbi-class">CFeatTree</span>.</p></li><li><p>Added possibility to debug <span class="nctnt ncbi-class">CScope</span> creation/deletion.</p></li><li><p>Many changes to the C++ cleanup functionality to imitate the cleanup functionality which already exists in C.  There is still more work to be done with BasicCleanup, but significant progress has been made.  Little work has been done for ExtendedCleanup as of yet.</p></li><li><p><span class="nctnt ncbi-class">CSeq_loc_Mapper</span> can now be initialized with a <span class="nctnt ncbi-app">GC-Assembly</span>.</p></li></ul></div><div class="sec" id="release_notes_7-05_.Bug_fixes"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Fixed mapping of mix locations on minus strand in <span class="nctnt ncbi-class">CFeat_CI</span>.</p></li><li><p>Many fixes in the way <span class="nctnt ncbi-class">CFeatTree</span> links features.</p></li><li><p>Several thread-safety fixes.</p></li><li><p>Fixed typo preventing adding aligns and graphs to <span class="nctnt ncbi-class">CSeq_annot_EditHandle</span>.</p></li><li><p>Safeguard against exceptions when sorting features in <span class="nctnt ncbi-class">CFeat_CI</span>.</p></li></ul></div></div><div class="sec" id="release_notes_7-05_.GENBANK_DATA_LOA"><h3><span class="title">GENBANK DATA LOADER</span></h3><div class="sec" id="release_notes_7-05_.New_functional_2"><h4><span class="title">New functionality:</span></h4><ul><li><p>Registered HPRD external annotations.</p></li><li><p>Added optional exclude_wgs_master param in pubseqos/pubseqos2 readers.</p></li><li><p>Implemented bulk loading for acc/ver, gi, label, and taxid.</p></li><li><p>Added <span class="nctnt ncbi-func">CGBDataLoader::CloseCache()</span>.</p></li></ul></div><div class="sec" id="release_notes_7-05_.Improvement"><h4><span class="title">Improvement:</span></h4><ul><li><p>Use bulk loading requests in <span class="nctnt ncbi-func">CScope::GetBioseqHandles()</span>.</p></li><li><p>Separate reader statistics by type of loaded blobs.</p></li><li><p>Added timestamp to GenBank debug messages.</p></li><li><p>Use <span class="nctnt ncbi-class">IConnValidator</span> for opening PubSeqOS connections.</p></li><li><p>Added split-version to chunk requests and chunk subkeys in GenBank cache to avoid using wrong chunks when blob split state is changed in ID.</p></li><li><p>Added secondary less confusing param names for open timeout.</p></li><li><p>Do not multiply retry count by number of connections.</p></li></ul></div></div><div class="sec" id="release_notes_7-05_.OBJECT_MANAGER_T"><h3><span class="title">OBJECT MANAGER TEST AND DEMO APPLICATIONS</span></h3><ul><li><p><span class="nctnt ncbi-app">id2_fetch_simple</span> -- added <span class="nctnt ncbi-monospace">-id</span> options for arbitrary <span class="nctnt ncbi-monospace">Seq-id</span>'s.</p></li><li><p><span class="nctnt ncbi-app">test_bulkinfo</span> -- new test application.</p></li></ul></div><div class="sec" id="release_notes_7-05_.FASTA"><h3><span class="title">FASTA</span></h3><ul><li><p>C++ feature table functionality has been made more functional such as for part of the BankIt project.</p></li></ul></div><div class="sec" id="release_notes_7-05_.asn2flat_utility"><h3><span class="title">asn2flat utility</span></h3><ul><li><p>Huge number of changes to flatfile formatter to bring it much closer to release-ready state (possibly release ready at this point, although some relatively minor issues remain).</p></li></ul></div><div class="sec" id="release_notes_7-05_.XMLWRAPP"><h3><span class="title">XMLWRAPP</span></h3><ul><li><p>Fixed segmentation fault in case of taking a reference to XPath expression running results.</p></li><li><p>Added helpers to get public ID, system ID and DTD name for external and internal subsets.</p></li><li><p>Added methods to lookup node attributes.</p></li><li><p>Fixed execution of XPath expression: it now starts from the given node.</p></li><li><p>Fixed searching attributes (including default) when a namespace is provided.</p></li><li><p>Added ability to run XPath expression without necessity to register namespaces explicitly.</p></li><li><p>Added ability to provide containers for collecting errors and warnings while parsing documents.</p></li><li><p>Added ability to modify values and namespaces of node’s default attributes.</p></li><li><p>Added ability to test if an attribute is default.</p></li><li><p>Added ability to insert or remove attributes while taking into account their namespaces.</p></li><li><p>Added ability to strip XML declaration when a document is saved.</p></li></ul></div><div class="sec" id="release_notes_7-05_.WindowMasker"><h3><span class="title">WindowMasker</span></h3><ul><li><p>Added a new input format, "seqids"; with this input format, the input is a file containing a sequence id on each line, and the algorithm uses the Bio-Object Manager to look up the sequences.</p></li><li><p>Added a new class <span class="nctnt ncbi-class">CWinMaskConfig</span>, for storing all the WindowMasker configuration parameters. The class can be used to add the needed command-line arguments to <span class="nctnt ncbi-class">CArgDescriptions</span>, and then get the configuration parameters from the command-line arguments.</p></li></ul></div><div class="sec" id="release_notes_7-05_.BUILD_FRAMEWORK"><h3><span class="title">BUILD FRAMEWORK (Unix)</span></h3><ul><li><p>Interpret command-line specifications of APP_PROJ or LIB_PROJ as a cue to clear out other *_PROJ settings not also provided there.  (Requires GNU Make; builds with Sun make continue to work as before.)</p></li><li><p>Supply more targets in subdirectories: *_f (using local flat makefiles produced on demand, ignoring dependencies on other parts of the tree), *_fd (wrapping the top-level Makefile.flat), clean_sources and purge_sources.</p></li><li><p>Configure and its convenience scripts (<span class="nctnt ncbi-path">compilers/unix/*.sh</span>):</p><ul><li><p>Noteworthy new flag <span class="nctnt ncbi-monospace">--without-3psw</span> – to not use with any 3<sup>rd</sup>-party software.</p></li><li><p>Added a check for GLEW.</p></li><li><p>Improved checks for Boost and OpenGL.</p></li></ul></li><li><p>Support specifying run paths on Darwin (Mac) systems with modern toolchains.</p></li></ul></div><div class="sec" id="release_notes_7-05_.BLAST"><h3><span class="title">BLAST</span></h3><ul><li><p>On Darwin (Mac OS X), build only for Intel processors even in otherwise universal builds due to a PowerPC toolchain limitation.</p></li><li><p>Added support for retrieving NCBI Taxonomy IDs for which WindowMasker support is available.</p></li><li><p>Allow the specification of a query sequence along with multiple sequence alignment file in psiblast.</p></li><li><p>Added database hard-masking support.</p></li><li><p>Added database soft-masking for translated searches.</p></li><li><p>Added support for btop (BLAST traceback operations) and query and subject length in the tabular report.</p></li><li><p>Command-line applications -- allow <span class="nctnt ncbi-app">psiblast</span> to search multiple queries, added optional <span class="nctnt ncbi-monospace">-input_type</span> for <span class="nctnt ncbi-app">makeblastdb</span></p></li><li><p>Allow use of best hit and XML in blast2sequences mode.</p></li><li><p>Improved formatting performance for remote searches.</p></li><li><p><span class="nctnt ncbi-app">makembindex</span> can now build masked MegaBLAST index directly from a BLAST nucleotide database using the masking information stored in the BLAST database. This is accomplished by new command line option -<span class="nctnt ncbi-monospace">db_mask</span> to <span class="nctnt ncbi-app">makembindex</span>. The option accepts the integer id of the filtering algorithm supported by the BLAST database. The option can only be applied in conjunction with <span class="nctnt ncbi-monospace">-iformat blastdb</span>.</p></li><li><p>To assist a user in finding out the numeric ids of filtering algorithms supported by a BLAST database, the flag <span class="nctnt ncbi-monospace">-show_filters</span> is introduced. Applying the flag with <span class="nctnt ncbi-monospace">-iformat blastdb</span> and BLAST database as an input causes <span class="nctnt ncbi-app">makembindex</span> to output a list of available filtering algorithms and exit.</p></li></ul></div><div class="sec" id="release_notes_7-05_.APPLICATIONS"><h3><span class="title">APPLICATIONS</span></h3><div class="sec" id="release_notes_7-05_.NETCACHE"><h4><span class="title">NETCACHE</span></h4><ul><li><p><span class="nctnt ncbi-app">NetCache</span> is reworked to include the following features:</p><ul><li><p>better management of disk space;</p></li><li><p>lock-less work with blobs, versioning is used instead;</p></li><li><p>multi-port listening and per-client settings differentiating.</p></li></ul></li><li><p><span class="nctnt ncbi-app">NetCache</span> and <span class="nctnt ncbi-app">ICache</span> APIs:</p><ul><li><p>Use Uint8 everywhere for blob size.</p></li><li><p>Allow partial blob retrieval.</p></li><li><p>Introduced blob password protection; empty passwords are treated as no password.</p></li></ul></li><li><p>Worker node APIs:</p><ul><li><p>New parameter for terminating the worker node if its memory consumption exceeds the specified limit (parameter "total_memory_limit").</p></li><li><p>New parameter for terminating the worker node if its run time exceeds the specified limit (parameter "total_time_limit").</p></li></ul></li></ul></div><div class="sec" id="release_notes_7-05_.GRID_APPLICATION"><h4><span class="title">GRID APPLICATIONS</span></h4><ul><li><p><span class="nctnt ncbi-app">netscheduled</span></p><ul><li><p>Fixed a bug that caused no reply to the queue deletion command.</p></li></ul></li><li><p><span class="nctnt ncbi-app">remote_app</span></p><ul><li><p>New configuration parameter ("tmp_dir") to control how temporary directory name is generated - to reduce its length.</p></li><li><p>Log  blob writing error.</p></li></ul></li><li><p><span class="nctnt ncbi-app">netcache_control</span></p><ul><li><p>Allow partial blob retrieval.</p></li><li><p>New command <span class="nctnt ncbi-monospace">-remove</span> to delete blobs by their ids.</p></li><li><p>New parameter <span class="nctnt ncbi-monospace">-auth</span> to specify authentication string to use.</p></li><li><p>New commands <span class="nctnt ncbi-monospace">-reconf</span> and -reinit for use by <span class="nctnt ncbi-app">NetCache</span> administrators.</p></li></ul></li><li><p><span class="nctnt ncbi-app">netschedule_control</span></p><ul><li><p>Enabled compatibility mode to make <span class="nctnt ncbi-app">netschedule_control</span> work with older worker nodes.</p></li></ul></li><li><p><span class="nctnt ncbi-app">cgi2rcgi.cgi</span></p><ul><li><p>Do not create an empty NetCache blob as a placeholder for the progress message.</p></li><li><p>Log Grid errors that are reported to the user.</p></li><li><p>Allow spaces in the job ID parameter.</p></li><li><p>Support output of the job status information in JSON format.</p></li><li><p>Allow custom HTML templates to be defined for GRID errors and other events.</p></li><li><p>Added no-cache HTTP headers to avoid caching of intermediate results.</p></li></ul></li><li><p><span class="nctnt ncbi-app">ncfetch.cgi</span></p><ul><li><p>New parameter to access password-protected blobs.</p></li><li><p>Interpret extra parameter "filename" as a file name for the downloaded file.</p></li></ul></li></ul></div></div></div><div class="sec" id="release_notes_7-05_.Documentation"><h2><span class="title">Documentation</span></h2><div class="sec" id="release_notes_7-05_.Location"><h3><span class="title">Location</span></h3><p>The documentation is available online as a searchable book "The NCBI C++ Toolkit": <a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/">https://www.ncbi.nlm.nih.gov/toolkit/doc/book/</a></p><p>The C++ Toolkit book also provides PDF version of the chapters. The PDF version can be accessed by a link that appears on each page.</p></div><div class="sec" id="release_notes_7-05_.Content"><h3><span class="title">Content</span></h3><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit" gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the newcomer.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the up-to-date public or in-house versions using source browsers <a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CTime&amp;d=">LXR</a>, <a href="https://www.ncbi.nlm.nih.gov/toolkit/?term=ctime">Doxygen</a> and <a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi?symbol=CTime">Library</a> - or do an <a href="https://www.ncbi.nlm.nih.gov/sites/entrez?Db=ToolkitAll&amp;term=CTime">overall</a> search.</p><p> <b>HEADS-UP</b>: We have switched our source control system from CVS to SVN (Subversion). Unfortunately, the SVN repository cannot (yet) be accessed from outside NCBI. </p></div></div><div class="sec" id="release_notes_7-05_.Supported_Platfo"><h2><span class="title">Supported Platforms (OS's and Compilers)</span></h2><ul><li><p><a href="release_notes_7-05_2011.html#release_notes_7-05_.HYPERLINK__boo_1">Unix</a></p></li><li><p><a href="release_notes_7-05_2011.html#release_notes_7-05_.HYPERLINK__boo_2">MS Windows</a></p></li><li><p><a href="release_notes_7-05_2011.html#release_notes_7-05_.HYPERLINK__boo_3">Mac OS X</a></p></li><li><p><a href="release_notes_7-05_2011.html#release_notes_7-05_.Added_Platforms">Added</a></p></li><li><p><a href="release_notes_7-05_2011.html#release_notes_7-05_.Discontinued_Pla">Discontinued</a></p></li></ul><p>This release was successfully tested on at least the following platforms (but may also work on other platforms). Since the previous release, some platforms were dropped from this list and some were added. Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages. In these cases, just skipping such projects (e.g. using flag "<span class="nctnt ncbi-monospace">-k</span>" for <span class="nctnt ncbi-app">make</span> on Unix), can get you through.</p><p>In cases where multiple versions of a compiler are supported, the mainstream version is shown in <b>bold</b>.</p><div class="sec" id="release_notes_7-05_.HYPERLINK__boo_1"><h3><span class="title"><a href="ch_config.html#ch_config.UNIX">Unix</a></span></h3><div class="table" id="release_notes_7-05_.T.unix_oss_and_suppo"><div class="caption"><p>Table 3. Unix OS's and Supported Compilers</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">SuSE-9.x (LIBC 2.3.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32 </td><td align="left" valign="top" rowspan="1" colspan="1">GCC  <b>3.4.2</b>, 4.1.2<sup>a</sup>, 4.2.3<sup>a</sup>, 4.3.3<sup>a</sup>, 4.4.2<sup>a</sup>, 4.6.0<sup>a</sup><br />ICC <b>10.1, </b>11.1<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">SuSE-9.x (LIBC 2.3.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64 </td><td align="left" valign="top" rowspan="1" colspan="1">GCC<b> 4.0.1</b>, 4.1.2<sup>a</sup>, 4.2.3<sup>a</sup>, 4.3.3<sup>a</sup>, 4.4.2<sup>a</sup>, 4.6.0<sup>a</sup><br />ICC <b>10.1, </b>11.1<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">CentOS 5.4 (LIBC 2.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC 4.4.2 </b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">CentOS 5.4 (LIBC 2.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC 4.4.5 </b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Ubuntu 9.04 ("jaunty")  (LIBC 2.9)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC 4.3.3 </b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10 </td><td align="left" valign="top" rowspan="1" colspan="1">SPARC </td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.1.1<sup>b</sup> <br /><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.2.3<br /><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">FreeBSD-8.0</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.2.1</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Darwin 8.x, 9.x (Mac OS X 10.4.x, 10.5.x)</td><td align="left" valign="top" rowspan="1" colspan="1">Native, Universal</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.0.1</td></tr></tbody></table></div></div><p> <sup>a</sup> some support</p><p> <sup>b</sup> 32-bit only </p></div><div class="sec" id="release_notes_7-05_.HYPERLINK__boo_2"><h3><span class="title"><a href="ch_config.html#ch_config.MS_Windows">MS Windows</a></span></h3><div class="table" id="release_notes_7-05_.T.ms_windows_and_sup"><div class="caption"><p>Table 4. MS Windows and Supported Compilers</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2008 (C++ 9.0),<br />2010 (C++ 10.0) – nominal support only. <br />NOTE: We also ship an easily buildable archive of <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/ThirdParty/README">3rd-party</a> packages (including <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/ARCHIVE/7_0_0/NCBI_C_Toolkit/README">NCBI C Toolkit</a>) for this platform.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2008 (C++ 9.0),<br />2010 (C++ 10.0) – nominal support only. <br /><br />NOTE: We also ship an easily buildable archive of <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2010/Jun_15_2010/ThirdParty/README">3rd-party</a> packages (including <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2010/Jun_15_2010/NCBI_C_Toolkit/README">NCBI C Toolkit</a>) for this platform, although not all libraries are available for x86-64 architecture.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Cygwin 1.5.25</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.Cygwin_GCC">GCC 4.3.4</a>, <a href="ch_config.html#ch_config.Cygwin_GCC">GCC 3.4.4</a> - nominal support only.</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_7-05_.HYPERLINK__boo_3"><h3><span class="title"><a href="ch_config.html#ch_config.Mac_OS_X">Mac OS X</a></span></h3><div class="table" id="release_notes_7-05_.T.mac_os_and_support"><div class="caption"><p>Table 5. Mac OS and Supported Compilers</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Mac OS X 10.5, <br />MacOS x 10.6</td><td align="left" valign="top" rowspan="1" colspan="1">Native (PowerPC <u>or</u> x86-32<u> or</u> x86-64 )</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.Xcode_30__31">Xcode</a>  3.0 - 3.2.6</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Darwin 8.x, 9.x, 10.x</td><td align="left" valign="top" rowspan="1" colspan="1">Native (PowerPC <u>or</u> x86-32<u> or</u> x86-64),<br />Universal (PowerPC <u>and</u> x86-32)</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.GCC">GCC 4.0.1</a> <br /><a href="ch_config.html#ch_config.GCC">GCC 4.2.1</a> (only available under Darwin 10.x)</td></tr></tbody></table></div></div><p><span class="nctnt highlight">NOTE:</span>  the correspondence between Darwin kernel versions and Mac OS versions:</p><p>Darwin  8.x = Mac OS 10.4.x</p><p>Darwin  9.x = Mac OS 10.5.x</p><p>Darwin 10.x = Mac OS 10.6.x</p></div><div class="sec" id="release_notes_7-05_.Added_Platforms"><h3><span class="title">Added Platforms</span></h3><div class="table" id="release_notes_7-05_.T.added_platforms"><div class="caption"><p>Table 6. Added Platforms</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32, x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2010 (C++ 10.0) – nominal support only. </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">CentOS 5.4 (LIBC 2.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC 4.4.5 </b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10 </td><td align="left" valign="top" rowspan="1" colspan="1">SPARC </td><td align="left" valign="top" rowspan="1" colspan="1"><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1">Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Mac OS X 10.5, <br />MacOS x 10.6</td><td align="left" valign="top" rowspan="1" colspan="1">Native (PowerPC <u>or</u> x86-32<u> or</u> x86-64 )</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.Xcode_30__31">Xcode</a>  3.2.3 - 3.2.6</td></tr></tbody></table></div></div><p><sup>a</sup> some support</p></div><div class="sec" id="release_notes_7-05_.Discontinued_Pla"><h3><span class="title">Discontinued Platforms</span></h3><div class="table" id="release_notes_7-05_.release_notes_7_05_."><div class="caption"><p>Table 7. Discontinued Platforms</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32, 64</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2005 (C++ 8.0)<br /></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">FreeBSD-6.1</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.4.6</td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_7-05_.Caveats_and_Hint"><h2><span class="title">Caveats and Hints</span></h2><div class="sec" id="release_notes_7-05_.GCC_34x"><h3><span class="title">GCC 3.4.x</span></h3><ul><li><p>At least on Linux, <span class="nctnt ncbi-func">ifstream::readsome()</span> does not always work for large files, as it calls an ioctl that doesn't work properly for large files (we didn't test whether 4.0.x fixed this).</p></li><li><p>GCC 3.4.4 has a bug in the C++ stream library that affects some parts of our code, notably CGI framework (fixed in 4.0.1).</p></li></ul></div></div><div class="sec" id="release_notes_7-05_.Last_Updated"><h2><span class="title">Last Updated</span></h2><p>This section last updated on May 9, 2014.</p></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (Version 7, May 2011)
                    .
        [Updated: May 9, 2014]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_7-05_2011</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_7-05_.Download">Download</a></li><li><a href="#release_notes_7-05_.Third_Party_Pack">Third Party Packages</a></li><li><a href="#release_notes_7-05_.Build">Build</a></li><li><a href="#release_notes_7-05_.New_Developments">New Developments</a></li><li><a href="#release_notes_7-05_.Documentation">Documentation</a></li><li><a href="#release_notes_7-05_.Supported_Platfo">Supported Platforms (OS's and Compilers)</a></li><li><a href="#release_notes_7-05_.Caveats_and_Hint">Caveats and Hints</a></li><li><a href="#release_notes_7-05_.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (Version 7, May 2011)
                    .
        [Updated: May 9, 2014]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_7-05_2011</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_7-05_.Download">Download</a></li><li><a href="#release_notes_7-05_.Third_Party_Pack">Third Party Packages</a></li><li><a href="#release_notes_7-05_.Build">Build</a></li><li><a href="#release_notes_7-05_.New_Developments">New Developments</a></li><li><a href="#release_notes_7-05_.Documentation">Documentation</a></li><li><a href="#release_notes_7-05_.Supported_Platfo">Supported Platforms (OS's and Compilers)</a></li><li><a href="#release_notes_7-05_.Caveats_and_Hint">Caveats and Hints</a></li><li><a href="#release_notes_7-05_.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
