<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (Version 9, May 2012) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_v9_2012" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_7-05_2011.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_v9_2012.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (Version 9, May 2012)</span></h1><p class="small">Created: January 20, 2013; Last Update: May 9, 2014.</p></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="#release_notes_v9_20.release_not.Download">Download</a></p></li><li><p><a href="#release_notes_v9_20.r.Third_Party_Packag">Third Party Packages</a></p></li><li><p><a href="#release_notes_v9_20.release_notes_.Build">Build</a></p></li><li><p><a href="#release_notes_v9_20.rel.New_Developments">New Developments</a></p><ul><li><p><a href="#release_notes_v9_20.r.release_notes_HIGH">HIGHLIGHTS</a></p></li><li><p><a href="#release_notes_v9_20.release_note.CORELIB">CORELIB</a></p></li><li><p><a href="#release_notes_v9_20.release_notes.SERIAL">DATA SERIALIZATION</a></p></li><li><p><a href="#release_notes_v9_20.release_not.DATATOOL">DATATOOL</a></p></li><li><p><a href="#release_notes_v9_20.release_notes_v9.CGI">CGI</a></p></li><li><p><a href="#release_notes_v9_20.release_note.NETWORK">NETWORK</a></p></li><li><p><a href="#release_notes_v9_20.release_not.UTILITES">UTILITIES</a></p></li><li><p><a href="#release_notes_v9_20.release_n.BIOOBJECTS">BIO-OBJECTS</a></p></li><li><p><a href="#release_notes_v9_20.release_not.BIOTOOLS">BIO-TOOLS</a></p></li><li><p><a href="#release_notes_v9_20.re.ALIGNMENT_MANAGER">ALIGNMENT MANAGER</a></p></li><li><p><a href="#release_notes_v9_20.re.BIOOBJECT_MANAGER">BIO-OBJECT MANAGER</a></p></li><li><p><a href="#release_notes_v9_20.r.GENBANK_DATA_LOADE">GENBANK DATA LOADER</a></p></li><li><p><a href="#release_notes_v9_20.rele.BAM_DATA_LOADER">BAM DATA LOADER</a></p></li><li><p><a href="#release_notes_v9_20.rele.SRA_DATA_LOADER">SRA DATA LOADER</a></p></li><li><p><a href="#release_notes_v9_20.rel.cSRA_DATA_LOADER">cSRA DATA LOADER</a></p></li><li><p><a href="#release_notes_v9_20.release_notes_.BLAST">BLAST</a></p></li><li><p><a href="#release_notes_v9_20.release.APPLICATIONS">APPLICATIONS</a></p></li><li><p><a href="#release_notes_v9_20.r.BUILD_FRAMEWORK_UN">BUILD FRAMEWORK (Unix)</a></p></li></ul></li><li><p><a href="#release_notes_v9_20.releas.Documentation">Documentation</a></p><ul><li><p><a href="#release_notes_v9_20.release_not.Location">Location</a></p></li><li><p><a href="#release_notes_v9_20.release_note.Content">Content</a></p></li></ul></li><li><p><a href="#release_notes_v9_20.r.Platforms_OSs__com">Supported Platforms (OS's and Compilers)</a></p><ul><li><p><a href="#release_notes_v9_20.HYPERLINK__bookp_1">Unix</a></p></li><li><p><a href="#release_notes_v9_20.release_n.MS_Windows">MS Windows</a></p></li><li><p><a href="#release_notes_v9_20.HYPERLINK__bookp_3">Mac OS X</a></p></li><li><p><a href="#release_notes_v9_20.rele.Added_Platforms">Added Platforms</a></p></li><li><p><a href="#release_notes_v9_20.release.Discontinued">Discontinued Platforms</a></p></li></ul></li></ul><div class="sec" id="release_notes_v9_20.release_not.Download"><h2><span class="title">Download</span></h2><p>Download the source code archives at: <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/">ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/</a>  </p><ul><li><p><a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ncbi_cxx--9_0_0.tar.gz">ncbi_cxx--9_0_0.tar.gz</a> — for Unix'es (see the list of Unix flavors below) and MacOSX</p></li><li><p><a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ncbi_cxx--9_0_0.exe">ncbi_cxx--9_0_0.exe</a> — for MS-Windows (32- and 64-bit) / MSVC++ 9.0, 10.0 — self-extracting</p></li><li><p><a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ncbi_cxx--9_0_0.zip">ncbi_cxx--9_0_0.zip</a> — for MS-Windows (32- and 64-bit) / MSVC++ 9.0, 10.0</p></li></ul><p>The sources correspond to the NCBI production tree <a href="ch_getcode_svn.html#ch_getcode_svn.chkout_production_tree">sources</a>, which are originally based on the development tree source snapshot from October 25, 2010 but also include many hundreds of important and safe code updates made since then and through March 28, 2011 (and then some).</p><p>There are also two sub-directories, containing easily buildable source distributives of the NCBI C Toolkit (for MS Windows and Unix) and selected 3rd-party packages (for MS Windows only). These are the versions that the NCBI C++ Toolkit should build with. For build instructions, see README files there:</p><ul><li><p> <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/NCBI_C_Toolkit/README">NCBI_C_Toolkit </a> </p></li><li><p> <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ThirdParty/README">ThirdParty</a> </p></li></ul></div><div class="sec" id="release_notes_v9_20.r.Third_Party_Packag"><h2><span class="title">Third Party Packages</span></h2><p>Some parts of the C++ Toolkit just cannot be built without 3<sup>rd</sup> party libraries, and other parts of the Toolkit will work more efficiently or provide more functionality if some 3rd-party packages (such as BerkeleyDB which is used for local data cache and for local data storage) are available.</p><p>For more information, see the <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ThirdParty/README">FTP README</a>.</p><div class="table" id="release_notes_v9_20.r.T.currently_suppor"><div class="caption"><p>Table 1. Currently Supported/Tested Versions of Third Party Packages</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Package</th><th align="left" valign="top" rowspan="1" colspan="1">Versions expected to work (obtained by build-environment inspection in some cases)</th><th align="left" valign="top" rowspan="1" colspan="1">Versions known to work (used in-house on any platform)</th></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"> <a href="http://www.oracle.com/us/products/database/berkeley-db/index.html">BerkeleyDB</a> </td><td align="left" valign="top" rowspan="1" colspan="1">4.3.0 or newer</td><td align="left" valign="top" rowspan="1" colspan="1">4.5.20, 4.6.21.NC, 4.6.21.1, 4.7.25</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.boost.org">Boost Test</a> </td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0 or newer</td><td align="left" valign="top" rowspan="1" colspan="1">1.35.0, 1.40.0, 1.41.0, 1.42.0, 1.45.0</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.fastcgi.com">FastCGI</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">2.1, 2.4.0</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.bzip.org">libbzip2</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.0.2, 1.0.3, 1.0.5, 1.0.6</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://freshmeat.net/projects/libjpeg">libjpeg</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">6b</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.libpng.org/pub/png/libpng.html">libpng</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.7, 1.2.10, 1.2.26, 1.2.42, 1.2.46, 1.4.11</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.libtiff.org">libtiff</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.1, 3.8.2, 3.9.2, 3.9.4, 3.9.5</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://sourceforge.net/projects/giflib/files/libungif-4.x/libungif-4.1.4/">libungif</a></td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">4.1.3 (libungif),<br />4.1.6 (giflib)</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://xmlsoft.org/">libxml2</a></td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">2.6.26, 2.7.6, 2.7.8</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://xmlsoft.org/">libxslt</a></td><td align="left" valign="top" rowspan="1" colspan="1">1.1.14</td><td align="left" valign="top" rowspan="1" colspan="1">1.1.17, 1.1.24, 1.1.26</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.oberhumer.com/opensource/lzo"><span class="small-caps">LZO</span></a> </td><td align="left" valign="top" rowspan="1" colspan="1">2.x</td><td align="left" valign="top" rowspan="1" colspan="1">2.03, 2.05</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.pcre.org">PCRE</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">6.6, 7.8, 7.9</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.sqlite.org">SQLite3</a> </td><td align="left" valign="top" rowspan="1" colspan="1">3.6.6 or newer</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2, 3.6.22</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.sybase.com">Sybase</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">12.0-EBF9209, 12.5, 12.5.1</td></tr><tr><td align="left" valign="middle" rowspan="1" colspan="1"><a href="http://www.zlib.org">zlib</a> </td><td align="left" valign="top" rowspan="1" colspan="1">All versions</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.3, 1.2.3.3</td></tr></tbody></table></div></div><p>For Mac OS X and Unix OS’s, the user is expected to download and build the 3<sup>rd</sup> party packages themselves.  The release’s package list includes links to download sites. However, the user still needs a list of the 3<sup>rd</sup> party packages and which versions of them are compatible with the release.</p><p>To facilitate the building of these 3rd-party libraries on Windows, there is an <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ThirdParty/README">archive</a> that bundles together source code of the 3rd-party packages, plus MSVC "solutions" to build all (or any combination) of them. </p><div class="table" id="release_notes_v9_20.release_notes_v9_.T2"><div class="caption"><p>Table 2. Versions of Third Party Packages Included in the FTP Archive</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Package</th><th align="left" valign="top" rowspan="1" colspan="1">Depends On</th><th align="left" valign="top" rowspan="1" colspan="1">Included Version <sup>a</sup> </th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">BerkeleyDB</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">4.6.21.NC </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Boost Test</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">1.42.0</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libbzip2</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">1.0.2</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libjpeg</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">6b</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libpng</td><td align="left" valign="top" rowspan="1" colspan="1">zlib 1.2.3</td><td align="left" valign="top" rowspan="1" colspan="1">1.2.7</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libtiff</td><td align="left" valign="top" rowspan="1" colspan="1">libjpeg 6b, zlib 1.2.3</td><td align="left" valign="top" rowspan="1" colspan="1">3.6.1</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">libungif</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">4.1.3</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">LZO</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">2.05</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">PCRE</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">7.9</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">SQLite3</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">3.6.14.2</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">zlib</td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">1.2.3</td></tr></tbody></table></div></div><p> <sup>a</sup> Applies to MSVC 9, MSVC 10</p></div><div class="sec" id="release_notes_v9_20.release_notes_.Build"><h2><span class="title">Build</span></h2><p>For guidelines to configure, build and install the Toolkit see <a href="ch_config.html">here</a>.</p></div><div class="sec" id="release_notes_v9_20.rel.New_Developments"><h2><span class="title">New Developments</span></h2><div class="sec" id="release_notes_v9_20.r.release_notes_HIGH"><h3><span class="title">HIGHLIGHTS</span></h3><ul><li><p>Added UNICODE support. Internally, the Toolkit works with UTF8-encoded strings; when needed, they are converted into UNICODE and back automatically.</p></li><li><p>Sequence Align Manager API and code have been significantly revised and improved.</p></li><li><p>Added BAM, SRA and cSRA (compressed SRA) readers and data loaders – based on the now incorporated NCBI SRA Toolkit.</p></li><li><p>Object writers and readers: Added support for the VCF and GVF files formats.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_note.CORELIB"><h3><span class="title">CORELIB</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_1"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added macros <span class="nctnt ncbi-macro">BEGIN_*_NAMESPACE</span> that require semicolon and do not break indentation.</p></li><li><p>Added <span class="nctnt ncbi-func">CObjectCounterLocker::MonitorObjectType()</span> for debugging.</p></li><li><p>Added <span class="nctnt ncbi-func">CObjectCounterLocker::TransferLock()</span> and updated <span class="nctnt ncbi-func">CRef&lt;&gt;::Swap()</span> to support <span class="nctnt ncbi-func">MonitorObjectType()</span>. </p></li><li><p>Added <span class="nctnt ncbi-class">ICanceled</span> interface.</p></li><li><p>Allow locking input data in stream buffer.</p></li><li><p>Added UNICODE support. Internally, the Toolkit works with UTF8-encoded strings; when needed, they are converted into UNICODE and back automatically.</p></li><li><p>Added condition variables.</p></li><li><p>Added fast double-to-string conversion, which is 5 to 10 times faster than standard sprintf.</p></li><li><p><span class="nctnt ncbi-func">CArg(Allow)_Int8</span> -- Accept explicitly marked (0x...) hexadecimal integer argument values.</p></li><li><p><span class="nctnt ncbi-func">MemoryAdvise()</span> -- Advise on memory usage for specified memory region (Unix).</p></li><li><p><span class="nctnt ncbi-class">CPerfLogger</span>/<span class="nctnt ncbi-class">CPerfLogGuard</span> -- Performance logger. Measure time spend somewhere and put result to the performance log.</p></li><li><p><span class="nctnt ncbi-func">FileAPI::SetHonorUmask()</span> -- add configurable parameter to honor umask settings (Unix). </p></li><li><p><span class="nctnt ncbi-class">NStr::StringToSizet</span>/<span class="nctnt ncbi-class">SizetToString</span> -- special case for size_t type, thats size depends from platform/OS.</p></li><li><p><span class="nctnt ncbi-func">NStr::</span> -- added <span class="nctnt ncbi-func">[U]LongToString()</span> variants. Deprecate incompatible types for <span class="nctnt ncbi-func">[U]IntToString().</span></p></li><li><p><span class="nctnt ncbi-func">CProcess::GetCurrentHandle()</span> added</p></li><li><p><span class="nctnt ncbi-func">CDirEntry::SetOwner()</span> added</p></li><li><p>An option for <span class="nctnt ncbi-func">CProcess::Daemonize()</span> to retain parent process</p></li></ul></div><div class="sec" id="release_notes_v9_20.relea.Improvements_1"><h4><span class="title">Improvements:</span></h4><ul><li><p>Changed macro <span class="nctnt ncbi-macro">NCBI_EAT_SEMICOLON</span> to extern declaration so that uniqueness is not required anymore.</p></li><li><p>Use more efficient direct type_info comparison.</p></li><li><p>Optimized errno access in <span class="nctnt ncbi-func">NStr::StringToXxx()</span>.</p></li><li><p>Optimize MD5 step functions.</p></li><li><p>Print stack trace with critical error messages.</p></li><li><p>Optimized <span class="nctnt ncbi-func">StringToDoublePosix()</span>.</p></li><li><p>Fixed processing [+-]infinity and error detection in number parsers.</p></li><li><p>Return starting ptr if no number parsed and there are leading spaces.</p></li><li><p>Added missing method necessary for <span class="nctnt ncbi-func">CIRef::Swap()</span>.</p></li><li><p>Added more restrictive bool operator.</p></li><li><p>Enable safer iterator-dealing code if _GLIBCXX_DEBUG is defined.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Bug_fixes_1"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Fixed parsing of "+" and "-" strings in <span class="nctnt ncbi-func">NStr::StringToXxx()</span>.</p></li><li><p>Fixed sorting of <span class="nctnt ncbi-class">CVersionInfo</span> and <span class="nctnt ncbi-class">SDriverInfo</span>.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.release_notes.SERIAL"><h3><span class="title">DATA SERIALIZATION</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_2"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added support of different forms of namespace qualification as described in XML Schema.</p></li><li><p>Added mechanism of matching ASN binary data against set of known data types</p></li><li><p>Added a suite of sample programs:</p></li></ul><p><span class="nctnt ncbi-app">asn_assign</span>, <span class="nctnt ncbi-app">asnwalk_read</span>, <span class="nctnt ncbi-app">asnwalk_type</span>, <span class="nctnt ncbi-app">asnwalk_write</span>, <span class="nctnt ncbi-app">bss_info, hooks_commented hooks_copy_member</span>, <span class="nctnt ncbi-app">hooks_copy_object</span>, <span class="nctnt ncbi-app">hooks_copy_variant</span>, <span class="nctnt ncbi-app">hooks_read_member</span>, <span class="nctnt ncbi-app">hooks_read_object</span>, <span class="nctnt ncbi-app">hooks_read_variant</span>, <span class="nctnt ncbi-app">hooks_skip_member</span>, <span class="nctnt ncbi-app">hooks_skip_object</span>, <span class="nctnt ncbi-app">hooks_skip_variant</span>, <span class="nctnt ncbi-app">hooks_write_member</span>, <span class="nctnt ncbi-app">hooks_write_object</span>, <span class="nctnt ncbi-app">hooks_write_variant</span>, <span class="nctnt ncbi-app">seqannot_splicer</span></p><ul><li><p>Allow cancellation via <span class="nctnt ncbi-class">ICanceled</span> callback.</p></li><li><p>Implemented global pre-read hooks for class members and choice variants.</p></li></ul></div><div class="sec" id="release_notes_v9_20.relea.Improvements_2"><h4><span class="title">Improvements:</span></h4><ul><li><p>Set internal names for inner classes, choices, and enums.</p></li><li><p>Make error message in <span class="nctnt ncbi-func">ThrowUnassidned()</span> more informative for inner classes.</p></li><li><p>Make configurable reading/writing of UTF8String tag.</p></li><li><p>Allow ambiguous class names until lookup.</p></li><li><p>Lock input data in buffer when guessing type.</p></li><li><p>Make datatool <span class="nctnt ncbi-monospace">-t</span> argument optional, guess data type if necessary.</p></li><li><p>Reimplemented pre-read member hooks as objects instead of function pointers.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Bug_fixes_2"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Stop AnyObject reading on closing brace.</p></li><li><p>Fixed UTF8 processing in <span class="nctnt ncbi-func">CopyString()</span> and <span class="nctnt ncbi-func">SkipString()</span>.</p></li><li><p>Use UTF8String tag in ASN.1 binary.</p></li><li><p>Correctly process UTF8 strings in datatool.</p></li><li><p>Allow trailing spaces in delayed-parsing buffer.</p></li><li><p>Reset typeinfo with null object pointer to make <span class="nctnt ncbi-class">CObjectTypeInfo </span>and <span class="nctnt ncbi-class">CConstObjectInfo</span> behave equally in bool context.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_not.XMLWrapp"><h4><span class="title">XMLWrapp:</span></h4><ul><li><p>Respect the document encoding (if so) while printing or streaming nodes.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.release_not.DATATOOL"><h3><span class="title">DATATOOL</span></h3><ul><li><p>Added support of different forms of namespace qualification as described in XML Schema.</p></li><li><p>Added possibility of matching ASN binary data against set of known data types. Modified code and schema generation to handle recursive element definitions.</p></li><li><p>Added a code auto-generator for efficient traversal of records.  Also eliminates the need for much hand-generated code.  Note that the specification for autogenerated code is still in flux and is likely to change in the future.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_notes_v9.CGI"><h3><span class="title">CGI</span></h3><ul><li><p><span class="nctnt ncbi-class">CCgiResponse</span> -- Let <span class="nctnt ncbi-func">BeginPart</span> and <span class="nctnt ncbi-func">SetFilenam</span>e take optional size arguments, with non-zero values yielding RFC 2183 size parameters (which are allowed to be approximate).</p></li><li><p><span class="nctnt ncbi-class">CCgiUserAgent</span> -- new methods <span class="nctnt ncbi-func">IsBrowser()</span>, <span class="nctnt ncbi-func">GetEngineName()</span> and <span class="nctnt ncbi-func">GetPlatformName()</span>.</p></li><li><p><span class="nctnt ncbi-class">CCgiUserAgent</span> -- update list of browsers and mobile devices</p></li><li><p><span class="nctnt ncbi-class">CCgiUserAgent</span> -- implemented case insensitive parsing for user-agent strings; improved "Version/" parsing.</p></li><li><p><span class="nctnt ncbi-class">CCgiUserAgent</span> -- extended <span class="nctnt ncbi-func">IsBot()</span> and <span class="nctnt ncbi-func">IsMobileDevice()</span> methods to support exclusions lists.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_note.NETWORK"><h3><span class="title">NETWORK</span></h3><div class="sec" id="release_notes_v9_20.release.New_features"><h4><span class="title">  New features:</span></h4><ul><li><p>Identify HTTP connections with AppName in User-Agent Unicode (UTF-8) system error messages</p></li><li><p><span class="nctnt ncbi-func">SOCK_SetCork()</span> added</p></li><li><p>CONN to use new supplement (look ahead) I/O modes for faster I/O SERVICE connectors to maintain NCBI_SID when detected <span class="nctnt ncbi-class">CConn_ServiceStream</span>'s ctor to accept customizable user-header <span class="nctnt ncbi-class">CConn_PipeStream</span> to be able to get from the child's both STDOUT and STDERR output streams (merged together)</p></li><li><p><span class="nctnt ncbi-func">CONN_Status()</span> to accept eIO_Open</p></li><li><p>Flex firewall mode (per checks with the help of <span class="nctnt ncbi-class">CConnTest</span>).  Allow proxy leak configuration (to bypass the proxy if needed). CONN cancellation via callbacks. Allow bound sockets (listening and datagram) to use port "0" for OS auto-assign</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_not.Bugfixes"><h4><span class="title">  Bugfixes:</span></h4><ul><li><p><span class="nctnt ncbi-func">SOCK_Poll()</span> to return error on closed / unopened sockets Detect quotes in the CONN environment / registry settings FTP stream to issue errors on incomplete uploads</p></li><li><p><span class="nctnt ncbi-func">ConnNetInfo_ParseURL()</span> to fail for too long URL elements (no truncation)</p></li><li><p><span class="nctnt ncbi-func">ConnNetInfo_ExtendHeader()</span> to check for duplicates (and avoid them)</p></li><li><p><span class="nctnt ncbi-func">CConn_MemoryStream::ToCStr()</span> deprecated and dropped</p></li><li><p><span class="nctnt ncbi-func">CRateMonitor::GetETA()</span> fixed</p></li><li><p>Workaround IAC-unclean implementations of FTPD</p></li><li><p><span class="nctnt ncbi-func">CORE_GetUsername()</span> to avoid using <span class="nctnt ncbi-func">getlogin()</span> if at all possible</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.release_not.UTILITES"><h3><span class="title">UTILITES</span></h3><ul><li><p><span class="nctnt ncbi-class">CFormatGuess</span> -- Recognize BAM and (c)SRA files.</p></li><li><p><span class="nctnt ncbi-path">misc/clog/ncbi_c_log.h</span> -- new C logging API.</p></li><li><p>Added <span class="nctnt ncbi-func">CTar::Append()</span> on a stream object</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_n.BIOOBJECTS"><h3><span class="title">BIO-OBJECTS</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_3"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added <span class="nctnt ncbi-class">CSeq_interval</span>, <span class="nctnt ncbi-class">CSeq_loc_mix</span>, <span class="nctnt ncbi-class">CPacked_seqint</span> getters for start and stop.</p></li><li><p>Added method <span class="nctnt ncbi-func">CompareOrdered()</span> and predicate <span class="nctnt ncbi-func">PLessOrdered</span> to <span class="nctnt ncbi-class">CSeq_id_Handle</span>.</p></li></ul></div><div class="sec" id="release_notes_v9_20.relea.Improvements_3"><h4><span class="title">Improvements:</span></h4><ul><li><p><span class="nctnt ncbi-func">CSeq_loc::Compare()</span> now checks Seq-id and circular locations.</p></li><li><p>Adjusted <span class="nctnt ncbi-class">CSeq_feat</span> comparison to match C Toolkit and CFeat_CI.</p></li><li><p>Added <span class="nctnt ncbi-func">CPDB_seq_id::Compare()</span>.</p></li><li><p>Prefer gnl|ti|* over gnl|TRACE|*.</p></li><li><p>Scoring functions made non-inlined.</p></li><li><p>Added matching of integer and string versions of the same general id.</p></li><li><p>Changed Seq-table's multi-utf8-string container type to vector.</p></li><li><p>Added vector reserve in pre-read hooks for Seq-table and Dense-seg.</p></li><li><p>Added reserving pre-read hooks for Seq-graph data.</p></li><li><p>Reserve Byte-graph vector too for text formats.</p></li><li><p><span class="nctnt ncbi-func">CSeq_id::IdentifyAccession</span> -- Implement or improve recognition for more prefixes (FN-FO, GB, HV-HX, JJ-JU, KA-KZ, and LA-LJ) and mixed-in EMBL TPA protein accessions.</p></li><li><p>Rework publication labeling logic, designating the historic implementation as version 1 and adding a version 2 corresponding to GenBank/GenPept REFERENCE JOURNAL fields and the like.  One difference (among many!) between version 1 and 2 labels is that the latter generally leave off item-specific author and title information, which would appear in neighboring flat-file fields.</p></li><li><p><span class="nctnt ncbi-class">CFastaReader</span> -- Rework split sequence packing to support optionally yielding gap literals as appropriate, potentially making for considerable savings (particularly for nucleotide data).  NB: Because this feature changes the FASTA reader's traditional interpretation of data, enabling it requires supplying the new <span class="nctnt ncbi-monospace">fLetterGaps</span> flag.  Add another avenue for space saving in the form of a new <span class="nctnt ncbi-monospace">fNoUserObjs</span> flag, which disables storage of raw deflines.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Bug_fixes_3"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Fixed <span class="nctnt ncbi-func">CTextseq_id::Compare()</span> for STL requirements.</p></li><li><p>Use MT-safe way to get <span class="nctnt ncbi-class">CSeq_id</span> from <span class="nctnt ncbi-class">CSeq_id_Handle</span>.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.release_not.BIOTOOLS"><h3><span class="title">BIO-TOOLS</span></h3><div class="sec" id="release_notes_v9_20.rele.New_Development"><h4><span class="title">New Development:</span></h4><ul><li><p>Object writers and readers: Added support for the VCF and GVF files formats.</p></li></ul></div><div class="sec" id="release_notes_v9_20.r.Improvements_of_ex"><h4><span class="title">Improvements of existing object readers:</span></h4><ul><li><p>Wiggle: Added support for input files containing multiple tracks.</p></li><li><p>Wiggle: Added support for input spanning multiple chromosomes.</p></li><li><p>Bed: Chunked down huge (more than 100,000 records) tracks into a sequence of manageably sized (at most 100,000 features) feature tables.</p></li><li><p>Bed: Recognized lines beginning with “#” as comment lines.</p></li><li><p>Newick: Made node labels in Newick tree input optional.</p></li></ul></div><div class="sec" id="release_notes_v9_20.r.Improvement_of_exi"><h4><span class="title">Improvement of existing object writers:</span></h4><ul><li><p>GFF3: Added support for writing alignment files.</p></li><li><p>GFF3: Added support for a wide variety of ASN.1 attributes to be emitted as GFF3 attributes.</p></li><li><p>GFF3: Added support for escaping characters that are considered special by the GFF3 spec.</p></li><li><p>GFF3: Added support for source features.</p></li><li><p>GFF3: Added support for assembly identifying, sequence region and species pragmas.</p></li><li><p>GFF3: Added support for SOFA vocabulary.</p></li><li><p>GFF3: Added support for circular sequences.</p></li><li><p>GFF3: Added full headers identifying GFF3 revision spec and generating agent.</p></li><li><p>Wiggle: Generated more compact byte graphs by pre-analyzing data before final processing.</p></li><li><p>GTF: Made output for gene features and exon numbers optional (flag controlled).</p></li></ul></div><div class="sec" id="release_notes_v9_20.r.Other_improvements"><h4><span class="title">Other improvements:</span></h4><ul><li><p>Enhanced parsing and formatting of RepeatMasker repeat data. The improvements include greater fidelity in preserving information from the native tabular RepeatMasker output, and better standardization in representing repeat features.</p></li><li><p>Flatfile generator: </p><ul><li><p>Numerous changes to fix various incompatibilities with the C flatfile generator or bugs</p></li></ul></li><li><p>Feature Table Reader:</p><ul><li><p>More lenient parsing</p></li><li><p>Better error reporting (e.g. line number)</p></li><li><p>6th column for features to indicate minus strand</p></li></ul></li><li><p>ASN.1 Record Cleanup:</p><ul><li><p>Changed to match C cleanup.</p></li><li><p>Uses autogenerated code, generated by changes to datatool</p></li></ul></li><li><p><span class="nctnt ncbi-class">CAgpReader</span>:</p><ul><li><p>Handle AGP 2.0</p></li></ul></li><li><p>AGP Writer</p><ul><li><p>Output AGP 2.0 instead of AGP 1.1</p></li></ul></li><li><p>FASTA reader</p><ul><li><p>Can keep track of bad mods and bad residues</p></li></ul></li><li><p><span class="nctnt ncbi-class">CFastaOstream</span></p><ul><li><p>Now lets you output [key=value] modifiers if requested.  Only a few basic ones currently appear, though.</p></li></ul></li></ul></div><div class="sec" id="release_notes_v9_20.release_notes.COBALT"><h4><span class="title">COBALT</span></h4><ul><li><p>COBALT now can align two multiple sequence alignments.</p></li><li><p>COBALT now can use pre-computed RPS-BLAST alignments saved in BLAST archive format</p></li><li><p>Fixed producing alignments in NEXUS format</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.re.ALIGNMENT_MANAGER"><h3><span class="title">ALIGNMENT MANAGER</span></h3><div class="sec" id="release_notes_v9_20.relea.Improvements_4"><h4><span class="title">Improvements:</span></h4><ul><li><p>Added support for gaps/indels in pairwise, anchored and sparse alignments.</p></li><li><p>Redesigned sparse alignment iterator to show more segment types.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.re.BIOOBJECT_MANAGER"><h3><span class="title">BIO-OBJECT MANAGER</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_4"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added API to work with annot zoom tracks.</p></li><li><p>Added <span class="nctnt ncbi-func">CSeq_annot_Handle::ReorderFtable()</span>.</p></li><li><p>Implemented <span class="nctnt ncbi-class">CFeat_CI</span> iterator of features by id/xref.</p></li><li><p>Allow cancelling all tasks in prefetch manager.</p></li><li><p>Added faster <span class="nctnt ncbi-func">CScope::IsSameBioseq()</span>.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release.Improvements"><h4><span class="title">Improvements:</span></h4><ul><li><p>Implemented gap annotation mapping on segset entries.</p></li><li><p>Avoid using catch (...). Use guard class for cleanup instead of catch(...).</p></li><li><p>Order Object-id.str with numeric value together with Object-id.id.</p></li><li><p>Treat Gene-ref.desc as locus if necessary.</p></li><li><p>Use gt/lt to indicate completely cut intervals in mapped annotations.</p></li><li><p><span class="nctnt ncbi-func">CTableFieldHandle::TryGet()</span> now returns false instead of throwing an exception.</p></li><li><p>Check all mRNA features passing expensive check in <span class="nctnt ncbi-func">GetBestMrnaForCds()</span>.</p></li><li><p>Try to find mRNA parent even if gene link exists in <span class="nctnt ncbi-class">CFeatTree</span>.</p></li><li><p>Keep blob cache size to avoid calling slow list&lt;&gt;::size().</p></li><li><p>Make <span class="nctnt ncbi-func">CAnnotCollector::x_MatchLimitObject()</span> non-inlined as it's used in other file.</p></li><li><p>Use xref before overlap check in <span class="nctnt ncbi-class">CFeatTree</span>.</p></li><li><p>Allow adding chunks to active split info in MT.</p></li><li><p>Optimized <span class="nctnt ncbi-func">CTSE_Info::GetAnnotIds()</span>.</p></li><li><p>Avoid multiple overlap test for multi-id locations in <span class="nctnt ncbi-func">CFeatTree</span>.</p></li><li><p>Avoid unnecessary loading split Seq-data via <span class="nctnt ncbi-func">GetInst()</span>.</p></li><li><p>Heavily redesigned <span class="nctnt ncbi-func">Compare()</span> and <span class="nctnt ncbi-func">TestForOverlap()</span> functions to fix many problems and add more flexibility.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_no.Bug_fixes"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Update annot index before looking for Seq-feat handle.</p></li><li><p>Skip empty ranges when looking up for a feature handle.</p></li><li><p>Fixed interval matching with circular locations in <span class="nctnt ncbi-class">CHandleRangeMap</span>.</p></li><li><p>Fixed IgnoreStrand matching in <span class="nctnt ncbi-class">CHandleRangeMap</span>.</p></li><li><p>Fixed loss of features with multi-id and circular locations.</p></li><li><p>Fixed self-alignment duplication.</p></li><li><p>Fixed alignment duplication on a boundary between chunks.</p></li><li><p>Fixed <span class="nctnt ncbi-func">objects::sequence::GetStop()</span> off-by-1 error.</p></li><li><p>Do not treat location with null as multi-id.</p></li><li><p>Safe-guard against exceptions from <span class="nctnt ncbi-func">TestForOverlap()</span>.</p></li><li><p>Fixed <span class="nctnt ncbi-func">CMappedFeat::GetLocationId()</span> for mix locations.</p></li><li><p>Repeat chunk loading if necessary as they may be added later.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.r.GENBANK_DATA_LOADE"><h3><span class="title">GENBANK DATA LOADER</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_5"><h4><span class="title">New functionality:</span></h4><ul><li><p>Allow cache writer without reader.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Improvement"><h4><span class="title">Improvement:</span></h4><ul><li><p>Added workaroung to empty skeleton data when it's attached to split-info.</p></li><li><p>Use more efficient retrieval of blobs together with versions.</p></li><li><p>Added one-time check for ICache version compatibility.</p></li><li><p>Added ID2 replies' retry-delay processing.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Bug_fixes_5"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Avoid slash in cache subkey due to conflict with storing blobs in files.</p></li><li><p>Skip zoom-level annotations instead of <span class="nctnt ncbi-func">failing _ASSERTion()</span>.</p></li><li><p>Fixed loss of chunks' split and blob versions after loading from cache. </p></li><li><p>Avoid copying uninitialized iterators.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.rele.BAM_DATA_LOADER"><h3><span class="title">BAM DATA LOADER</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_6"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added support for SRZ cfg and graph.</p></li><li><p>Added map quality and annot name.</p></li><li><p>Added mate pairing getters.</p></li><li><p>Added <span class="nctnt ncbi-func">GetFlags()</span> for real BAM files.</p></li><li><p>Added configurable SPOT_ID detector.</p></li><li><p>Allow independent creation of <span class="nctnt ncbi-class">CSeq_align</span>, <span class="nctnt ncbi-class">CSeq_annot</span>, and short <span class="nctnt ncbi-class">CBioseq</span>.</p></li></ul></div><div class="sec" id="release_notes_v9_20.relea.Improvements_6"><h4><span class="title">Improvements:</span></h4><ul><li><p>Added detailed SRA SDK error reporting.</p></li><li><p>Added more CIGAR types.</p></li><li><p>Relax alignment size limit.</p></li><li><p>Allow direct BAM file name in arguments.</p></li><li><p>Convert numeric refseq ids &lt; 1000 to local Seq-id.</p></li><li><p>Added <span class="nctnt ncbi-func">CDataLoader::GetPossibleAnnotNames()</span>.</p></li><li><p>Implemented separate loading of alignments and short sequences.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Bug_fixes_6"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Fixed skipping entries with low quality.</p></li><li><p>Reverse short sequences aligned with minus strand.</p></li><li><p>Fixed soft clipping at the end of sequence.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.rele.SRA_DATA_LOADER"><h3><span class="title">SRA DATA LOADER</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_7"><h4><span class="title">New functionality:</span></h4><ul><li><p>Implemented SRA and SRZ lookup in readers.</p></li></ul></div><div class="sec" id="release_notes_v9_20.relea.Improvements_7"><h4><span class="title">Improvements:</span></h4><ul><li><p>Added explicit schema registration for XCODE and Windows.</p></li><li><p>Separated SRA, SRZ, and BAM paths.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.rel.cSRA_DATA_LOADER"><h3><span class="title">cSRA DATA LOADER</span></h3><div class="sec" id="release_notes_v9_20.New_functionalit_8"><h4><span class="title">New functionality:</span></h4><ul><li><p>Implemented cSRA reader for alignments, short reads, coverage graphs, and pileup graphs.</p></li><li><p>Added IIdMapper to cSRA reader.</p></li><li><p>Added mate read info.</p></li><li><p>Added <span class="nctnt ncbi-func">GetShortRead()</span> to get reversed read sequence.</p></li><li><p>Added <span class="nctnt ncbi-func">GetMatchAnnot()</span> to get annot with align only.</p></li><li><p>Added generation of read quality graph.</p></li><li><p>Added generation of reference sequence if necessary.</p></li></ul></div><div class="sec" id="release_notes_v9_20.relea.Improvements_8"><h4><span class="title">Improvements:</span></h4><ul><li><p>Added check for SRA value read success.</p></li><li><p>Added query arguments to test applications.</p></li><li><p>Added mate ref-id if different.</p></li><li><p>Changed local ids to a more efficient ints.</p></li><li><p>Explicit mate pair info generation made optional.</p></li><li><p>Added Seq-annot.desc for mate match indication.</p></li><li><p>Allow SRR accession to be passed to RegisterInObjectManager.</p></li><li><p>Added <span class="nctnt ncbi-func">GetIds()</span>.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release_.Bug_fixes_7"><h4><span class="title">Bug fixes:</span></h4><ul><li><p>Fixed soft clipping shift.</p></li><li><p>Fixed seq-id mapping.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.release_notes_.BLAST"><h3><span class="title">BLAST</span></h3><div class="sec" id="release_notes_v9_20.r.New_functionality_"><h4><span class="title">New functionality:</span></h4><ul><li><p>Added DELTA-BLAST - an new protein sequence alignment tool that constructs a PSSM using the results of a Conserved Domain Database search and searches a sequence database.</p></li><li><p>Added a new API <span class="nctnt ncbi-class">CDeltaBlast</span> (see above)</p></li><li><p>Added a new command line binary deltablast (see above)</p></li><li><p>Added support for bl2seq in blast_formatter</p></li><li><p>Expose psiblast command line option to ignore the MSA master sequence</p></li><li><p>Added blastdbcp, seedtop, deltablast, and makeprofiledb application</p></li><li><p>Added new FSC statistics</p></li><li><p>For more details, see BLAST+ 2.2.26 release notes.</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.release.APPLICATIONS"><h3><span class="title">APPLICATIONS</span></h3><div class="sec" id="release_notes_v9_20.r.Improvements_to_ut"><h4><span class="title">Improvements to utilities:</span></h4><ul><li><p>Multireader: Add support for Newick trees, GVF, VCF and (bare bones) text alignment files.</p></li><li><p>Annotwriter: Add support for GVF, and VCF formats.</p></li></ul></div><div class="sec" id="release_notes_v9_20.release.agp_validate"><h4><span class="title">agp_validate</span></h4><ul><li><p>AGP version 2.0. Extra checks for chromosome-from-scaffold and scaffold-from-component files (-scaf, -chr).</p></li><li><p>It is OK to not include the red-lined commits.</p></li><li><p>Added agp/fasta comparison functionality</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.r.BUILD_FRAMEWORK_UN"><h3><span class="title">BUILD FRAMEWORK (Unix)</span></h3><ul><li><p>The build system has been optimized to take advantage of GNU Make, which is now a requirement.  (Support for Solaris Make is limited to chaining to gmake.)</p></li><li><p>On Mac OS X, built bundles in addition to dynamic libraries only when explicitly given the new <span class="nctnt ncbi-monospace">--with-bundles</span> flag.</p></li><li><p>Added a <span class="nctnt ncbi-monospace">--with-relative-runpath=...</span> flag to simplify specifying executable-relative runpaths, which could be cumbersome to quote and escape (and which previous releases mishandled slightly), and a <span class="nctnt ncbi-monospace">--with-hard-runpath</span> flag to favor old dynamic tags when the choice exists.  (The default is now to use new tags, honoring LD_LIBRARY_PATH).</p></li><li><p>Added <span class="nctnt ncbi-monospace">--with(out)-symbols</span> flags to allow decoupling -g from -D_DEBUG.</p></li></ul><div class="sec" id="release_notes_v9_20.r.CHANGES_TO_COMPILE"><h4><span class="title">CHANGES TO COMPILER SUPPORT</span></h4><ul><li><p>Supported GCC versions now range from 4.0.x (released in 2005) to 4.6.x (the current stable series).  In addition, the Toolkit now supports building with LLVM Clang 3.0, at least on Mac OS X.</p></li></ul></div></div></div><div class="sec" id="release_notes_v9_20.releas.Documentation"><h2><span class="title">Documentation</span></h2><div class="sec" id="release_notes_v9_20.release_not.Location"><h3><span class="title">Location</span></h3><p>The documentation is available online as a searchable book "The NCBI C++ Toolkit": <a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/">https://www.ncbi.nlm.nih.gov/toolkit/doc/book/</a>.</p><p>The C++ Toolkit book also provides PDF version of the chapters. The PDF version can be accessed by a link that appears on each page.</p></div><div class="sec" id="release_notes_v9_20.release_note.Content"><h3><span class="title">Content</span></h3><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit" gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the newcomer.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the up-to-date public or in-house versions using source browsers <a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CTime&amp;d=">LXR</a>, <a href="https://www.ncbi.nlm.nih.gov/toolkit/?term=ctime">Doxygen</a> and <a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi?symbol=CTime">Library</a> - or do an <a href="https://www.ncbi.nlm.nih.gov/toolkitall?term=CTime">overall</a> search.</p><p> <b>HEADS-UP</b>: We have switched our source control system from CVS to SVN (Subversion). From outside NCBI it’s available:</p><ul><li><p>For browsing:  <a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++">https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++</a></p></li><li><p>For retrieval:  <a href="http://anonsvn.ncbi.nlm.nih.gov/repos/v1/trunk/c++">http://anonsvn.ncbi.nlm.nih.gov/repos/v1/trunk/c++</a> (<span class="nctnt highlight">NOTE:</span> Some WebDAV clients may require dav:// instead of http://)</p></li></ul></div></div><div class="sec" id="release_notes_v9_20.r.Platforms_OSs__com"><h2><span class="title">Supported Platforms (OS's and Compilers)</span></h2><ul><li><p><a href="release_notes_v9_2012.html#release_notes_v9_20.HYPERLINK__bookp_1">Unix</a></p></li><li><p><a href="release_notes_v9_2012.html#release_notes_v9_20.release_n.MS_Windows">MS Windows</a></p></li><li><p><a href="release_notes_v9_2012.html#release_notes_v9_20.HYPERLINK__bookp_3">Mac OS X</a></p></li><li><p><a href="release_notes_v9_2012.html#release_notes_v9_20.rele.Added_Platforms">Added</a></p></li><li><p><a href="release_notes_v9_2012.html#release_notes_v9_20.release.Discontinued">Discontinued</a></p></li></ul><p>This release was successfully tested on at least the following platforms (but may also work on other platforms). Since the previous release, some platforms were dropped from this list and some were added. Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages. In these cases, just skipping such projects (e.g. using flag "<span class="nctnt ncbi-monospace">-k</span>" for <span class="nctnt ncbi-app">make</span> on Unix), can get you through.</p><p>In cases where multiple versions of a compiler are supported, the mainstream version is shown in <b>bold</b>.</p><div class="sec" id="release_notes_v9_20.HYPERLINK__bookp_1"><h3><span class="title"><a href="ch_config.html#ch_config.UNIX">Unix</a></span></h3><div class="table" id="release_notes_v9_20.release_notes_v9_.T3"><div class="caption"><p>Table 3. Unix OS's and Supported Compilers</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">CentOS 5.4 (LIBC 2.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC 4.4.2, </b>4.0.1<sup>a</sup>, 4.1.2<sup>a</sup>, 4.3.3<sup>a</sup>, 4.6.0<sup>a</sup>, 4.6.3<sup>a</sup></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">CentOS 5.4 (LIBC 2.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC </b>4.4.5<sup> a</sup><b>, 4.6.0</b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Ubuntu 9.04 ("jaunty")  (LIBC 2.9)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32<br />x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><b>GCC 4.3.3 </b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10 </td><td align="left" valign="top" rowspan="1" colspan="1">SPARC </td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.1.1<sup>b</sup>, 4.5.3<sup>b</sup><br /><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.2.3<br /><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Solaris 10</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><b>Sun Studio 12 (C++ 5.9)</b>, Sun Studio 12 Update 1 (C++ 5.10)<sup>a</sup><br />Oracle Studio 12.2 (C++ 5.11)<sup>a</sup> </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">FreeBSD-8.0</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.2.1</td></tr></tbody></table></div></div><p> <sup>a</sup> some support</p><p> <sup>b</sup> 32-bit only </p></div><div class="sec" id="release_notes_v9_20.release_n.MS_Windows"><h3><span class="title"><a href="ch_config.html#ch_config.MS_Windows">MS Windows</a></span></h3><div class="table" id="release_notes_v9_20.release_notes_v9_.T4"><div class="caption"><p>Table 4. MS Windows and Supported Compilers</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2008 (C++ 9.0), 2010 (C++ 10.0) <br />NOTE: We also ship an easily buildable archive of <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/ThirdParty/README">3rd-party</a> packages (including <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/ARCHIVE/9_0_0/NCBI_C_Toolkit/README">NCBI C Toolkit</a>) for this platform.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2008 (C++ 9.0), 2010 (C++ 10.0) <br />NOTE: We also ship an easily buildable archive of <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/2010/Jun_15_2010/ThirdParty/README">3rd-party</a> packages (including <a href="ftp://ftp.ncbi.nlm.nih.gov/toolbox/ncbi_tools++/2010/Jun_15_2010/NCBI_C_Toolkit/README">NCBI C Toolkit</a>) for this platform, although not all libraries are available for x86-64 architecture.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Cygwin 1.5.25</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.Cygwin_GCC">GCC 4.5.3</a>- nominal support only.</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_v9_20.HYPERLINK__bookp_3"><h3><span class="title"><a href="ch_config.html#ch_config.Mac_OS_X">Mac OS X</a></span></h3><div class="table" id="release_notes_v9_20.release_notes_v9_.T5"><div class="caption"><p>Table 5. Mac OS and Supported Compilers</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Mac OS X 10.5, <br />MacOS x 10.6</td><td align="left" valign="top" rowspan="1" colspan="1">Native (PowerPC <u>or</u> x86-32<u> or</u> x86-64 )</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.Xcode_30__31">Xcode</a>  3.0 - 3.2.6</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Darwin 8.x, 9.x, 10.x</td><td align="left" valign="top" rowspan="1" colspan="1">Native (PowerPC <u>or</u> x86-32<u> or</u> x86-64),<br />Universal (PowerPC <u>and</u> x86-32)</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.GCC">GCC 4.0.1</a> <br /><a href="ch_config.html#ch_config.GCC">GCC 4.2.1</a> (only available under Darwin 10.x)<br />Clang 3.0</td></tr></tbody></table></div></div><p><span class="nctnt highlight">NOTE:</span>  the correspondence between Darwin kernel versions and Mac OS versions:</p><p>Darwin  8.x = Mac OS 10.4.x</p><p>Darwin  9.x = Mac OS 10.5.x</p><p>Darwin 10.x = Mac OS 10.6.x</p></div><div class="sec" id="release_notes_v9_20.rele.Added_Platforms"><h3><span class="title">Added Platforms</span></h3><div class="table" id="release_notes_v9_20.release_notes_v9_.T6"><div class="caption"><p>Table 6. Added Platforms</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32, x86-64</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2010 (C++ 10.0) </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">CentOS 5.4 (LIBC 2.5)</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 4.4.5<sup> a</sup><b>, </b>4.6.0</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">Mac OS X 10.5, <br />MacOS x 10.6</td><td align="left" valign="top" rowspan="1" colspan="1">Native (PowerPC <u>or</u> x86-32<u> or</u> x86-64 )</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.Xcode_30__31">Xcode</a>  3.2.3 - 3.2.6<br />Clang 3.0</td></tr></tbody></table></div></div><p><sup>a</sup> some support</p></div><div class="sec" id="release_notes_v9_20.release.Discontinued"><h3><span class="title">Discontinued Platforms</span></h3><div class="table" id="release_notes_v9_20.release_notes_v9_.T7"><div class="caption"><p>Table 7. Discontinued Platforms</p></div><div class="table-scroll"><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">MS Windows</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32, 64</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_config.html#ch_config.MS_Visual_C_2008">MS Visual C++</a> 2005 (C++ 8.0)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">FreeBSD-6.1</td><td align="left" valign="top" rowspan="1" colspan="1">x86-32</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.4.6</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">All</td><td align="left" valign="top" rowspan="1" colspan="1">All</td><td align="left" valign="top" rowspan="1" colspan="1">All GCC 4.0.1 and below</td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_v9_20.release.Last_Updated"><h2><span class="title">Last Updated</span></h2><p>This section last updated on May 9, 2014.</p></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (Version 9, May 2012)
                    .
        [Updated: May 9, 2014]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_v9_2012</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_v9_20.release_not.Download">Download</a></li><li><a href="#release_notes_v9_20.r.Third_Party_Packag">Third Party Packages</a></li><li><a href="#release_notes_v9_20.release_notes_.Build">Build</a></li><li><a href="#release_notes_v9_20.rel.New_Developments">New Developments</a></li><li><a href="#release_notes_v9_20.releas.Documentation">Documentation</a></li><li><a href="#release_notes_v9_20.r.Platforms_OSs__com">Supported Platforms (OS's and Compilers)</a></li><li><a href="#release_notes_v9_20.release.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (Version 9, May 2012)
                    .
        [Updated: May 9, 2014]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_v9_2012</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_v9_20.release_not.Download">Download</a></li><li><a href="#release_notes_v9_20.r.Third_Party_Packag">Third Party Packages</a></li><li><a href="#release_notes_v9_20.release_notes_.Build">Build</a></li><li><a href="#release_notes_v9_20.rel.New_Developments">New Developments</a></li><li><a href="#release_notes_v9_20.releas.Documentation">Documentation</a></li><li><a href="#release_notes_v9_20.r.Platforms_OSs__com">Supported Platforms (OS's and Compilers)</a></li><li><a href="#release_notes_v9_20.release.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
