<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (July 8, 2004) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_7_8_2004" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_10_2_2004.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_april_16_2004.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_7_8_2004.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (July 8, 2004)</span></h1></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.download_loc_7_8_">Download Location</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.source_archive_7_">Source Archive Contents</a>
					<ul><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.source_code_7_8_2">Source Code Archives</a></p></li></ul>
				</p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.new_development_7">New Development</a>
					<ul><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.corelib_7_8_2004">CORELIB -- Portability and Application Framework</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.connect_7_8_2004">CONNECT -- Data streaming, Networking, and Dispatching</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.html_7_8_2004">HTML -- HTML Generation Library</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.util_7_8_2004">UTIL -- Miscellaneous Low-Level APIs</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.util_comp_7_8_200">UTIL/COMPRESSION -- Data Compression (GZIP, BZ2)</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.serial_7_8_2004">SERIAL -- Data Serialization (ASN.1, XML)</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.datatool_7_8_2004">DATATOOL -- Code Generator and Data Converter Tool</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.cgi_7_8_2004">CGI  --  CGI and Fast-CGI Application Framework</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.berkeley_dbapi_7_">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.dbapi_7_8_2004">DBAPI -- Generic SQL Database Connectivity</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.algo_7_8_2004">ALGO -- Advanced Algorithms Library</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.algoalign_7_8_200">ALGO/ALIGN -- Generic Alignment Algorithms</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.blast_7_8_2004">BLAST</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.algnmgr_7_8_2004">ALNMGR -- Bio-sequence Alignment Manager</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.bioobj_7_8_2004">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.om_7_8_2004">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.omlr_7_8_2004">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.id2_7_8_2004">ID2</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.biotools_7_8_2004">BIO-TOOLS</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.lds_7_8_2004">LDS</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.bf_7_8_2004">BUILD FRAMEWORK</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.apps_7_8_2004">APPLICATIONS</a></p></li></ul>
				</p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.documentation_7_8">Documentation</a>
					<ul><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.doc_loc_7_8_2004">Document Location</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.doc_changes_7_8_2">Document Content</a></p></li></ul>
				</p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.build_macos_7_8_2">Building on the MacOS</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.platforms_7_8_200">Platforms (OS's, compilers used inside NCBI)</a>
					<ul><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.platform_unix_7_8">Unix</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.platform_windows_">MS Windows</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.platform_mac_7_8_">Mac OS X</a></p></li></ul>
				</p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.caveats_7_8_2004">Caveats and Hints</a>
					<ul><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.macos_10x_7_8_200">MacOS 10.X / CodeWarrior 9.1</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.macos_10_X_7_8_20">MacOS 10.2/GCC 3.3</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.gcc_2_95_3_7_8_20">GCC 2.95</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.gcc_3_0_4_7_8_200">GCC 3.0.4</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.gcc_3_3_7_8_2004">GCC 3.3</a></p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.gcc_3_4_7_8_2004">GCC 3.4</a></p></li></ul>
				</p></li><li><p><a href="release_notes_7_8_2004.html#release_notes_7_8_2004.last_update_7_8_2">Last Updated</a></p></li></ul><div class="sec" id="release_notes_7_8_2004.release_notes7_8_2004.content"><h2><span class="title">Release Notes (July 8, 2004)</span></h2><div class="sec" id="release_notes_7_8_2004.download_loc_7_8_"><h3><span class="title">Download Location</span></h3><p>
				<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2004/Jul_08_2004/">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2004/Jul_08_2004/</a>
			</p></div><div class="sec" id="release_notes_7_8_2004.source_archive_7_"><h3><span class="title">Source Archive Contents</span></h3><div class="sec" id="release_notes_7_8_2004.source_code_7_8_2"><h4><span class="title">Source Code Archives</span></h4><ul><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_unix--Jul_08_2004.tar.gz</span>-- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_unix--Jul_08_2004.gtar.gz</span>-- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win--Jul_08_2004.exe</span>-- for MS-Windows: MSVC++ 6.0 and 7.1 (self-extracting)</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win--Jul_08_2004.zip</span>-- for MS-Windows: MSVC++ 6.0 and 7.1</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win--Jul_08_2004.tar.gz</span>-- for MS-Windows: MSVC++ 6.0 and 7.1</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_cw--Jul_08_2004.tar.gz</span>-- for MacOS 10.3.4 / CodeWarrior DevStudio for MacOS 9.2</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_gcc--Jul_08_2004.tar.gz</span>-- for MacOS 10.2, 10.3.4 / GCC 3.3</p></li></ul><p>
					<span class="nctnt highlight">The sources correspond to the NCBI production tree sources from patch "GCC34_MSVC7," which in turn corresponds to the development tree sources from around end of may, 2004.</span>
				</p></div></div><div class="sec" id="release_notes_7_8_2004.new_development_7"><h3><span class="title">New Development</span></h3><p>
				<span class="nctnt highlight">Some of the new development may introduce potentially backward-incompatible changes.</span>
			</p><div class="sec" id="release_notes_7_8_2004.corelib_7_8_2004"><h4><span class="title">CORELIB -- Portability and Application Framework</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CRef&lt;&gt;</span>-- added methods for atomic release and reset.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-type">EDiagPostFlag</span>-- added optional conversion of multi-line diagnostic messages into single-line ones.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CDirEntry</span>-- the methods that support determining entries' types now let the caller specify how to treat non-dangling symlinks.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-type">CFile::ETmpFileCreationMode</span>-- to control whether to actually create the temporary file (to get rid of a possible race condition).</p></dd><dt>5</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CDir</span>-- new method<span class="nctnt ncbi-func">GetTmpDir()</span>to get path to the temporary directory; old method<span class="nctnt ncbi-func">GetEntries()</span>now has an additional parameter to ignore self recursive directory entries (".", "..").</p></dd><dt>6</dt><dd><p class="no_top_margin">The generic<span class="nctnt ncbi-app">FindFile</span>algorithm -- now can do recursive searches.</p></dd><dt>7</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CTime</span>-- new formatting symbol 'z' (local time in format GMT{+|-}HHMM). Also added output timezone parameter to<span class="nctnt ncbi-func">CTime::AsString()</span>.</p></dd><dt>8</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CProcess</span>-- added<span class="nctnt ncbi-func">WaitForAlive()</span>and<span class="nctnt ncbi-func">WaitForTerminate()</span>methods.</p></dd><dt>9</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CPIDGuard</span>-- now does smart reference counting in the PID guard file to avoid confusion with multiple locks;  also added<span class="nctnt ncbi-func">CPIDGuard::Remove()</span>method to force the removal of PID guard file.</p></dd><dt>10</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CPluginManager</span>-- new methods to better control DLL resolution:<span class="nctnt ncbi-func">FreezeResolution()</span>and<span class="nctnt ncbi-func">DetachResolver()</span>.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.connect_7_8_2004"><h4><span class="title">CONNECT -- Data streaming, Networking, and Dispatching</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CNamedPipe</span>-- replaced static const with enum for the default Toolkit and system pipe buffer size.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CONN_ReadLine()</span>-- new function, to read from CONN line-by-line.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.html_7_8_2004"><h4><span class="title">HTML -- HTML Generation Library</span></h4><p>
					<span class="nctnt ncbi-class">CHTMLPopupMenu</span>-- added support for Sergey Kurdin's popup menu with configurations, extended the API to account for the new menu features and functionality.</p></div><div class="sec" id="release_notes_7_8_2004.util_7_8_2004"><h4><span class="title">UTIL -- Miscellaneous Low-Level APIs</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">"Bitset" API -- the NCBI adaptation of the BitMagic template library to manipulate with compressed bitsets (by BitMagic author).</p></dd><dt>2</dt><dd><p class="no_top_margin">Added UNICODE/UTF8 conversion functions.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CResourcePool</span>- added access to the list of available objects.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CResourcePool</span>- added access to the list of available objects.</p></dd><dt>5</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">C[RW]Stream[buf]</span>-- added optional ownership of the underlying IReader/IWriter-derived objects to allow for their automagic destruction.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.util_comp_7_8_200"><h4><span class="title">UTIL/COMPRESSION -- Data Compression (GZIP, BZ2)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CZipDecompressor</span>-- now can recognize and skip GZIP file format header.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CCompressionStream</span>-- added optional ownership of the underlying C++ iostreams and/or compression processors to allow for their automatic destruction.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CCompression[IO]Stream</span>-- new methods GetProcessedSize() and GetOutputSize() to get, respectively, the amount of the (de)compressed and original data processed so far.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.serial_7_8_2004"><h4><span class="title">SERIAL -- Data Serialization (ASN.1, XML)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Context-sensitive serialization hooks -- now, can be set for the given read context, using a pattern string, with wildcards allowed: "TypeName.Member1.Member2.HookedMember".</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">CObjectIStream::SetStreamOffset()</span>-- to rollback from the current position of input stream, if possible.</p></dd><dt>3</dt><dd><p class="no_top_margin">Improved XML serialization, added "Windows-1252" encoding.</p></dd><dt>4</dt><dd><p class="no_top_margin">Implemented optional skipping of unknown data members in XML and ASN streams.</p></dd><dt>5</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">MSerial_{AsnText,AsnBinary,Xml}</span>-- new I/O manipulators for regular C++ iostreams to simplify the serialization of serial data objects.</p></dd><dt>6</dt><dd><p class="no_top_margin">Implemented non-recursive copying and comparing of serial objects.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.datatool_7_8_2004"><h4><span class="title">DATATOOL -- Code Generator and Data Converter Tool</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added the type conversion check (when it is defined in DEF file) to prohibit inappropriate conversions.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added optional generation of doxygen-style comments and inserting of pre-compiled headers when generating C++ code.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.cgi_7_8_2004"><h4><span class="title">CGI -- CGI and Fast-CGI Application Framework</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Turned off background exception reporting, by default.</p></dd><dt>2</dt><dd><p class="no_top_margin">Convert multi-line diagnostic messages into one-line ones, by default.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CCgiRedirectApplication</span>-- use "&amp;{var}" rather than "$var" to refer to the value of the variable;  also, all new string values in the registry file must be URL-encoded now.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.berkeley_dbapi_7_"><h4><span class="title">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Query parser and interpreter -- implemented NOT and comparison operators.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added support for Berkeley DB transactions.</p></dd><dt>3</dt><dd><p class="no_top_margin">Implemented transaction protected BLOB Stream.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">ICache</span>-- improved stability, added transaction protection, greatly reduced probability of file corruption.</p></dd><dt>5</dt><dd><p class="no_top_margin">Implemented transaction protected cursors.</p></dd><dt>6</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">CBDB_Env::JoinEnv</span>improved to support several program instances working with the same database.</p></dd><dt>7</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBDB_FieldLString</span>-- new field type to support length prefixed strings (for better compatibility with<span class="nctnt ncbi-class">std::string</span>class).</p></dd><dt>8</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBDB_Env</span>-- added error logging (<span class="nctnt ncbi-func">OpenErrFile</span>).</p></dd><dt>9</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">BDB_find_field</span>-- new search method using<span class="nctnt ncbi-class">CBoyerMooreMatcher</span>search condition.</p></dd><dt>10</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">CBDB_Cache::Open()</span>-- added RAM cache size parameter.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.dbapi_7_8_2004"><h4><span class="title">DBAPI -- Generic SQL Database Connectivity</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">FreeTDS -- the bundled FreeTDS code now supports using<span class="nctnt ncbi-func">getaddrinfo()</span>and<span class="nctnt ncbi-func">getnameinfo()</span>, enabling multithreaded builds on platforms that lack<span class="nctnt ncbi-func">gethostbyname_r()</span>et al.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">IConnection</span>-- new public methods<span class="nctnt ncbi-func">GetStatement()</span>,<span class="nctnt ncbi-func">GetCallableStatement()</span>,<span class="nctnt ncbi-func">GetCursor()</span>and<span class="nctnt ncbi-func">GetBulkInsert()</span>. All objects obtained by calling these methods work within a single connection, opened by<span class="nctnt ncbi-func">IConnection::Open()</span>and cannot be used simultaneously. Don't mix them with the old<span class="nctnt ncbi-func">Create()</span>methods, otherwise an exception will be thrown.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">IStatement</span>-- new method<span class="nctnt ncbi-func">PurgeResults()</span>, to purge all results of the given query.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">IStatement</span>-- new method<span class="nctnt ncbi-func">ExecuteQuery()</span>, to return the first resultset of the given query and to discard any additional results.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.algo_7_8_2004"><h4><span class="title">ALGO -- Advanced Algorithms Library</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Tree specific algorithms and manipulation methods -- added<span class="nctnt ncbi-func">TreeFindCommonParent</span>,<span class="nctnt ncbi-func">TreeTraceToRoot</span>,<span class="nctnt ncbi-func">TreeReRoot</span>,<span class="nctnt ncbi-func">MoveSubnodes</span>,<span class="nctnt ncbi-class">CTreeMinimalSet</span>, operations on node lists, etc.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBioTree</span>- new, advanced tree container for phylogenetic and taxonomy trees.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.algoalign_7_8_200"><h4><span class="title">ALGO/ALIGN -- Generic Alignment Algorithms</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">XALGOSPLIGN -- new library to share the state-of-the-art        transcript alignment functionality.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSplign</span>--  provides access to all parameters of the XALGOSPLIGN algorithm and controls the algorithm's execution. To use<span class="nctnt ncbi-class">CSplign</span>directly, BLAST hits must be computed externally and a sequence accessor class must be provided (derived from class<span class="nctnt ncbi-class">CSplignSeqAccessor</span>).</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSplignSimple</span>-- a lightweight version of<span class="nctnt ncbi-class">CSplign</span>, which needs only two<span class="nctnt ncbi-class">CSeq_loc</span>and a<span class="nctnt ncbi-class">CScope</span>object to access sequences, and does BLAST computation internally.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSplignFormatter</span>-- class to format output produced by<span class="nctnt ncbi-class">CSplign</span>or<span class="nctnt ncbi-class">CSplignSimple</span>as a tab-separated list of exons and gaps (unaligned cDNA regions) or as a<span class="nctnt ncbi-class">CSeq_align_set</span>object.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.blast_7_8_2004"><h4><span class="title">BLAST</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Introduction of remote BLAST API and demo program.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added support for RPS-BLAST.</p></dd><dt>3</dt><dd><p class="no_top_margin">Merged TwoSequences and Database engines.</p></dd><dt>4</dt><dd><p class="no_top_margin">Complete rewrite of low-level gapped alignment functions.</p></dd><dt>5</dt><dd><p class="no_top_margin">Cleaned up and modularized API for saving HSPs, HSPLists and HitLists</p></dd><dt>6</dt><dd><p class="no_top_margin">Refactoring of link hsps code to make it more self-contained.</p></dd><dt>7</dt><dd><p class="no_top_margin">Addition of<span class="nctnt ncbi-class">CSeqDb</span>and multiple sequences implementations of the BlastSeqSrc API.</p></dd><dt>8</dt><dd><p class="no_top_margin">Introduction of the BlastDiagnostics API in CORE BLAST.</p></dd><dt>9</dt><dd><p class="no_top_margin">Fixed various naming conflicts with the NCBI C Toolkit..</p></dd><dt>10</dt><dd><p class="no_top_margin">Introduction of the BlastHspStream abstraction.</p></dd><dt>11</dt><dd><p class="no_top_margin">Introduction of PSSM calculation engine.</p></dd><dt>12</dt><dd><p class="no_top_margin">Unify handling of gapped Karlin blocks for protein and nucleotide searches.</p></dd><dt>13</dt><dd><p class="no_top_margin">Consolidated validation of all options structures in one function in CORE BLAST.</p></dd><dt>14</dt><dd><p class="no_top_margin">Added check for word size value in options validation for MegaBLAST:  word size must be &gt;=12 for contiguous MegaBLAST, and can be 11 or 12 for discontiguous MegaBLAST.</p></dd><dt>15</dt><dd><p class="no_top_margin">Added setting of a most reasonable database scanning stride value dependent on word size W and lookup table width L. The best stride is calculated as W - L + 1 or as W - L + 4 if the "variable word size" is set to true.  However the resulting stride is rounded to a number divisible by 4 for all values except 6 and 7. This is done because scanning database with stride divisible by 4 does not require splitting bytes of compressed sequences. For values 6 and 7 however the advantage of a larger stride outweighs the disadvantage of splitting the bytes.</p></dd><dt>16</dt><dd><p class="no_top_margin">Started separation of preliminary and traceback stages in CORE BLAST.</p></dd><dt>17</dt><dd><p class="no_top_margin">Rewrote BlastKarlinLHtoK..</p></dd><dt>18</dt><dd><p class="no_top_margin">Port of RedoAlignmentCore from the NCBI C toolkit to Kappa_RedoAlignmentCore.</p></dd><dt>19</dt><dd><p class="no_top_margin">Provide an interface to retrive the matrices' frequency ratios.</p></dd><dt>20</dt><dd><p class="no_top_margin">Modified ungapped protein alignment code to handle PSSMs, optimizations to the construction of the protein lookup table.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.algnmgr_7_8_2004"><h4><span class="title">ALNMGR -- Bio-sequence Alignment Manager</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CAlnMap</span>-- added<span class="nctnt ncbi-func">GetResidueIndexMap()</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CAlnVec</span>-- added<span class="nctnt ncbi-func">GetColumnVector()</span>and<span class="nctnt ncbi-func">CalculatePercentIdentity()</span>; made<span class="nctnt ncbi-class">CScope</span>a required parameter.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CAlnMix</span>-- added forcing translation of nucleotide sequences; emade<span class="nctnt ncbi-class">CScope</span>a required parameter.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.bioobj_7_8_2004"><h4><span class="title">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Changed type of Seq-align containers from list&lt;&gt; to vector&lt;&gt; for better performance.</p></dd><dt>2</dt><dd><p class="no_top_margin">Moved<span class="nctnt ncbi-func">CSeq_feat::CompareLocations()</span>to<span class="nctnt ncbi-func">CSeq_loc::Compare</span>
						</p></dd><dt>3</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-func">FindGi()</span>,<span class="nctnt ncbi-func">FindTextseq_id()</span>and<span class="nctnt ncbi-func">GetSeq_idByType()</span>to simpify ID lookups in containers.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeq_id</span>-- support several new prefixes in<span class="nctnt ncbi-func">IdentifyAccession()</span>method.</p></dd><dt>5</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeq_id</span>-- improved support for ranking IDs by utility.</p></dd><dt>6</dt><dd><p class="no_top_margin">New flat-file generator (`xformat` in "objects/format"). The previous generator (in "objects/flat") is no longer in active development.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.om_7_8_2004"><h4><span class="title">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CDesc_CI</span>-- renamed to<span class="nctnt ncbi-class">CSeq_descr_CI</span>for consistency with the ASN.1 class naming.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CAnnotTypes_CI</span>-- is no longer inherited from SAnnotSelector, but accepts it as an argument.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">SAnnotSelector</span>-- renamed methods of from *Choice to *Type.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CAlign_CI</span>-- implemented alignment mapping.</p></dd><dt>5</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CFeat_CI</span>-- now, the annotations can be filtered by a combination (with both inclusions and exclusions) of types and subtypes.</p></dd><dt>6</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeqdesc_CI</span>-- now, more than one type can be used for the filtering.</p></dd><dt>7</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CScope</span>-- allow to fine-tune the getting of BioseqHandle depending on resolution flags.</p></dd><dt>8</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CScope</span>-- new method GetAllTSEs to get all already resolved TSEs.</p></dd><dt>9</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeqVector</span>-- added ncbi2na ambiguity randomizer.</p></dd><dt>10</dt><dd><p class="no_top_margin">Significantly improved performance of annotation collecting and mapping functions.</p></dd><dt>11</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CPrefetchToken</span>-- new, to provides background data pre-fetching.</p></dd><dt>12</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeq_loc_Mapper</span>-- new, to map locations and alignments.</p></dd><dt>13</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeq_{feat,align,graph}_Handle</span>-- handles for annotation types, to be used instead of "real" bio-objects.</p></dd><dt>14</dt><dd><p class="no_top_margin">OM++ headers -- reduced inter-dependency. Some code may require to include headers which were included implicitly before.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>15</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBioseq_set_Handle</span>and set of<span class="nctnt ncbi-class">C*_EditHandle's</span>-- added for working with different bioseq objects (<span class="nctnt ncbi-class">CBioseq</span>,<span class="nctnt ncbi-class">CBioseq_set</span>,<span class="nctnt ncbi-class">CSeq_entry</span>, and<span class="nctnt ncbi-class">CSeq_annot</span>).</p></dd><dt>16</dt><dd><p class="no_top_margin">A start-up edit API for bioseq objects through various edit handles.<span class="nctnt ncbi-class">ConvertSeqToSet()</span>and<span class="nctnt ncbi-class">ConvertSetToSeq()</span>for<span class="nctnt ncbi-class">CSeq_entry</span>. Methods to move Bioseq withing Seq-entry. Implemented<span class="nctnt ncbi-func">MoveTo</span>and<span class="nctnt ncbi-func">CopyTo</span>methods for various handles.</p></dd><dt>17</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">CScope::GetSynonyms()</span>-- do not try to resolve any Seq-ids, just assume that any unresolved Seq-id is a synonym. Seq-id conflict messages made clearer.</p></dd><dt>18</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBioseq_Handle</span>-- new method<span class="nctnt ncbi-func">GetExactComplexityLevel()</span>.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.omlr_7_8_2004"><h4><span class="title">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CReader</span>-- revamped to use<span class="nctnt ncbi-class">CPluginManager</span>.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>2</dt><dd><p class="no_top_margin">GenBank loader and its readers -- moved sources to new location (in OBJTOOLS). Libraries with the OM++ data loaders renamed to "libncbi_xloader_*". Libraries with the OM++ GenBank data loader's readers renamed to "libncbi_xreader_*". OBJMGR_LIBS macro updated correspondingly.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
						</p></dd><dt>3</dt><dd><p class="no_top_margin">GenBank data readers -- added support for TRACEs: Seq-id ::= general { db "ti", tag id NNN }.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CCachedId1Reader</span>-- now supports new<span class="nctnt ncbi-class">ICache</span>interface.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.id2_7_8_2004"><h4><span class="title">ID2</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">ID2 ASN.1 specification file was split into two parts. Now, the ID2 communication protocol is specified in "id2.asn" module, while the ID2 split data structure specifications are moved to "seqsplit.asn" module.</p></dd><dt>2</dt><dd><p class="no_top_margin">Moved ID2 and ID1 specific code out of object manager. Protocol is processed by corresponding readers. ID2 split parsing is processed by ncbi_xreader library - used by all GenBank readers.</p></dd><dt>3</dt><dd><p class="no_top_margin">Various parts of the ID2 communication protocol and the ID2 split data structures have been modified;  these are still changing, albeit at much slower rate, and are expected to by and large stabilize by around mid-fall.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.biotools_7_8_2004"><h4><span class="title">BIO-TOOLS</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">GFF/GTF files -- added [semi-experimental] support for reading these.</p></dd><dt>2</dt><dd><p class="no_top_margin">CObjectsSniffer -- bug fixes and perfomance optimizations.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.lds_7_8_2004"><h4><span class="title">LDS</span></h4><p>Implemented sequence search using query (using BDB query).</p></div><div class="sec" id="release_notes_7_8_2004.bf_7_8_2004"><h4><span class="title">BUILD FRAMEWORK</span></h4><p>The Toolkit can now take advantage of preexisting installations of zlib, bz2lib, and libpcre, though it still includes copies of all three libraries in order to maintain support for systems that lack them.  You may need to adjust some of your makefiles accordingly.<span class="nctnt highlight">May introduce potentially backward-incompatible changes.</span>
				</p></div><div class="sec" id="release_notes_7_8_2004.apps_7_8_2004"><h4><span class="title">APPLICATIONS</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">ASN2FLAT and CONVERT_SEQ -- new applications to support conversion between various common representations of annotated biosequence data.</p></dd><dt>2</dt><dd><p class="no_top_margin">ID1_FETCH_SIMPLE -- added possibility to send any ASN.1 request to ID1 (for testing purposes).</p></dd><dt>3</dt><dd><p class="no_top_margin">SPLIGN -- improved quality of transcript alignments. Added a powerful compartment identification algorithm (implemented within the XALGOSPLIGN library) to identify pseudogenes and paralogs. The updated output format includes the compartment identifier preceded by strand sign.</p></dd></dl></div></div><div class="sec" id="release_notes_7_8_2004.documentation_7_8"><h3><span class="title">Documentation</span></h3><div class="sec" id="release_notes_7_8_2004.doc_loc_7_8_2004"><h4><span class="title">Document Location</span></h4><p>The documentation is available online at<!--ro--><a href="toc.html">https://www.ncbi.nlm.nih.gov/books/bv.fcgi?call=bv.View..ShowTOC&amp;rid=toolkit.TOC&amp;depth=2</a>as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release. The PDF version can be accessed by a link that appears on each page.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and  "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_7_8_2004.doc_changes_7_8_2"><h4><span class="title">Document Content</span></h4><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit"  gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
					<a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/">https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
				</p><p>The above link is also available under the "Source Code Browsers" that appears on each page.</p><p>A C/C++ Symbol Search query has been added to each page of the online Toolkit documentation. You can use this to perform a symbol search on the public or in-house versions of LXR, Doxygen and Library. The Library search runs a CGI script that lists the library name in which a symbol occurs.</p><p>The current release notes as well as past release notes are now in an appendix in the C++ Toolkit Book and a link to the current release notes appears on each page of the online Toolkit document.</p></div></div><div class="sec" id="release_notes_7_8_2004.build_macos_7_8_2"><h3><span class="title">Building on the MacOS</span></h3><p>We now build all the libraries and most of the applications including the Genome Workbench (<span class="nctnt ncbi-app">gbench</span>).  We do not build any of the many test or demo apps (except<span class="nctnt ncbi-app">testvalidator</span>).  We also build the FLTK library's GUI editor,<span class="nctnt ncbi-app">fluid</span>.</p><p>All apps are built as application bundles except<span class="nctnt ncbi-app">gbench_plugin_scan</span>and<span class="nctnt ncbi-app">datatool</span>which are built as command line apps.  Any of the applications can be built as command line apps by tweaking the build scripts or the Codewarrior projects.</p><p>Build procedure is still the same: use an AppleScript editor to open and run the script files makeLibs.met and makeApps.met.  You must use a script editor capable of opening a script file larger than 32K, such as Apple's Script Editor v2.0, or Smile.  Script Editor v1.9 will not work since makeLibs.met just got too big.  The command line tool osascript also works.</p><p>Projects include targets to compile with BSD/Apple headers and libraries and with MSL headers and libraries, but plugins (for gbench) built with MSL do not link properly, and so, because of lack of interest to keep up with changes, some source code does not currently  compile with MSL.  Hopefully this will become easier to work with and get fixed with Codewarrior v.9.</p><p>Targets to be compiled can be controlled by including an empty file or folder in the compilers:mac_prj folder with the name 'Build' followed by the keywords of the targets you want built.  The keywords are: MSL, BSD, Debug and Final.  For example, to build only the BSD Debug targets use: "Build BSD Debug", to build both BSD debug and release (final) versions: "Build BSD".  The default is to build everything.</p><p>If you install the C++ toolkit under a different name than "ncbi_cxx" or in a different location than your home directory, you can edit the  script's properties, pRootFolderName and pRootFolderPath, to override these defaults.  Note: these paths, and those mentioned below, must be entered in mac format (e.g. disk:Users:username:) not Unix format (e.g.<span class="nctnt ncbi-path">/Users/username/</span>).  The disk name (and its following colon) may be omitted.</p><p>Certain third party libraries (see<a href="release_notes_7_8_2004.html#release_notes_7_8_2004.T1">Table 10</a>) are required to build the C++ toolkit. The scripts will try and find them if they are in your home directory, or you can specify where they were installed using properties at the beginning of the script.</p><div class="table" id="release_notes_7_8_2004.T1"><h4><span class="label">Table 10</span><span class="title">Third Party Libraries</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Library</th><th align="left" valign="top" rowspan="1" colspan="1">Property</th><th align="left" valign="top" rowspan="1" colspan="1">Example</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">FLTK 1.1.5rc2 (w/ NCBI patches)</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pFLTKRootFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"usr:local:src:fltk-1.1.5rc2:"</span>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">BerkeleyDB 4.2.52</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pBdbRootFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"Users:myhome:mylibs:db-4.2.52"</span>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">SQLite 2.8.15</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">gSqliteFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">

</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">dlcompat</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pDLRootFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"usr:"</span>
							</td></tr></tbody></table></div></div><p>You do not have to build the FLTK or BDB libraries separately.  This is done by the scripts and Codewarrior along with the toolkit libraries.  Just unpack the source bundles in your home directory or where ever you have specified in the appropriate properties.  The root folders for FLTK and BDB do not have to have any particular names.  If there is more than one version the scripts will grab whichever is last alphabetically (e.g. fltk-1.1.4r2 will get used instead of fltk-1.1.3).</p><p>The scripts normally halt on any Codewarrior compilation errors.  If you want them to continue and save errors, set the next script property, pSaveContinueOnErrors, to true.  Compilation errors for a project will be saved in a file in the same folder as the project being built, with a name in the following format: projectName-targetNumber.errs (e.g. xncbi-2.errs).</p><p>The Genome Workbench's configuration file(s) is stored in the users Library:Application Support:gbench folder.</p><p>CFM builds are not supported.  OS 8 or 9 are not supported.  We know 10.2 works.  We think 10.1 still works, and 10.3 might work.</p></div><div class="sec" id="release_notes_7_8_2004.platforms_7_8_200"><h3><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h3><p>This release was successfully tested on the following platforms -- but may also work on other platforms. Since the previous release, some platforms were dropped from this list, just because we do not use them here anymore, and some were added (these new platforms are highlighted using bold font). Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages -- in these cases, just skipping such projects (e.g. using flag "-k" for<span class="nctnt ncbi-app">make</span>on UNIX), can get you through.</p><div class="sec" id="release_notes_7_8_2004.platform_unix_7_8"><h4><span class="title">Unix</span></h4><div class="table" id="release_notes_7_8_2004.T2"><h4><span class="label">Table 11</span><span class="title">Unix OS's and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Platform</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.2.5)</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 2.95.3, 3.0.4,<b>3.4.0</b></span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.2.5)</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">ICC 7.1</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-8</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SPARC</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13  (64-bit, 32-bit)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-8</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SPARC</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-7</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-7</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">IRIX64-6.5</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SGI-Mips</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MIPSpro 7.3.1.2m (64-bit, 32-bit)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">FreeBSD-4.5</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">ALPHA</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.3.2</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">ALPHA</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">Compaq C++ V6.5-014</span>
								</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_7_8_2004.platform_windows_"><h4><span class="title">MS Windows</span></h4><div class="table" id="release_notes_7_8_2004.T3"><h4><span class="label">Table 12</span><span class="title">MS Windows and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MS Windows</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MSVC++ 6.0 Service Pack 5</span>.  To be DEPRECATED -- this is the last release where it is supported.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MS Windows</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MSVC++<b>7.1</b></span>. See documentation on<a href="ch_config.html#ch_config.MS_Windows">MS Visual C++.NET</a>.</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_7_8_2004.platform_mac_7_8_"><h4><span class="title">Mac OS X</span></h4><div class="table" id="release_notes_7_8_2004.T4"><h4><span class="label">Table 13</span><span class="title">Mac OS, and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS 10.2,<b>10.3.4</b></span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.3</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS<b>10.3.4</b></span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">CodeWarrior<b>9.2</b></span>
								</td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_7_8_2004.caveats_7_8_2004"><h3><span class="title">Caveats and Hints</span></h3><div class="sec" id="release_notes_7_8_2004.macos_10x_7_8_200"><h4><span class="title">MacOS 10.X / CodeWarrior 9.1</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Not all of the test or demo applications are built.</p></dd><dt>2</dt><dd><p class="no_top_margin">The source code for the latest release of FLTK (1.1.x), BerkeleyDB (4.x) and SQLite (2.x) should be present. See the  installation instructions for details.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.macos_10_X_7_8_20"><h4><span class="title">MacOS 10.2/GCC 3.3</span></h4><p>At least the GCC 3.3 update for Dec. 2002 Developers Tools required from Apple.</p></div><div class="sec" id="release_notes_7_8_2004.gcc_2_95_3_7_8_20"><h4><span class="title">GCC 2.95</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Poor MT-safety record.</p></dd><dt>2</dt><dd><p class="no_top_margin">Relatively incomplete/incorrect (comparing to modern compilers) STL implementation.</p></dd><dt>3</dt><dd><p class="no_top_margin">It is going to be deprecated in NCBI rather soon -- as soon as we have any significant trouble with its maintenance.</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.gcc_3_0_4_7_8_200"><h4><span class="title">GCC 3.0.4</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body (fixed in GCC 3.3).</p></dd><dt>2</dt><dd><p class="no_top_margin">STL stream uses locale in thread unsafe way which may result to segmentation fault when run in multithread mode (fixed in GCC 3.3).</p></dd><dt>3</dt><dd><p class="no_top_margin">Long-file support for C++ streams is disabled/broken (first broken in 3.0, fixed in 3.4).</p></dd></dl></div><div class="sec" id="release_notes_7_8_2004.gcc_3_3_7_8_2004"><h4><span class="title">GCC 3.3</span></h4><p>Other than the feature described below, GCC 3.3.2 has been very good for us; it has a lot of very ugly bugs finally fixed.</p><p>Painfully slow linking in debug mode on Linux with GCC-3.3 compiler:<dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.</p><p>GCC starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking.</p><p>GCC 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4.</p><p>The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug.</p><p>We had to patch GCC 3.3 in-house with the fix described at<a href="http://lists.boost.org/MailArchives/boost/msg53004.php">http://lists.boost.org/MailArchives/boost/msg53004.php</a>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Long-file support still broken (first broken in 3.1).</p></dd></dl>
				</p></div><div class="sec" id="release_notes_7_8_2004.gcc_3_4_7_8_2004"><h4><span class="title">GCC 3.4</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">The "Painfully slow linking..." (see GCC3.3 [1] above) is still an issue. -- Again, had to patch it in-house to speed it up, a la GCC 3.3.</p></dd><dt>2</dt><dd><p class="no_top_margin">At least on Linux, ifstream::readsome() does not always work for large files, as it calls an ioctl  that doesn't work properly for large files.</p></dd></dl></div></div><div class="sec" id="release_notes_7_8_2004.last_update_7_8_2"><h3><span class="title">Last Updated</span></h3><p>This section last updated on August 25, 2004</p></div></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (July 8, 2004)
                    . .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_7_8_2004</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (July 8, 2004)
                    . .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_7_8_2004</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
