<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (April, 2005) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_05_05_2005" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_10_03_2005.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_03_09_2005.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_05_05_2005.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (April, 2005)</span></h1><p class="small">Created: April 1, 2005; Last Update: April 27, 2005.</p></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.1">Download Location</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.2">Source Archive Contents</a></p><ul><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.2.1">Source Code Archives</a></p></li></ul></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3">New Development</a></p><ul><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.1">CORELIB -- Portability and Application Framework</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.2">CONNECT -- Data streaming, Networking, and Dispatching</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.3">UTIL -- Miscellaneous Low-Level APIs</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.4">SERIAL -- Data Serialization (ASN.1, XML)</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.5">DATATOOL -- Code Generator and Data Converter Tool</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.6">CGI -- CGI and Fast-CGI Application Framework</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.7">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.8">DBAPI -- Generic SQL Database Connectivity</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.9">ALGO/ALIGN -- Generic Alignment Algorithms</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.10">BLAST</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.11">ALNMGR -- Bio-sequence Alignment Manager</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.12">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.14">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.15">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.16">OM++ TEST programs</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.17">OM++ DEMO program (objmgr_demo)</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.18">ID2</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.19">BUILD FRAMEWORK (MSVC++.NET)</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.20">PTB -- Project Tree Builder for MSVC++ .NET</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.21">3RD-PARTY PACKAGES</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.3.22">PYTHON -- Scripting Language Support</a></p></li></ul></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.4">FRAMEWORKS</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.5">Documentation</a></p><ul><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.5.1">Document Location</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.5.2">Document Content</a></p></li></ul></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.6">Building on the MacOS</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.7">Platforms (OS's, compilers used inside NCBI)</a></p><ul><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.7.1">Unix</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.7.2">MS Windows</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.7.3">Mac OS X</a></p></li></ul></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8">Caveats and Hints</a></p><ul><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8.1">MacOS 10.X / CodeWarrior 9.2</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8.2">MacOS 10.2/GCC 3.3</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8.3">GCC 2.95</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8.4">GCC 3.0.4</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8.5">GCC 3.3</a></p></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.8.6">GCC 3.4</a></p></li></ul></li><li><p><a href="release_notes_05_05_2005.html#release_notes_05_05_2005.1.9">Last Updated</a></p></li></ul><div class="sec" id="release_notes_05_05_2005.1.1"><h2><span class="title">Download Location</span></h2><p>
      <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2005/Apr_22_2005/">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2005/Apr_22_2005/</a>
    </p></div><div class="sec" id="release_notes_05_05_2005.1.2"><h2><span class="title">Source Archive Contents</span></h2><div class="sec" id="release_notes_05_05_2005.1.2.1"><h3><span class="title">Source Code Archives</span></h3><ul><li><p><span class="nctnt ncbi-path">ncbi_cxx_unix--Apr_22_2005.tar.gz</span> -- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_unix-- Apr_22_2005.gtar.gz</span> -- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_win-- Apr_22_2005.exe</span> -- for MS-Windows / MSVC++ 7.1 (self-extracting)</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_win-- Apr_22_2005.zip</span> -- for MS-Windows / MSVC++ 7.1</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_mac_cw-- Apr_22_2005.tgz</span> -- for MacOS 10.3.4 / CodeWarrior DevStudio for MacOS 9.2</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_mac_xcode-- Apr_22_2005.tgz</span> -- for MacOS 10.3.4 / xCode 1.[1-5]</p></li></ul><p>
        <span class="nctnt highlight">The sources correspond to the NCBI </span>
        production tree
        <span class="nctnt highlight"> sources from patch "RELEASE_APR_2005", which in turn roughly corresponds to the </span>
        development tree
        <span class="nctnt highlight"> sources from the very end of March, 2005.</span>
      </p></div></div><div class="sec" id="release_notes_05_05_2005.1.3"><h2><span class="title">New Development</span></h2><p>
      <span class="nctnt highlight">Some of the new development that may introduce potentially backward-incompatible changes are marked with an asterisk(*).</span>
    </p><div class="sec" id="release_notes_05_05_2005.1.3.1"><h3><span class="title">CORELIB -- Portability and Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Modified <span class="nctnt ncbi-class">CNcbiApplication</span> class to add yet another command line flag - to distinguish between "short" and "detailed" help printout. The detailed help is printed only upon request.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CDll</span> - added 'auto_unload' parameter to constructors to be able to unload DLL in a destructor.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CPluginManager_DllResolver</span> - added 'unload_dll' parameter to a constructor to be able to unload driver DLLs in the destructor.</p></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CPluginManager</span> - CPluginManager won't unload its drivers by default. All drivers that used to be unloaded previously kept old behavior.</p></dd><dt>5</dt><dd><p class="no_top_margin">File API (<span class="nctnt ncbi-class">CDirEntry, CFile, CDir</span>) -- dropped native (non-<span class="nctnt ncbi-monospace">UNIX</span>)Â  <span class="nctnt ncbi-monospace">MacOS 9</span> support.</p></dd><dt>6</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CDirEntry</span> added methods:</p><ul><li><p><span class="nctnt ncbi-func">CreateObject()</span> -- construct <span class="nctnt ncbi-class">CDirEntry</span> based object of specified type (file, dir, symlink).</p></li><li><p><span class="nctnt ncbi-func">operator=</span> -- assignment operator.</p></li><li><p><span class="nctnt ncbi-func">Copy()</span> -- copy one dir entry to another (should have equal type).</p></li><li><p><span class="nctnt ncbi-func">CopyToDir()</span> -- copy dir entry to to some directory.</p></li><li><p><span class="nctnt ncbi-func">SetBackupSuffix()</span> -- set suffix used to create backup copies of dir entries.</p></li><li><p><span class="nctnt ncbi-func">GetBackupSuffix()</span> -- get current backup suffix.</p></li><li><p><span class="nctnt ncbi-func">Backup()</span> -- create backup copy for dir entry.</p></li><li><p><span class="nctnt ncbi-func">IsLink()</span> -- check if directory entry a symbolic link.</p></li><li><p><span class="nctnt ncbi-func">LookupLink()</span> -- get entry name that link is pointed to.</p></li><li><p><span class="nctnt ncbi-func">DereferenceLink()</span> -- dereference symbolic link, replace current object entry name with entry name that link is pointed to</p></li><li><p><span class="nctnt ncbi-func">IsNewer()</span> -- check if current entry is newer than some other.</p></li><li><p><span class="nctnt ncbi-func">GetOwner()</span> -- get owner of dir entry.</p></li><li><p><span class="nctnt ncbi-func">SetOwner()</span> -- set new owner for dir entry.</p></li><li><p><span class="nctnt ncbi-func">Get/SetTimeT()</span> -- get/set dir entry time using <span class="nctnt ncbi-class">time_t</span>.</p></li></ul></dd><dt>7</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CDirEntry</span> -- added <span class="nctnt ncbi-class">CMask</span> versions of <span class="nctnt ncbi-func">MatchesMask(), GetEntries(), FindFiles()</span> and <span class="nctnt ncbi-func">FindFilesInDir()</span> methods.</p></dd><dt>8</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CDirEntry::Rename()</span> -- added flags parameter.</p></dd><dt>9</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CFile</span> -- added new methods:</p><ul><li><p><span class="nctnt ncbi-func">Copy()</span> -- copy one file to another.</p></li><li><p><span class="nctnt ncbi-func">Compare()</span> -- compare file's content.</p></li></ul></dd><dt>10</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CDir::Copy()</span> -- new method to copy directories.</p></dd><dt>11</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CDir::GetEntries(), CDirEntry::MatchesMask()</span> - added parameter for case sensitive/insensitive matching.</p></dd><dt>12</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSymLink</span> -- new class to work with symbolic links (UNIX specific).</p></dd><dt>13</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">SuppressSystemMessageBox()</span> -- new function to suppress popup messages on execution errors in both runtime and in debug libraries, as well as all <span class="nctnt ncbi-var">General Protection Fault</span> messages. (<span class="nctnt ncbi-monospace">Windows</span> specific).</p></dd><dt>14</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CDllResolver::AddExtraDllPath()</span> -- added support of multiply paths in the NCBI runpath.</p></dd><dt>15</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">NStr::MatchesMask()</span> -- added parameter for case sensitive/insensitive matching.</p></dd><dt>16</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CTime::IsValid()</span> -- added leapsecond support.</p></dd><dt>17</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CTimeSpan::IsEmpty()</span> -- new method to check if time span is empty.</p></dd><dt>18</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CFastLocalTime</span> -- new class for support quick-and-dirty getter of local time.</p></dd><dt>19</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CNcbiRegistry</span> -- support taking settings from the environment.</p></dd><dt>20</dt><dd><p class="no_top_margin">Added support for safe boolean operators to prevent their accidental use in non-boolean context.</p></dd><dt>21</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CPluginManager</span> now keeps track of registered entry points.</p></dd><dt>22</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CPluginManager</span>: add lib prefix to dll name mask.</p></dd><dt>23</dt><dd><p class="no_top_margin">Configuration files: added possibility to include sections. Merge subnodes and values having the same id.</p></dd><dt>24</dt><dd><p class="no_top_margin">New <span class="nctnt ncbi-func">CProcess::GetParentPid()</span> -- to retrieve parent's PID on <span class="nctnt ncbi-monospace">UNIX/Windows</span>.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.2"><h3><span class="title">CONNECT -- Data streaming, Networking, and Dispatching</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">New <span class="nctnt ncbi-var">BASE64_{En|De}code</span> API to perform <span class="nctnt ncbi-var">BASE64 (RFC1521)</span> encoding and decoding.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">ConnNetInfo_Create()</span> to always set <span class="nctnt ncbi-var">SConnNetInfo::user_header</span> from environment/registry key <span class="nctnt ncbi-var">CONN_HTTP_USER_HEADER</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">New method <span class="nctnt ncbi-func">CConn_IOStream::Close()</span> for forced stream closure.</p></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">HTTP_CreateConnector()</span> is now overriding additional user header instead of formerly setting it (thus the value is able to get preserved if it is not conflicting with the overriding one).</p></dd><dt>5</dt><dd><p class="no_top_margin">Major bug fix in "\r\n" parsing in both header and body.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.3"><h3><span class="title">UTIL -- Miscellaneous Low-Level APIs</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CRequestRateControl</span> -- new class to cap the rate of requests to a resource.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CTar</span> -- new class for basic <span class="nctnt ncbi-var">TAR</span> archive support.</p></dd><dt>3</dt><dd><p class="no_top_margin">Implemented <span class="nctnt ncbi-func">CIStreamBuffer::HasMore()</span>.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.4"><h3><span class="title">SERIAL -- Data Serialization (ASN.1, XML)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Implemented serialization of XML elements with mixed content.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CClassPrePostReadWrite</span> -- implements <span class="nctnt ncbi-var">PreRead</span> and <span class="nctnt ncbi-var">PostWrite</span> serialization hooks.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.5"><h3><span class="title">DATATOOL -- Code Generator and Data Converter Tool</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Implemented XML elements with mixed content in DTD parsing and writing.</p></dd><dt>2</dt><dd><p class="no_top_margin">Modified DATATOOL to produce correct output when converting DTD to DTD.</p></dd><dt>3</dt><dd><p class="no_top_margin">Corrected generation of XML schema:  fixed syntax errors, corrected definition of default values, fixed "real" data type (changed "decimal" to "double").</p></dd><dt>4</dt><dd><p class="no_top_margin">Corrected C++ code generation to correctly handle custom data types for ASN data of <span class="nctnt ncbi-var">ENUMERATED</span> type.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.6"><h3><span class="title">CGI -- CGI and Fast-CGI Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Major fixes in the use of CArgs argument description mechanism  for CGI applications.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.7"><h3><span class="title">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added cursor reopen operation.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">PrintStat</span> method to output <span class="nctnt ncbi-var">BTREE</span> statistics.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.8"><h3><span class="title">DBAPI -- Generic SQL Database Connectivity</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Database driver manager revamped to use "core" <span class="nctnt ncbi-class">CPluginManager</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added a new method <span class="nctnt ncbi-func">I_DriverMgr::AddDllSearchPath</span> to be able to get <span class="nctnt ncbi-class">I_DriverContext</span> from a <span class="nctnt ncbi-class">TPluginManagerParamTree</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added a new method <span class="nctnt ncbi-func">I_DriverMgr::GetDriverContextFromTree</span> to allow adding a driver search path for the driver manager.</p></dd><dt>4</dt><dd><p class="no_top_margin">Developed a unit-test for DBAPI.</p></dd><dt>5</dt><dd><p class="no_top_margin">An initial version of a <span class="nctnt ncbi-var">Python</span> DBAPI module was released.</p></dd><dt>6</dt><dd><p class="no_top_margin">Renamed <span class="nctnt ncbi-lib">bdb_cache</span> and <span class="nctnt ncbi-lib">dbapi_cache</span> libs to <span class="nctnt ncbi-lib">ncbi_xcache_*</span>.</p></dd><dt>7</dt><dd><p class="no_top_margin">CDB_Connection::Abort() -- new method to severe a connection to database on socket level (so there are no post-mortem communications with the database server and related risks of network hanging).</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.9"><h3><span class="title">ALGO/ALIGN -- Generic Alignment Algorithms</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSeq_id</span> -- support several new prefixes in <span class="nctnt ncbi-func">IdentifyAccession()</span> method.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-var">FASTA</span> reader -- fix handling of non-<span class="nctnt ncbi-var">IUPAC</span> residues, runs of hyphens, and semicolon-started comments.</p></dd><dt>3</dt><dd><p class="no_top_margin">The formatting classes are now using <span class="nctnt ncbi-class">CSeq_id</span> objects instead of plain strings to represent sequences IDs.</p></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CNWAligner::SetTranscript()</span> -- new method to simulate the alignment.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.10"><h3><span class="title">BLAST</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Introduction of structures to represent the filtering string in <span class="nctnt ncbi-var">CORE BLAST</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">BlastSeqSrc</span> implementations clean up (<span class="nctnt ncbi-var">API BLAST</span>).</p></dd><dt>3</dt><dd><p class="no_top_margin">Addition of composition-based statistics for <span class="nctnt ncbi-var">PSI-BLAST</span> (code to perform <span class="nctnt ncbi-var">IMPALA</span>-style scaling of <span class="nctnt ncbi-var">PSSM</span>s in <span class="nctnt ncbi-path">blast_posit.[ch]</span>).</p></dd><dt>4</dt><dd><p class="no_top_margin">Introduction of interval trees, a binary tree data structure that greatly speeds up the search for alignments that envelop/contain other alignments. Used in the preliminary gapped alignment phase of all <span class="nctnt ncbi-var">BLAST</span> programs and in the traceback phase for all programs except <span class="nctnt ncbi-var">MEGABLAST</span>.</p></dd><dt>5</dt><dd><p class="no_top_margin">Improvements in speed and sensitivity to <span class="nctnt ncbi-var">DUST</span> (<span class="nctnt ncbi-path">blast_dust.c</span>).</p></dd><dt>6</dt><dd><p class="no_top_margin">Changes to how score matrices are stored in <span class="nctnt ncbi-var">CORE BLAST</span> (<span class="nctnt ncbi-class">SBlastScoreMatrix</span>).</p></dd><dt>7</dt><dd><p class="no_top_margin">Major refactoring of greedy alignment routines, minor refactoring of in-frame and out-of-frame gapped alignments routines (in-frame and out-of-frame gapped alignment outines now allocate memory for auxiliary structures dynamically, instead of allocating the worst case amount of memory when initializing).</p></dd><dt>8</dt><dd><p class="no_top_margin">Change in tie-breakers for score comparisons.</p></dd><dt>9</dt><dd><p class="no_top_margin">Rewrite of uneven gap <span class="nctnt ncbi-var">HSP</span> linking.</p></dd><dt>10</dt><dd><p class="no_top_margin">Unification of traceback generation functions (memory usage has been greatly reduced).</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.11"><h3><span class="title">ALNMGR -- Bio-sequence Alignment Manager</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMap</span> -- added a print class <span class="nctnt ncbi-class">CAlnMapPrinter</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnVec</span> -- added a print class <span class="nctnt ncbi-class">CAlnVecPrinter</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMix</span> -- rearranged into a set of classes with a preserved CAlnMix front-end interface.  The new classes are: </p><dl class="labeled-list"><dt>a</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMixMatches</span> -- container for <span class="nctnt ncbi-class">CAlnMixMatch</span> plus methods,</p></dd><dt>b</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMixSegments</span> -- container for <span class="nctnt ncbi-class">CAlnMixSegment</span> plus methods,</p></dd><dt>c</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMixSequences</span> -- container for <span class="nctnt ncbi-class">CAlnMixSeq</span> plus methods,</p></dd><dt>d</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMixMerger</span> -- the merging algorithm.</p></dd></dl></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMix</span> -- abstracted the <span class="nctnt ncbi-func">CalculateScore</span> method so that it could be delegated to the caller. Added handling for best reciprocal hits.</p></dd><dt>5</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CAlnMrg</span> --  added a <span class="nctnt ncbi-var">sortseqbyscore</span> flag, extended to reading multiple dense-segs, and added handling for best reciprocal hits.</p></dd><dt>6</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CAlnVwr::GetAlnPosFromSeqPosDemo()</span> -- new method.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.12"><h3><span class="title">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSeq_id</span> -- support several new prefixes in <span class="nctnt ncbi-func">IdentifyAccession()</span> method.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-var">FASTA</span> reader -- fix handling of non-<span class="nctnt ncbi-var">IUPAC</span> residues, runs of hyphens, and semicolon-started comments.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added possibility to use <span class="nctnt ncbi-func">GetSeq_idByType()</span> with containers of <span class="nctnt ncbi-class">CSeq_id_Handle</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">Preserve fuzz when merging/adding <span class="nctnt ncbi-var">seq-locs</span> in <span class="nctnt ncbi-var">seq-loc</span> operations.</p></dd><dt>5</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">Sequence::GetId()</span> now returns <span class="nctnt ncbi-class">CSeq_id_Handle</span>.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.13"><h3><span class="title">ALGO/PHY_TREE -- Phylogenetic and bio tree algorithms</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Minor bug fixes and improvements.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.14"><h3><span class="title">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</span></h3><dl class="labeled-list"><dt>2</dt><dd><p class="no_top_margin">Implemented setters for '<span class="nctnt ncbi-var">descr</span>' field of <span class="nctnt ncbi-class">CBioseq_Handle</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Implemented <span class="nctnt ncbi-func">CBioseq_EditHandle AddId()</span> &amp; <span class="nctnt ncbi-func">RemoveId()</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added methods to get feature type and subtype.</p></dd><dt>5</dt><dd><p class="no_top_margin">Added getter for feature type and subtype to <span class="nctnt ncbi-class">CMappedFeat</span>.</p></dd><dt>6</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">SSNP_Info</span> structure is defined in separate header to reduce dependencies.</p></dd><dt>7</dt><dd><p class="no_top_margin">Added "<span class="nctnt ncbi-var">SNP</span>" symbols to names of <span class="nctnt ncbi-class">CSeq_feat_Handle</span> methods used to access SNP table.</p></dd><dt>8</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">CScope::RemoveFromHistory()</span>, <span class="nctnt ncbi-func">CScope::RemoveTopLevelSeqEntry()</span> and <span class="nctnt ncbi-func">CScope::RemoveDataLoader()</span>.</p></dd><dt>9</dt><dd><p class="no_top_margin">Added state flags to <span class="nctnt ncbi-class">CBioseq_Handle</span>. Report conflicts using state flag, do not throw exception.</p></dd><dt>10</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">CBioseq_Handle::ContainsSegment</span> with optional depth and limit object.</p></dd><dt>11</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CBioseq_Handle::RemoveDesc()</span> -- now returns <span class="nctnt ncbi-class">CRef&lt;CSeqdesc&gt;</span> (was <span class="nctnt ncbi-var">bool</span>).</p></dd><dt>12</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">SAnnotSelector(TFeatSubtype)</span>. Added flag to <span class="nctnt ncbi-class">SAnnotSelector</span> for skipping multiple SNPs from the same <span class="nctnt ncbi-var">seq-annot</span>.</p></dd><dt>13</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSeq_loc_Mapper</span> -- added direction flag for mapping between top level sequence and segments.</p></dd><dt>14</dt><dd><p class="no_top_margin">Iterators: ignore <span class="nctnt ncbi-var">e_Locs</span> annotations with unknown format (used to fail before).</p></dd><dt>15</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CBioseq_CI</span> iterator with <span class="nctnt ncbi-var">CSeq_inst::eMol_na</span> now includes both dna and rna bioseqs. Added constructor accepting <span class="nctnt ncbi-class">CBioseq_set_Handle</span>.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.15"><h3><span class="title">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added auto cleaning of indexes with id resolution information.</p></dd><dt>2</dt><dd><p class="no_top_margin">Commented out obsolete classes <span class="nctnt ncbi-class">CRefresher, CMutexPool</span>, and <span class="nctnt ncbi-class">CGBLGuard</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-var">GenBank</span> data loader now has new schema of readers/writers.</p></dd><dt>4</dt><dd><p class="no_top_margin">It's now possible to use <span class="nctnt ncbi-var">GenBank</span> loader cache by specifying it in registry, e.g. "<span class="nctnt ncbi-var">[GENBANK] ReaderName = cache;id1</span>", or via environment "<span class="nctnt ncbi-var">GENBANK_LOADER_METHOD=cache;id1</span>".</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.16"><h3><span class="title">OM++ TEST programs</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added few tests for loading split data from <span class="nctnt ncbi-var">ID2</span> server.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.17"><h3><span class="title">OM++ DEMO program (objmgr_demo)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added flag -print_mapper to check <span class="nctnt ncbi-class">CSeq_loc_Mapper</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added scan of all features in TSE when -whole_tse is specified.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added option to allow loading <span class="nctnt ncbi-var">Seq-annot</span> from file.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.18"><h3><span class="title">ID2</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added split-version to <span class="nctnt ncbi-class">ID2S-Request-Get-Chunks</span> structure of ID2 protocol.</p></dd><dt>2</dt><dd><p class="no_top_margin">Allow storing skeleton <span class="nctnt ncbi-class">Seq-entry</span> together with split info in ID2.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.19"><h3><span class="title">BUILD FRAMEWORK (MSVC++.NET)</span></h3></div><div class="sec" id="release_notes_05_05_2005.1.3.20"><h3><span class="title">PTB -- Project Tree Builder for MSVC++ .NET</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Now can keep track of subproject types (expendable, potential) and propagate it down the project tree, and to generate MSVC projects for all subprojects, not only for those listed in <span class="nctnt ncbi-path">Makefile.in</span>.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.21"><h3><span class="title">3RD-PARTY PACKAGES</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">CPPUNIT</span> was added to 3rd-party packages. To use <span class="nctnt ncbi-lib">CPPUNIT</span> just add <span class="nctnt ncbi-var">REQUIRES = CPPUNIT</span> to your <span class="nctnt ncbi-path">Makefile.in</span>, <span class="nctnt ncbi-var">$(CPPUNIT_INCLUDE)</span> to <span class="nctnt ncbi-var">CPPFLAGS</span> and <span class="nctnt ncbi-var">$(CPPUNIT_LIBS)</span> to <span class="nctnt ncbi-var">LIBS</span> respectively in your app file.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.3.22"><h3><span class="title">PYTHON -- Scripting Language Support</span></h3><p>1. An initial version of a Python DBAPI module was released.</p></div></div><div class="sec" id="release_notes_05_05_2005.1.4"><h2><span class="title">FRAMEWORKS</span></h2><p>A beta version of grid computing components, including:</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">NetCache</span>  -- daemon for network-distributed data cache,</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">NetSchedule</span>  -- daemon to support network-distributed request queues,</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">GridWorker</span> -- API to create grid computing nodes,</p></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">Cgi_Tunnel2Grid</span> -- CGI to tunnel request to <span class="nctnt ncbi-lib">GridWorkers</span>.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.5"><h2><span class="title">Documentation</span></h2><div class="sec" id="release_notes_05_05_2005.1.5.1"><h3><span class="title">Document Location</span></h3><p>The documentation is available online at <!--ro--><a href="toc.html">http://www.ncbi.nlm.nih.gov/books/bv.fcgi?rid=toolkit.TOC&amp;depth=2</a> as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release. The PDF version can be accessed by a link that appears on each page.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_05_05_2005.1.5.2"><h3><span class="title">Document Content</span></h3><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit" gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
        <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/">http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
      </p><p>The above link is also available under the "Browsers" that appears on each page.</p><p>You can also access the CVS code repository via a Web interface. These links also appear in the sidebar box on each page.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the public or in-house versions of LXR, Doxygen and Library. The Library search runs a CGI script that lists the library name where the symbol (such as function) is defined in.</p><p>The current release notes as well as past release notes are now in an appendix in the C++ Toolkit Book and a link to the current release notes appears on each page of the online Toolkit document.</p></div></div><div class="sec" id="release_notes_05_05_2005.1.6"><h2><span class="title">Building on the MacOS</span></h2><p>We now build all the libraries and most of the applications including the Genome Workbench (<span class="nctnt ncbi-app">gbench</span>), some test and a few demo applications. We also build the FLTK library's GUI editor, <span class="nctnt ncbi-app">fluid</span>.</p><p>All apps are built as application bundles except <span class="nctnt ncbi-app">gbench_plugin_scan</span> and <span class="nctnt ncbi-app">datatool</span> which are built as command line apps. Any of the applications can be built as command line apps by tweaking the build scripts or the CodeWarrior projects.</p><p>When building the toolkit with xCode, the latest version of xCode (at least 1.5) is required for the trouble-free build. Build procedure is as follows: open, build and run a project file in compilers/xCode.  This is a GUI tool to generate a new NCBI C++ Toolkit xCode project.  You'll have an option to specify third-party installation directories and choose which packages (libs, applications and tests) to include into the final project.</p><p>xCode build fully support all the latest Apple innovations: distributed  builds, optional CPU specific optimization , pre-compiled headers, fix  &amp; continue, code cleanup and Zero Link (with few exceptions).</p><p>xCode builds all libraries as a Mach-O dynamically linked shared ones (.dylib) and all Genome Workbench plugins as Mach-O bundles (also .dylib extension). Note, that xCode will place Genome Workbench plugins inside Genome Workbench application bundle (Genome Workbench.app/Contents/MacOS/plugins).</p><p>To build the toolkit with CodeWarrior use an AppleScript editor to open and run the script files makeLibs.met and makeApps.met. You must use a script editor capable of opening a script file larger than 32K, such as Apple's Script Editor v2.0, or Smile. Script Editor v1.9 will not work since makeLibs.met just got too big. The command line tool osascript also works.</p><p>On running the scripts you will be prompted as to which targets you want to build. Or if you always build the same targets they can be specified by including an empty file or folder in the compilers:mac_prj  folder with the name 'Build' followed by the keywords of the targets  you want built. The keywords are: Debug and Final. For example, to  build only the Debug targets use: "Build Debug", to build both debug  and release (final) versions: "Build".</p><p>If you install the C++ toolkit under a different name than "ncbi_cxx" or in a different location than your home directory, you can edit the script's properties, pRootFolderName and pRootFolderPath, to override these defaults. Note: these paths, and those mentioned below, must be entered in mac format (e.g. disk:Users:username:) not Unix format (e.g. <span class="nctnt ncbi-path">/Users/username/</span>). The disk name (and its following colon) may be omitted.</p><p>Certain third party libraries (see <a href="release_notes_05_05_2005.html#release_notes_05_05_2005.T1">Table 1</a>) are required to build some parts of the C++ toolkit. The scripts will try and find them if they are in your home directory, or you can specify where they were installed using properties at the beginning of the script.</p><div class="table" id="release_notes_05_05_2005.T1"><h4><span class="title">Table 1. Third Party Libraries</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Library </th><th align="left" valign="top" rowspan="1" colspan="1">Property </th><th align="left" valign="top" rowspan="1" colspan="1">Example</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-lib">FLTK 1.1.6 </span>(w/ NCBI patches) </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">pFLTKRootFolder </span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-monospace">"</span>
              <span class="nctnt ncbi-monospace">home:mhome:fltk-1.1.6-ncbi3"</span>
            </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-lib">BerkeleyDB 4.2 </span>
              <span class="nctnt ncbi-lib">(or 4.3)</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">pBdbRootFolder </span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-monospace">"Users:myhome:mylibs:db-4.2.52"</span>
              <span class="nctnt ncbi-monospace">(or "... db-4.3.21")</span>
            </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-lib">SQLite 2.8.13</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">gSqliteFolder</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr></tbody></table></div></div><p>You do not have to build the FLTK or BDB libraries separately. This is done by the scripts and CodeWarrior along with the toolkit libraries. Just unpack the source bundles in your home directory or where ever you have specified in the appropriate properties. The root folders for FLTK and BDB do not have to have any particular names. If there is more than one version the scripts will grab whichever is last alphabetically (e.g. fltk-1.1.4r2 will get used instead of fltk-1.1.3).</p><p>The scripts normally halt on any CodeWarrior compilation errors. If you want them to continue and save errors, set the next script property, pSaveContinueOnErrors, to true. Compilation errors for a project will be saved in a file in the same folder as the project being built, with a name in the following format: projectName-targetNumber.errs (e.g. xncbi-2.errs).</p><p>The Genome Workbench's configuration file(s) is stored in the user's <span class="nctnt ncbi-path">Library:Application Support:gbench</span> folder.</p><p>CFM builds are not supported. OS 8 or 9 are not supported. We know 10.3 works. We think 10.1 still works, and 10.2 might work.</p></div><div class="sec" id="release_notes_05_05_2005.1.7"><h2><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h2><p>This release was successfully tested on at least the following platforms -- but may also work on other platforms. Since the previous release, some platforms were dropped from this list, just because we do not use them here anymore, and some were added (these new platforms are highlighted using bold font). Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages -- in these cases, just skipping such projects (e.g. using flag "-k" for <span class="nctnt ncbi-app">make</span> on UNIX), can get you through.</p><div class="sec" id="release_notes_05_05_2005.1.7.1"><h3><span class="title">Unix</span></h3><div class="table" id="release_notes_05_05_2005.T2"><h4><span class="title">Table 2. Unix OS's and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC </span>2.3.2<span class="nctnt ncbi-monospace">) </span></td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL </span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.0, 3.0.4, 2.95.3</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Linux-2.4.26 (w/ LIBC 2.3.2)</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.0</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Linux-2.6.7 (w/ LIBC 2.3.3)</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL/64</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.3</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.2.5)</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">ICC 8.0</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-8,9</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">SPARC</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-</span><span class="nctnt ncbi-app">21</span> (64-bit, 32-bit)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-8</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">SPARC</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.3</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-9</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-9</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.3</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">IRIX64-6.5</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">SGI-Mips</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">MIPSpro 7.3.1.3m (64-bit, 32-bit)</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">FreeBSD-4.10</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.2</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">ALPHA</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.3.2</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">ALPHA</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">Compaq C++ V6.5-014</span>
              </td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_05_05_2005.1.7.2"><h3><span class="title">MS Windows</span></h3><div class="table" id="release_notes_05_05_2005.T3"><h4><span class="title">Table 3. MS Windows and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MS Windows</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-app">MSVC++ 7.1</span>. See documentation on <a href="ch_config.html#ch_config.MS_Windows">MS Visual C++.NET</a>.</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_05_05_2005.1.7.3"><h3><span class="title">Mac OS X</span></h3><div class="table" id="release_notes_05_05_2005.T4"><h4><span class="title">Table 4. Mac OS, and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MacOS 10.3</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.3</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MacOS 10.3</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">CodeWarrior 9.2</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MacOS 10.3</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">xCode 1.5</span>
              </td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_05_05_2005.1.8"><h2><span class="title">Caveats and Hints</span></h2><div class="sec" id="release_notes_05_05_2005.1.8.1"><h3><span class="title">MacOS 10.X / CodeWarrior 9.2</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Not all of the test or demo applications are built.</p></dd><dt>2</dt><dd><p class="no_top_margin">The source code for the latest release of FLTK (1.1.x), BerkeleyDB (4.x) and SQLite (2.x) should be present. See the installation instructions for details.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.8.2"><h3><span class="title">MacOS 10.2/GCC 3.3</span></h3><p>At least the GCC 3.3 update for Dec. 2002 Developers Tools required from Apple.</p></div><div class="sec" id="release_notes_05_05_2005.1.8.3"><h3><span class="title">GCC 2.95</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Poor MT-safety record.</p></dd><dt>2</dt><dd><p class="no_top_margin">Relatively incomplete/incorrect (comparing to modern compilers) STL implementation.</p></dd><dt>3</dt><dd><p class="no_top_margin">It is going to be deprecated in NCBI rather soon -- as soon as we have any significant trouble with its maintenance.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.8.4"><h3><span class="title">GCC 3.0.4</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body (fixed in GCC 3.3).</p></dd><dt>2</dt><dd><p class="no_top_margin">STL stream uses locale in thread unsafe way which may result to segmentation fault when run in multithread mode (fixed in GCC 3.3).</p></dd><dt>3</dt><dd><p class="no_top_margin">Long-file support for C++ streams is disabled/broken (first broken in 3.0, fixed in 3.4).</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.8.5"><h3><span class="title">GCC 3.3</span></h3><p>Other than the feature described below, GCC 3.3.2 had been very good to us; it had a lot of very ugly bugs finally fixed.</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Painfully slow linking in debug mode on Linux with GCC-3.3 compiler. -- Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.  GCC starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking. GCC 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4. The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug (N^2). Binutils 2.15 fixes this. The link time still 2 times slower than without symbol merge, but the resultant executable is about two times smaller in size, and no compiler patching is necessary. We are still testing it in-house. We had to patch GCC 3.3 in-house with the fix described at http://lists.boost.org/MailArchives/boost/msg53004.php.</p></dd><dt>2</dt><dd><p class="no_top_margin">Long-file support still broken.</p></dd></dl></div><div class="sec" id="release_notes_05_05_2005.1.8.6"><h3><span class="title">GCC 3.4</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">The "Painfully slow linking..." (see GCC3.3,! [1] above) was an issue, and we had to patch it in-house to speed up, a la GCC 3.3 -- until we finally upgraded to binutils 2.15.</p></dd><dt>2</dt><dd><p class="no_top_margin">At least on Linux, ifstream::readsome() does not always work for large files, as it calls an ioctl that doesn't work properly for large files.</p></dd><dt>3</dt><dd><p class="no_top_margin">At least on Linux, GCC 3.4.[0,1] optimizer (very rarely) generates incorrect code when comparing enumerated values in else-ifs. (Fixed in 3.4.2)</p></dd></dl></div></div><div class="sec" id="release_notes_05_05_2005.1.9"><h2><span class="title">Last Updated</span></h2><p>This section last updated on April 27, 2005.</p></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (April, 2005)
                    .
        [Updated: April 27, 2005]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_05_05_2005</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_05_05_2005.1.1">Download Location</a></li><li><a href="#release_notes_05_05_2005.1.2">Source Archive Contents</a></li><li><a href="#release_notes_05_05_2005.1.3">New Development</a></li><li><a href="#release_notes_05_05_2005.1.4">FRAMEWORKS</a></li><li><a href="#release_notes_05_05_2005.1.5">Documentation</a></li><li><a href="#release_notes_05_05_2005.1.6">Building on the MacOS</a></li><li><a href="#release_notes_05_05_2005.1.7">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_05_05_2005.1.8">Caveats and Hints</a></li><li><a href="#release_notes_05_05_2005.1.9">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />Â Â 					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (April, 2005)
                    .
        [Updated: April 27, 2005]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_05_05_2005</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_05_05_2005.1.1">Download Location</a></li><li><a href="#release_notes_05_05_2005.1.2">Source Archive Contents</a></li><li><a href="#release_notes_05_05_2005.1.3">New Development</a></li><li><a href="#release_notes_05_05_2005.1.4">FRAMEWORKS</a></li><li><a href="#release_notes_05_05_2005.1.5">Documentation</a></li><li><a href="#release_notes_05_05_2005.1.6">Building on the MacOS</a></li><li><a href="#release_notes_05_05_2005.1.7">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_05_05_2005.1.8">Caveats and Hints</a></li><li><a href="#release_notes_05_05_2005.1.9">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />Â Â 					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
