<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (February, 2005) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_03_09_2005" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_05_05_2005.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_11_22_2004.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_03_09_2005.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (February, 2005)</span></h1><p class="small">Created: February 1, 2005; Last Update: February 1, 2005.</p></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.1">Download Location</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.2">Source Archive Contents</a></p><ul><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.2.1">Source Code Archives</a></p></li></ul></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3">New Development</a></p><ul><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.1">CORELIB -- Portability and Application Framework</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.2">CONNECT -- Data streaming, Networking, and Dispatching</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.3">UTIL -- Miscellaneous Low-Level APIs</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.4">SERIAL -- Data Serialization (ASN.1, XML)</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.5">DATATOOL -- Code Generator and Data Converter Tool</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.6">CGI -- CGI and Fast-CGI Application Framework</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.7">HTML -- HTML Generation Library</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.8">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.9">DBAPI -- Generic SQL Database Connectivity</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.10">ALGO/ALIGN -- Generic Alignment Algorithms</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.11">BLAST</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.12">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.13">ALGO/PHY_TREE -- Phylogenetic and bio tree algorithms</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.14">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.15">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.16">BUILD FRAMEWORK (UNIX)</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.17">BUILD FRAMEWORK (MSVC++.NET)</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.18">PTB -- Project Tree Builder for MSVC++ .NET</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.19">3RD-PARTY PACKAGES</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.3.20">PYTHON -- Scripting Language Support</a></p></li></ul></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.4">APPLICATIONS</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.5">Documentation</a></p><ul><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.5.1">Document Location</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.5.2">Document Content</a></p></li></ul></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.6">Building on the MacOS</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.7">Platforms (OS's, compilers used inside NCBI)</a></p><ul><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.7.1">Unix</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.7.2">MS Windows</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.7.3">Mac OS X</a></p></li></ul></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8">Caveats and Hints</a></p><ul><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8.1">MacOS 10.X / CodeWarrior 9.1</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8.2">MacOS 10.2/GCC 3.3</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8.3">GCC 2.95</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8.4">GCC 3.0.4</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8.5">GCC 3.3</a></p></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.8.6">GCC 3.4</a></p></li></ul></li><li><p><a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.1.9">Last Updated</a></p></li></ul><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.1"><h2><span class="title">Download Location</span></h2><p>
      <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2005/Feb_17_2005/">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2005/Feb_17_2005/</a>
    </p></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.2"><h2><span class="title">Source Archive Contents</span></h2><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.2.1"><h3><span class="title">Source Code Archives</span></h3><ul><li><p><span class="nctnt ncbi-path">ncbi_cxx_unix--Feb_17_2005.tar.gz</span> -- for UNIX'es (see the list of UNIX flavors below) and MacOSX 10.[2-3] /GCC 3.3</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_unix--Feb_17_2005.gtar.gz</span> -- for UNIX'es (see the list of UNIX flavors below) and MacOSX 10.[2-3] /GCC 3.3</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_win--Feb_17_2005.exe</span> -- for MS-Windows / MSVC++ 7.1 (self-extracting)</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_win--Feb_17_2005.zip</span> -- for MS-Windows / MSVC++ 7.1</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_mac_cw--Feb_17_2005.tgz</span> -- for MacOS 10.3.X / CodeWarrior 9.2</p></li><li><p><span class="nctnt ncbi-path">ncbi_cxx_mac_xcode--Feb_17_2005.tgz</span> -- for MacOS 10.3.X / xCode 1.5</p></li></ul><p>
        <span class="nctnt highlight">The sources correspond to the NCBI </span>
        production tree
        <span class="nctnt highlight"> sources from patch "RELEASE_FEB_2005", which in turn closely corresponds to the </span>
        development tree
        <span class="nctnt highlight"> sources from January 18-19, 2005.</span>
      </p></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3"><h2><span class="title">New Development</span></h2><p>
      <span class="nctnt highlight">Some of the new development that may introduce potentially backward-incompatible changes are marked with an asterisk(*).</span>
    </p><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.1"><h3><span class="title">CORELIB -- Portability and Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CNcbiRegistry</span> -- refactored, with base classes split out at various levels.  Along the way, it gained support for prioritized collections of sub-registries.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CMetaRegistry</span> -- the default search path has changed, and become easier to tune via environment settings.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CNcbiRegistry</span> -- added functions for easy retrieval of configuration parameters: <span class="nctnt ncbi-func">GetConfigString</span>, <span class="nctnt ncbi-func">GetConfigInt</span>, and <span class="nctnt ncbi-func">GetConfigFlag</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CArgDescriptions</span> -- new flag to allow multiple command-line arguments (<span class="nctnt ncbi-var">EFlags::fAllowMultiple</span>).</p></dd><dt>5</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">CDir::Create()</span> -- return <span class="nctnt ncbi-var">TRUE</span> if creating directory already exists.</p></dd><dt>6</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-func">NStr::MatchesMask()</span> -- added string version of this function.</p></dd><dt>7</dt><dd><p class="no_top_margin">Command line arguments processing, added constraint inversion (<span class="nctnt ncbi-var">NOT</span>).</p></dd><dt>8</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CVersionInfo</span> -- added support for version srtings (like "2.5.1")</p></dd><dt>9</dt><dd><p class="no_top_margin">Diagnostics -- now allows configuring message filtering by severity.</p></dd><dt>10</dt><dd><p class="no_top_margin">Removed <span class="nctnt ncbi-func">operator bool()</span> from <span class="nctnt ncbi-class">CRef&lt;&gt;</span> and <span class="nctnt ncbi-class">CConstRef&lt;&gt;</span> templates. Pointer conversion will be used instead. Added helper template and macro for easy implementation of boolean operator via pointer.</p></dd><dt>11</dt><dd><p class="no_top_margin">Moved <span class="nctnt ncbi-path">obj_store</span> and <span class="nctnt ncbi-path">plugin_manager_store</span> to <span class="nctnt ncbi-lib">CORELIB</span>.</p></dd><dt>12</dt><dd><p class="no_top_margin">Plugin Manager -- added <span class="nctnt ncbi-func">CDllResolver_Getter()</span> to get default DLL resolver.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.2"><h3><span class="title">CONNECT -- Data streaming, Networking, and Dispatching</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CNamedPipe</span> -- added automatic generation of OS-specific pipe names.</p></dd><dt>2</dt><dd><p class="no_top_margin">FTP client connector (<span class="nctnt ncbi-class">FTP_CreateDownloadConnector</span>) and stream (<span class="nctnt ncbi-class">CConn_FTPDownloadStream</span>) have been implemented (only for download).</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">NetCache</span> client -- implemented BLOB update, allow to engage NCBI load balancer.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.3"><h3><span class="title">UTIL -- Miscellaneous Low-Level APIs</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CZipDecompressor</span> -- fixed bug in <span class="nctnt ncbi-func">Process()</span> when <span class="nctnt ncbi-var">fCheckFileHeader</span> flag is set.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-path">util/bitset</span> -- integrated new version of BM library into NCBI toolkit.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.4"><h3><span class="title">SERIAL -- Data Serialization (ASN.1, XML)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">XML output stream -- corrected writing namespace of <span class="nctnt ncbi-var">NamedType</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">XML input stream -- fixed the reading of empty sequence with all optional members represented by self-closed tag.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CTreeIterator</span> -- implemented context-dependent filtering.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.5"><h3><span class="title">DATATOOL -- Code Generator and Data Converter Tool</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Made it possible to process multiple objects in input stream when converting data from one format to another.</p></dd><dt>2</dt><dd><p class="no_top_margin">Implemented parsing of conditional sections in DTD, processing of entities in attribute definition, and processing of compound identifier names (those made of several entities).</p></dd><dt>3</dt><dd><p class="no_top_margin">Corrected generation of XML schema for "sequence of choice" types with attributes, and simple type with a default.</p></dd><dt>4</dt><dd><p class="no_top_margin">Corrected generation of links in DOXYGEN comments.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.6"><h3><span class="title">CGI -- CGI and Fast-CGI Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CCgiApplication</span> -- new callback <span class="nctnt ncbi-func">OnEvent()</span> to allow one catch and handle a variety of states and events happening in the CGI and Fast-CGI applications.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CCgiEntry</span> -- keep track of Content-Type headers from POST submissions. Added missing <span class="nctnt ncbi-func">operators !=</span>. Moved <span class="nctnt ncbi-func">operators ==</span> and <span class="nctnt ncbi-func">!=</span> into class.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-class">CArgDescriptions</span> class support for CGI. Now it is possible to describe CGI arguments as if they were command-line key arguments.</p></dd><dt>4</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CCgiRequest</span> -- new flag <span class="nctnt ncbi-var">fCaseInsensitiveArgs</span> to allow case-insensitive arguments in query string.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.7"><h3><span class="title">HTML -- HTML Generation Library</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CHTML_map</span> and <span class="nctnt ncbi-class">CHTML_area</span> -- new classes to support HTML <span class="nctnt ncbi-var">MAP</span> and <span class="nctnt ncbi-var">AREA</span> tags.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.8"><h3><span class="title">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CBDB_BlobFile</span> -- added <span class="nctnt ncbi-func">UpdateInsert().</span></p></dd><dt>2</dt><dd><p class="no_top_margin">Cursors -- implemented BLOB update, added support for Read-Modify-Write (RMW) locks.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CBDB_Cache</span> -- implemented adaptive checkpoints and individual BLOB timeouts, improved performance. Code moved into a separate library, which is set up as a "standard NCBI plug-in", with registration entry point, fixed interface, class factory, driver name, versioning, etc.</p></dd><dt>4</dt><dd><p class="no_top_margin">Ported to work with <span class="nctnt ncbi-lib">Berkeley DB 4.3</span>.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.9"><h3><span class="title">DBAPI -- Generic SQL Database Connectivity</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">FreeTDS</span> database driver -- now works on <span class="nctnt ncbi-monospace">MS Windows</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Refactoring of <span class="nctnt ncbi-path">dbapi/driver/samples</span> DBAPI sample applications -- refactored, now all are based on a common class <span class="nctnt ncbi-class">CDbapiSampleApp</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CDBAPI_Cache</span> -- code moved into a separate library, which is set up as a "standard NCBI plug-in", with registration entry point, fixed interface, class factory, driver name, versioning, etc.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.10"><h3><span class="title">ALGO/ALIGN -- Generic Alignment Algorithms</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Reorganzation of the <span class="nctnt ncbi-path">algo/align</span> branch:  <span class="nctnt ncbi-lib">XALGOALIGN</span> moved under <span class="nctnt ncbi-path">algo/align/nw</span> and renamed to <span class="nctnt ncbi-lib">XALGOALIGNNW</span>. <span class="nctnt ncbi-lib">XALGOSPLIGN</span> renamed to <span class="nctnt ncbi-lib">XALGOALIGNSPLIGN</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSplignFormatter</span> -- Seq-align output now includes percent identity score. </p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSplignFormatter</span> -- was modified to use CSeq_id objects in place of strings.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.11"><h3><span class="title">BLAST</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CBlastNucleotideOptionsHandle</span> -- removed some member functions to simplify its usage.</p></dd><dt>2</dt><dd><p class="no_top_margin">Applications under <span class="nctnt ncbi-path">src/algo/blast/api/demo</span> and <span class="nctnt ncbi-path">src/app/blast_client</span> are NOT meant to provide full functionality of the BLAST C Toolkit binaries, please do not use them as such.</p></dd><dt>3</dt><dd><p class="no_top_margin">Improved error handling when initializing <span class="nctnt ncbi-class">BlastSeqSrc</span> implementations.</p></dd><dt>4</dt><dd><p class="no_top_margin">Removal of <span class="nctnt ncbi-var">eSkipTbck</span> from <span class="nctnt ncbi-var">EBlastTbackExt</span> enumeration, consolidation of means to avoiding the traceback stage of the algorithm.</p></dd><dt>5</dt><dd><p class="no_top_margin">Removal of total HSP limit option from low level structures and options API.</p></dd><dt>6</dt><dd><p class="no_top_margin">Made fetching sequences via the C++ <span class="nctnt ncbi-lib">Object Manager</span> more robust and consistent with the C Toolkit.</p></dd><dt>7</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">BlastQueryInfo</span> structure -- refactored.</p></dd><dt>8</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-lib">RPS-BLAST</span> now supports multiple queries. Streamlined the initialization of <span class="nctnt ncbi-lib">RPS-BLAST </span>structures.</p></dd><dt>9</dt><dd><p class="no_top_margin">Reorganization of <span class="nctnt ncbi-class">Blast*Options</span> and <span class="nctnt ncbi-class">Blast*Parameter</span> structures.</p></dd><dt>10</dt><dd><p class="no_top_margin">Various file and functions renamed to improve clarity of the code and adherence to C++ Toolkit coding conventions.</p></dd><dt>11</dt><dd><p class="no_top_margin">Implemented reevaluation with ambiguities for translated ungapped searches.</p></dd><dt>12</dt><dd><p class="no_top_margin">Added methods to retrieve warnings/errors from <span class="nctnt ncbi-class">CBl2Seq</span> and <span class="nctnt ncbi-class">CDbBlast</span>.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.12"><h3><span class="title">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSeq_id</span> -- support several new prefixes in <span class="nctnt ncbi-func">IdentifyAccession()</span> method.</p></dd><dt>2</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CGen_code_table</span> -- new <span class="nctnt ncbi-func">LoadTransTable()</span> method to support optionally replacing the built-in copy with an updated or customized version.</p></dd><dt>3</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CSeq_id_Handle</span> -- added helper functions to avoid getting intermediate instance of <span class="nctnt ncbi-class">CSeq_id_Mapper</span>: <span class="nctnt ncbi-func">HaveMatchingHandles()</span>, <span class="nctnt ncbi-func">HaveReverseMatch()</span>, <span class="nctnt ncbi-func">GetMatchingHandles()</span>, <span class="nctnt ncbi-func">GetReverseMatchingHandles()</span>.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.13"><h3><span class="title">ALGO/PHY_TREE -- Phylogenetic and bio tree algorithms</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added new <span class="nctnt ncbi-app">test-demo</span> to <span class="nctnt ncbi-path">algo/phy_tree/test/test_biotree.cpp.</span></p></dd><dt>2</dt><dd><p class="no_top_margin">New converters for phylogenetic trees. <span class="nctnt ncbi-class">PhyTree</span> serialization support.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.14"><h3><span class="title">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added class <span class="nctnt ncbi-class">CTSE_Handle</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Implemented auto-release of unused <span class="nctnt ncbi-var">TSE</span>s in scope. <span class="nctnt ncbi-var">TSE</span>s are locked from release by any object manager handle or iterator. Scope keeps last few unlocked <span class="nctnt ncbi-var">TSE</span>s from immediate release. Scope also maintains links from one <span class="nctnt ncbi-var">TSE</span> to indirectly used ones,    like <span class="nctnt ncbi-var">TSE</span>s with far segments of segmented sequence. Number of unlocked <span class="nctnt ncbi-var">TSE</span>s can be changed by configuration variable <span class="nctnt ncbi-var">SCOPE_AUTORELEASE_SIZE</span> in <span class="nctnt ncbi-var">OBJMGR</span> section of configuration file. <span class="nctnt ncbi-var">TSE</span> auto-release can be turned off by setting <span class="nctnt ncbi-var">SCOPE_AUTORELEASE</span>  boolean flag to <span class="nctnt ncbi-var">FALSE</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-class">CSeqVector</span> constructor from <span class="nctnt ncbi-class">CBioseq_Handle</span> to allow used <span class="nctnt ncbi-var">TSE</span> linking.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">CSeqMap::CanResolveRange()</span> with <span class="nctnt ncbi-var">SSeqMapSelector</span> argument to allow used <span class="nctnt ncbi-var">TSE</span> linking.</p></dd><dt>5</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">SAnnotSelector::SetExcludeExternal()</span>.</p></dd><dt>6</dt><dd><p class="no_top_margin">Added helper class <span class="nctnt ncbi-class">CBlobIdKey</span> to use it as key, and to print <span class="nctnt ncbi-var">BlobId</span>.</p></dd><dt>7</dt><dd><p class="no_top_margin">Added method <span class="nctnt ncbi-func">CSeq_annot_Handle GetAnnot()</span> to various annotation iterators: <span class="nctnt ncbi-class">CFeat_CI</span>, <span class="nctnt ncbi-class">CGraph_CI</span>,and <span class="nctnt ncbi-class">CAlign_CI</span>.</p></dd><dt>8</dt><dd><p class="no_top_margin">Changed <span class="nctnt ncbi-var">TSE</span> suppression levels to blob state flags.</p></dd><dt>9</dt><dd><p class="no_top_margin">Removed many methods previously marked as deprecated.</p></dd><dt>10</dt><dd><p class="no_top_margin">Fixed sorting of circular features.</p></dd><dt>11</dt><dd><p class="no_top_margin">Redesigned <span class="nctnt ncbi-class">CBioseq_CI</span> not to collect all bioseqs in constructor.</p></dd><dt>12</dt><dd><p class="no_top_margin">Added <span class="nctnt ncbi-func">CBioseq_Handle::GetRangeSeq_loc()</span>: creates <span class="nctnt ncbi-class">CSeq_loc</span> to be used with annotation iterators instead of start/stop values.</p></dd><dt>13</dt><dd><p class="no_top_margin">Added proxy methods for <span class="nctnt ncbi-class">CSeq_annot</span> getters.</p></dd><dt>14</dt><dd><p class="no_top_margin">Reduced number of <span class="nctnt ncbi-func">CSeqMap::FindResolved()</span> methods, simplified <span class="nctnt ncbi-func">BeginResolved()</span> and <span class="nctnt ncbi-func">EndResolved()</span>.</p></dd><dt>15</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-class">CTSE_Handle</span> -- added new methods to get blob state <span class="nctnt ncbi-func">IsSuppressed*()</span> and <span class="nctnt ncbi-func">IsDead()</span>.</p></dd><dt>16</dt><dd><p class="no_top_margin">Simplified resulting seq-loc in Add/Merge/Subtract. Moved seq-loc operations to <span class="nctnt ncbi-class">CSeq_loc</span>, modified flags. Removed old <span class="nctnt ncbi-func">SeqLocMerge</span>.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.15"><h3><span class="title">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">LDS data loader now can read sequence entries out of binary ANS.1 <span class="nctnt ncbi-var">RefSeq</span> release files.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added possibility to reload <span class="nctnt ncbi-var">TSE</span>s by their <span class="nctnt ncbi-var">BlobId</span> in GenBank loader.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added conversion of <span class="nctnt ncbi-var">BlobId</span> to string in GenBank loader.</p></dd><dt>4</dt><dd><p class="no_top_margin">Merged "ID1" and "cached ID1" libraries. ID1 reader class factory now can create any of these two -- depending on parameters.</p></dd><dt>5</dt><dd><p class="no_top_margin">Loaders and readers are now set up as "standard NCBI plug-ins", with registration entry point, fixed interface, class factory, driver name, versioning, etc.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.16"><h3><span class="title">BUILD FRAMEWORK (UNIX)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">CONFIGURE now allows for a wider range of <span class="nctnt ncbi-lib">Berkeley DB</span> layouts by honoring settings of <span class="nctnt ncbi-var">BERKELEYDB_{INCLUDE,LIBPATH,LIBS}</span> from the environment when --with-berkeleydb=... has not been supplied. </p></dd><dt>2</dt><dd><p class="no_top_margin">Makefiles may specify <span class="nctnt ncbi-var">DTD_PROJ</span>, a la <span class="nctnt ncbi-var">ASN_PROJ</span>.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.17"><h3><span class="title">BUILD FRAMEWORK (MSVC++.NET)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Modified <span class="nctnt ncbi-path">import_project.wsf</span> and <span class="nctnt ncbi-path">new_project.wsf</span> to reduce dependency on CVS.</p></dd><dt>2</dt><dd><p class="no_top_margin">Changed <span class="nctnt ncbi-path">new_project.wsf</span> to generate new project in a separate folder, and to make it possible to enter the data interactively.</p></dd><dt>3</dt><dd><p class="no_top_margin">Improved diagnostics output.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.18"><h3><span class="title">PTB -- Project Tree Builder for MSVC++ .NET</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Generate and use configuration-dependent site localization file - <span class="nctnt ncbi-path">ncbiconf_msvc_site.$(ConfigurationName).h</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Configure libraries with choice and 3rd-party library dependencies and macros for each build configuration independently.</p></dd><dt>3</dt><dd><p class="no_top_margin">In 3rd-party library description (<span class="nctnt ncbi-path">project_tree_builder.ini</span>): allow macros, treat library <span class="nctnt ncbi-var">INCLUDE</span> as list, and made it possible to add standard libraries along with 3rd-party ones (<span class="nctnt ncbi-var">STDLIB</span>).</p></dd><dt>4</dt><dd><p class="no_top_margin">Corrected handling of project lists, and configurable macros. Correctly process MSVC project tuning files for DLLs (<span class="nctnt ncbi-path">Makefile.*.msvc</span>). In generated solution, correctly set dependency on user projects.</p></dd><dt>5</dt><dd><p class="no_top_margin">Process <span class="nctnt ncbi-var">DTD_PROJ</span> entry in <span class="nctnt ncbi-path">Makefile.in</span> - to generate library project by a DTD specification.</p></dd><dt>6</dt><dd><p class="no_top_margin">Improved diagnostic output.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.19"><h3><span class="title">3RD-PARTY PACKAGES</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Upgraded code to work with newer versions of some 3rd-party packages.</p></dd><dt>2</dt><dd><p class="no_top_margin">The 3rd-party sources' bundle for MSVC++ has some packages upgraded to their newer version.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.3.20"><h3><span class="title">PYTHON -- Scripting Language Support</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin"><span class="nctnt ncbi-path">dbapi/lang/python</span> -- the first draft of a Python database extension module based on the NCBI C++ Toolkit's DBAPI. It implements a big part of the Python Database API Specification v2.0 (http://www.python.org/peps/pep-0249.html), and allows Python developers to work with the databases supported by DBAPI (such Sybase and MS SQL Server.</p></dd></dl></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.4"><h2><span class="title">APPLICATIONS</span></h2><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.4.1"><h3><span class="title">APP/NetCache -- General Purpose Network Cache</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Fixed a number of bugs, improved performance, added logging. Some optimization changes to reduce server choking under heavy load.</p></dd></dl></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.5"><h2><span class="title">Documentation</span></h2><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.5.1"><h3><span class="title">Document Location</span></h3><p>The documentation is available online at <!--ro--><a href="toc.html">https://www.ncbi.nlm.nih.gov/books/bv.fcgi?call=bv.View..ShowTOC&amp;rid=toolkit.TOC&amp;depth=2</a> as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release. The PDF version can be accessed by a link that appears on each page.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.5.2"><h3><span class="title">Document Content</span></h3><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit" gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
        <a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/">https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
      </p><p>The above link is also available under the "Browsers" that appears on each page.</p><p>You can also access the CVS code repository via a Web interface. These links also appear in the sidebar box on each page.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the public or in-house versions of LXR, Doxygen and Library. The Library search runs a CGI script that lists the library name where the symbol (such as function) is defined in.</p><p>The current release notes as well as past release notes are now in an appendix in the C++ Toolkit Book and a link to the current release notes appears on each page of the online Toolkit document.</p></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.6"><h2><span class="title">Building on the MacOS</span></h2><p>We now build all the libraries and most of the applications including the Genome Workbench (<span class="nctnt ncbi-app">gbench</span>), some test and a few demo applications. We also build the FLTK library's GUI editor, <span class="nctnt ncbi-app">fluid</span>.</p><p>All apps are built as application bundles except <span class="nctnt ncbi-app">gbench_plugin_scan</span> and <span class="nctnt ncbi-app">datatool</span> which are built as command line apps. Any of the applications can be built as command line apps by tweaking the build scripts or the CodeWarrior projects.</p><p>When building the toolkit with xCode, the latest version of xCode (at least 1.5) is required for the trouble-free build. Build procedure is as follows: open, build and run a project file in compilers/xCode.  This is a GUI tool to generate a new NCBI C++ Toolkit xCode project.  You'll have an option to specify third-party installation directories and choose which packages (libs, applications and tests) to include into the final project.</p><p>xCode build fully support all the latest Apple innovations: distributed  builds, optional CPU specific optimization , pre-compiled headers, fix  &amp; continue, code cleanup and Zero Link (with few exceptions).</p><p>xCode builds all libraries as a Mach-O dynamically linked shared ones (.dylib) and all Genome Workbench plugins as Mach-O bundles (also .dylib extension). Note, that xCode will place Genome Workbench plugins inside Genome Workbench application bundle (Genome Workbench.app/Contents/MacOS/plugins).</p><p>To build the toolkit from a command line use an AppleScript editor to open and run the script files makeLibs.met and makeApps.met. You must use a script editor capable of opening a script file larger than 32K, such as Apple's Script Editor v2.0, or Smile. Script Editor v1.9 will not work since makeLibs.met just got too big. The command line tool osascript also works.</p><p>Projects include targets to compile with BSD/Apple headers and libraries and with MSL headers and libraries, but plugins (for gbench) built with MSL do not link properly, and so, because of lack of interest to keep up with changes, some source code does not currently compile with MSL. Hopefully this will become easier to work with and get fixed with CodeWarrior v.9.</p><p>Targets to be compiled can be controlled by including an empty file or folder in the compilers:mac_prj folder with the name 'Build' followed by the keywords of the targets you want built. The keywords are: MSL, BSD, Debug and Final. For example, to build only the BSD Debug targets use: "Build BSD Debug", to build both BSD debug and release (final) versions: "Build BSD". The default is to build everything.</p><p>If you install the C++ toolkit under a different name than "ncbi_cxx" or in a different location than your home directory, you can edit the script's properties, pRootFolderName and pRootFolderPath, to override these defaults. Note: these paths, and those mentioned below, must be entered in mac format (e.g. disk:Users:username:) not Unix format (e.g. <span class="nctnt ncbi-path">/Users/username/</span>). The disk name (and its following colon) may be omitted.</p><p>Certain third party libraries (see <a href="release_notes_03_09_2005.html#release_notes_03_09_2005.release_notes_2_2005.T1">Table 1</a>) are required to build some parts of the C++ toolkit. The scripts will try and find them if they are in your home directory, or you can specify where they were installed using properties at the beginning of the script.</p><div class="table" id="release_notes_03_09_2005.release_notes_2_2005.T1"><h4><span class="title">Table 1. Third Party Libraries</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Library </th><th align="left" valign="top" rowspan="1" colspan="1">Property </th><th align="left" valign="top" rowspan="1" colspan="1">Example</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-lib">FLTK 1.1.5rc2</span> (w/ NCBI patches) </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">pFLTKRootFolder </span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-monospace">"usr:local:src:fltk-1.1.5rc2:"</span>
            </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-lib">BerkeleyDB 4.3.21</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">pBdbRootFolder </span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-monospace">"Users:myhome:mylibs:db-4.3.21"</span>
            </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-lib">SQLite 2.8.13</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">gSqliteFolder</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-lib">dlcompat</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-var">pDLRootFolder</span>
            </td><td align="left" valign="top" rowspan="1" colspan="1">
              <span class="nctnt ncbi-monospace">"usr:"</span>
            </td></tr></tbody></table></div></div><p>You do not have to build the FLTK or BDB libraries separately. This is done by the scripts and CodeWarrior along with the toolkit libraries. Just unpack the source bundles in your home directory or where ever you have specified in the appropriate properties. The root folders for FLTK and BDB do not have to have any particular names. If there is more than one version the scripts will grab whichever is last alphabetically (e.g. fltk-1.1.4r2 will get used instead of fltk-1.1.3).</p><p>The scripts normally halt on any CodeWarrior compilation errors. If you want them to continue and save errors, set the next script property, pSaveContinueOnErrors, to true. Compilation errors for a project will be saved in a file in the same folder as the project being built, with a name in the following format: projectName-targetNumber.errs (e.g. xncbi-2.errs).</p><p>The Genome Workbench's configuration file(s) is stored in the user's <span class="nctnt ncbi-path">Library:Application Support:gbench</span> folder.</p><p>CFM builds are not supported. OS 8 or 9 are not supported. We know 10.2 works. We think 10.1 still works, and 10.3 might work.</p></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.7"><h2><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h2><p>This release was successfully tested on at least the following platforms -- but may also work on other platforms. Since the previous release, some platforms were dropped from this list, just because we do not use them here anymore, and some were added (these new platforms are highlighted using bold font). Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages -- in these cases, just skipping such projects (e.g. using flag "-k" for <span class="nctnt ncbi-app">make</span> on UNIX), can get you through.</p><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.7.1"><h3><span class="title">Unix</span></h3><div class="table" id="release_notes_03_09_2005.release_notes_2_2005.T2"><h4><span class="title">Table 2. Unix OS's and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC </span>2<b>.3.2</b><span class="nctnt ncbi-monospace">) </span></td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL </span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 2.95.3, 3.0.4, 3.4.0</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Linux-2.4.26 (w/ LIBC 2.3.2)</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.0</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Linux-2.6.7 (w/ LIBC 2.3.3)</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL/64</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4<b>.3</b></span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.2.5)</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">ICC 8.0</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-8</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">SPARC</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13</span> (64-bit, 32-bit)
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-8</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">SPARC</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3<b>.4.3</b></span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-9</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Solaris-9</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3<b>.4.3</b></span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">IRIX64-6.5</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">SGI-Mips</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">MIPSpro 7.3.1<b>.3</b>m (64-bit, 32-bit)</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">FreeBSD-4.10</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">INTEL</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.4.2</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">ALPHA</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.3.2</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">ALPHA</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">Compaq C++ V6.5-014</span>
              </td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.7.2"><h3><span class="title">MS Windows</span></h3><div class="table" id="release_notes_03_09_2005.release_notes_2_2005.T3"><h4><span class="title">Table 3. MS Windows and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MS Windows</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-app">MSVC++ 7.1</span>. See documentation on
              <a href="ch_config.html#ch_config.MS_Windows">MS Visual C++.NET</a>.</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.7.3"><h3><span class="title">Mac OS X</span></h3><div class="table" id="release_notes_03_09_2005.release_notes_2_2005.T4"><h4><span class="title">Table 4. Mac OS, and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MacOS 10.2, 10.3</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">GCC 3.3</span>
              </td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-monospace">MacOS 10.3</span>
              </td><td align="left" valign="top" rowspan="1" colspan="1">
                <span class="nctnt ncbi-app">CodeWarrior 9.2</span>
              </td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8"><h2><span class="title">Caveats and Hints</span></h2><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8.1"><h3><span class="title">MacOS 10.X / CodeWarrior 9.2</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Not all of the test or demo applications are built.</p></dd><dt>2</dt><dd><p class="no_top_margin">The source code for the latest release of FLTK (1.1.x), BerkeleyDB (4.x) and SQLite (2.x) should be present. See the installation instructions for details.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8.2"><h3><span class="title">MacOS 10.2/GCC 3.3</span></h3><p>At least the GCC 3.3 update for Dec. 2002 Developers Tools required from Apple.</p></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8.3"><h3><span class="title">GCC 2.95</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Poor MT-safety record.</p></dd><dt>2</dt><dd><p class="no_top_margin">Relatively incomplete/incorrect (comparing to modern compilers) STL implementation.</p></dd><dt>3</dt><dd><p class="no_top_margin">It is going to be deprecated in NCBI rather soon -- as soon as we have any significant trouble with its maintenance.</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8.4"><h3><span class="title">GCC 3.0.4</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body (fixed in GCC 3.3).</p></dd><dt>2</dt><dd><p class="no_top_margin">STL stream uses locale in thread unsafe way which may result to segmentation fault when run in multithread mode (fixed in GCC 3.3).</p></dd><dt>3</dt><dd><p class="no_top_margin">Long-file support for C++ streams is disabled/broken (first broken in 3.0, fixed in 3.4).</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8.5"><h3><span class="title">GCC 3.3</span></h3><p>Other than the feature described below, GCC 3.3.2 had been very good to us; it had a lot of very ugly bugs finally fixed.</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Painfully slow linking in debug mode on Linux with GCC-3.3 compiler. -- Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.  GCC starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking. GCC 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4. The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug (N^2). Binutils 2.15 fixes this. The link time still 2 times slower than without symbol merge, but the resultant executable is about two times smaller in size, and no compiler patching is necessary. We are still testing it in-house. We had to patch GCC 3.3 in-house with the fix described at http://lists.boost.org/MailArchives/boost/msg53004.php.</p></dd><dt>2</dt><dd><p class="no_top_margin">Long-file support still broken (first broken in 3.1).</p></dd></dl></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.8.6"><h3><span class="title">GCC 3.4</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">The "Painfully slow linking..." (see GCC3.3 [1] above) was an issue, and we had to patch it in-house to speed up, a la GCC 3.3 -- until we finally upgraded to binutils 2.15.</p></dd><dt>2</dt><dd><p class="no_top_margin">At least on Linux, ifstream::readsome() does not always work for large files, as it calls an ioctl that doesn't work properly for large files.</p></dd><dt>3</dt><dd><p class="no_top_margin">At least on Linux, GCC 3.4.[0,1] optimizer (very rarely) generates incorrect code when comparing enumerated values in else-ifs. (Fixed in 3.4.2)</p></dd></dl></div></div><div class="sec" id="release_notes_03_09_2005.release_notes_2_2005.1.9"><h2><span class="title">Last Updated</span></h2><p>This section last updated on March 9, 2005.</p></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (February, 2005)
                    .
        [Updated: February 1, 2005]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_03_09_2005</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.1">Download Location</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.2">Source Archive Contents</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.3">New Development</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.4">APPLICATIONS</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.5">Documentation</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.6">Building on the MacOS</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.7">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.8">Caveats and Hints</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.9">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (February, 2005)
                    .
        [Updated: February 1, 2005]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: https://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_03_09_2005</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.1">Download Location</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.2">Source Archive Contents</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.3">New Development</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.4">APPLICATIONS</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.5">Documentation</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.6">Building on the MacOS</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.7">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.8">Caveats and Hints</a></li><li><a href="#release_notes_03_09_2005.release_notes_2_2005.1.9">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="https://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
