<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (December 8, 2003) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_12_08_2003" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_april_16_2004.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_08_01_2003.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_12_08_2003.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (December 8, 2003)</span></h1></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.download_loc_12">Download Location</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.source_archive_">Source Archive Contents</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.new_development">New Development</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.build_framework">Build Framework</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.documentation_1">Documentation</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.platforms_12_08">Platforms (OS's, compilers used inside NCBI)</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.caveats_12_08_2">Caveats and Hints</a></p></li><li><p><a href="release_notes_12_08_2003.html#release_notes_12_08_2003.last_updated_12">Last Updated</a></p></li></ul><div class="sec" id="release_notes_12_08_2003.content"><h2><span class="title">Release Notes (December 8, 2003)</span></h2><div class="sec" id="release_notes_12_08_2003.download_loc_12"><h3><span class="title">Download Location</span></h3><p>
				<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2003/Dec_08_2003/">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2003/Dec_08_2003/</a>
			</p></div><div class="sec" id="release_notes_12_08_2003.source_archive_"><h3><span class="title">Source Archive Contents</span></h3><div class="sec" id="release_notes_12_08_2003.source_code_12_"><h4><span class="title">Source Code Archives</span></h4><ul><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_unix--Dec_08_2003.tar.gz</span>-- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win_msvc6--Dec_08_2003.exe</span>-- for MS-Windows / MSVC++ 6.0 (self-extracting)</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win_msvc6--Dec_08_2003.zip</span>-- for MS-Windows / MSVC++ 6.0</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win_msvc7--Dec_08_2003.exe</span>-- for MS-Windows / MSVC++ 7.0 (semi-experimental)</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win_msvc7--Dec_08_2003.zip</span>-- for MS-Windows / MSVC++ 7.0 (semi-experimental)</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_cw--Dec_08_2003.tar.gz</span>-- for MacOS 10.2 / CodeWarrior DevStudio for MacOS 9.1</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_gcc--Dec_08_2003.tar.gz</span>-- for MacOS 10.2 / GCC 3.3</p></li></ul></div><div class="sec" id="release_notes_12_08_2003.other_files_12_"><h4><span class="title">Other Files</span></h4><ul><li><p>timestamp -- when the sources' snapshot was made</p></li></ul></div></div><div class="sec" id="release_notes_12_08_2003.new_development"><h3><span class="title">New Development</span></h3><div class="sec" id="release_notes_12_08_2003.corelib_12_08_2"><h4><span class="title">Portability and Application Framework (corelib)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CPipe</span>class moved to xconnect library.</p></dd><dt>2</dt><dd><p class="no_top_margin">New process management class<span class="nctnt ncbi-class">CProcess</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">The<span class="nctnt ncbi-class">CPIDGuard</span>class moved from<span class="nctnt ncbi-path">ncbi_system.[ch]pp</span>to<span class="nctnt ncbi-path">ncbi_process.[ch]pp</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-class">CMetaRegistry</span>to centralize configuration file search logic and allow potential reuse of parsed contents.</p></dd><dt>5</dt><dd><p class="no_top_margin">Added a cross-platform<span class="nctnt ncbi-class">CPIDGuard</span>for the sake of daemons.</p></dd><dt>6</dt><dd><p class="no_top_margin">Added optional symlink resolution to<span class="nctnt ncbi-func">CDirEntry::NormalizePath</span>.</p></dd><dt>7</dt><dd><p class="no_top_margin">Made it possible to alter the verbosity of trace messages.</p></dd><dt>8</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-class">CDllResolver</span>- class scanning for DLLs and looking for  specified entry points.</p></dd><dt>9</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-path">corelib/plugin_manager.hpp</span>- Developed initial version of plugin manager (cannot be considered  as a true release, but rather a pre-beta version)</p></dd><dt>10</dt><dd><p class="no_top_margin">Fixes and extensions<ul><li><p>File API: Fixed bug in the File API<span class="nctnt ncbi-func">GetType()</span>with get type of a   dir entry on UNIX. Replaced using<span class="nctnt ncbi-func">tmpnam()</span>with<span class="nctnt ncbi-func">tempnam()</span>/<span class="nctnt ncbi-func">mkstemp()</span>for creating temporary files in the<span class="nctnt ncbi-func">GetTmpName[Ex]</span>. Added functions<span class="nctnt ncbi-func">DeleteTrailingPathSeparator()</span>and<span class="nctnt ncbi-func">NormalizePath()</span>to the<span class="nctnt ncbi-class">CDirEntry</span>class. Added month and day of week names conversion functions to the<span class="nctnt ncbi-class">CTime</span>class.</p></li><li><p>Added month and day of week names conversion functions to the  CTime class.</p></li><li><p>Exec API: Used correct function for run a processes on Unix for   the eV-mode.</p></li><li><p>DLL API: Changed<span class="nctnt ncbi-class">CDll</span>class to better conform C/C++ standards.</p></li><li><p>Revamped<span class="nctnt ncbi-func">GetEntryPoint()</span>method to avoid mixing a pointers to data and functions.</p></li><li><p>File API: Added<span class="nctnt ncbi-func">CDirEntry::SetTime()</span>method to change file access and modification times.</p></li><li><p>
										<span class="nctnt ncbi-class">CTime</span>class: Added setters for various components of time, added  functions to determine number of week in the year and month, number of  days in the month, reimplemented<span class="nctnt ncbi-func">AddYear()</span>, changed<span class="nctnt ncbi-func">DayOfWeek()</span>function  to use other computation method.</p></li></ul>
						</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.connect_12_08_2"><h4><span class="title">Data streaming, Networking, and Dispatching (connect)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-path">ncbi_socket.[ch][pp]</span>:<ul><li><p>Bug-fixes for proper destruction of underlying SOCK in case of<span class="nctnt ncbi-func">Close()</span>or<span class="nctnt ncbi-func">Abort()</span>methods got called, but the<span class="nctnt ncbi-class">CSocket</span>is still in use.</p></li><li><p>Stall protection against waiting indefinitely if SOCK that is being  waited for is closed from other thread (Linux specific issue).</p></li></ul>
						</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-path">ncbi_http_connector.[ch]</span>:<ul><li><p>Do not call<span class="nctnt ncbi-func">shutdown()</span>after sending HTTP request. Although 100% valid this call could cause some buggy software and  hardware to react inappropriately as if the connection is going to be  closed soon.</p></li><li><p>Support VHosts by passing "Host: " tag in the HTTP headers (essential  if VHosts are being used with Apache web server software).</p></li></ul>
						</p></dd><dt>3</dt><dd><p class="no_top_margin">Pipes and named pipes moved from CORELIB into CONNECT.</p></dd><dt>4</dt><dd><p class="no_top_margin">Abstract classes<span class="nctnt ncbi-class">IReader</span>and<span class="nctnt ncbi-class">IWriter</span>in UTIL library. Implementation of stream buffer on top of these classes.</p></dd><dt>5</dt><dd><p class="no_top_margin">Fix for<span class="nctnt ncbi-func">CStreamUtils::Readsome()</span>and<span class="nctnt ncbi-func">Pushback()</span>API. Make all basic Toolkit streambufs (including those in CONNECT) standard-compliant with respect to<span class="nctnt ncbi-func">xsgetn()</span>operation.</p></dd><dt>6</dt><dd><p class="no_top_margin">Performance and standard-compilance of FCGI I/O streams: FCGI I/O streams (cgi library) are significantly improved by  eliminating intermediary buffers, and making the streams fully  buffered directly on top of FCGIX structures.</p></dd><dt>7</dt><dd><p class="no_top_margin">Performance of fstreams on Windows: MS-Windows (MSVC/6.0): fstreams are significantly improved by making  an ad-hoc fix to make fstreams buffered by means of direct use of  FILE* buffer pointers, and thus eliminating 1-byte<span class="nctnt ncbi-func">fgetc()</span>calls on  every I/O.</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.dbapi"><h4><span class="title">Database Connectivity (DBAPI) (dbapi)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">util/cache: Added specification for local cache interface (Work in progress,  current version looks unstable and likely to change)</p></dd><dt>2</dt><dd><p class="no_top_margin">Local Data Storage (LDS):<ul><li><p>Fixed number of bugs</p></li><li><p>Implemented recursive objects directory scan, now all subdirs are  part of the database (Improved LDS data management)</p></li><li><p>Implemented multiple LDS databases. Changed genome workbench to  reflect this new feature.</p></li><li><p>Added PID protection againts running several instances of program  with the same local database (not supported yet and caused crash)</p></li></ul>
						</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.berkeley_dbapi_"><h4><span class="title">Berkeley DB API (bdb)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added class for Berkeley DB environment support</p></dd><dt>2</dt><dd><p class="no_top_margin">Implemented stream like class on top of BDB BLOB (<span class="nctnt ncbi-class">CBDB_BLobStream</span>).</p></dd><dt>3</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-class">CBDB_FileDumper</span>class to dump BDB files into text format</p></dd><dt>4</dt><dd><p class="no_top_margin">Implemented local cache for biological objects.</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.serial_12_08_20"><h4><span class="title">Data Serialization (serial)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Data initialization verification mechanism in I/O streams has been  elaborated to allow for handling (skipping or using defaults) the  unset mandatory data members both on reading and writing.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added support of XML ANY data type to the<span class="nctnt ncbi-app">DATATOOL</span>(C++ code  generator) and XML serialization I/O streams.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added possibility to specify XML namespaces for serial objects.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added support of multiple namespaces in XML streams.</p></dd><dt>5</dt><dd><p class="no_top_margin">Implemented basic XML SOAP packager, which includes an API to  create a SOAP message, serialize (read or write) it, and investigate  its contents.</p></dd><dt>6</dt><dd><p class="no_top_margin">Added possibility to reset an object member from a read hook  (including non-optional members).</p></dd><dt>7</dt><dd><p class="no_top_margin">Redesigned<span class="nctnt ncbi-app">DATATOOL</span>to use new approach for type aliases. Instead of using wrapper classes, the new<span class="nctnt ncbi-app">DATATOOL</span>derives alias class  from the aliased one if the aliased type is a class, or from a special  template class in case of primitive types.</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.obj_mgr_12_08_2"><h4><span class="title">Object Manager (objects/objmgr)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added or modified several seq-loc related functions to handle  circular locations correctly (<span class="nctnt ncbi-func">GetStart()</span>,<span class="nctnt ncbi-func">GetEnd()</span>,<span class="nctnt ncbi-func">GetCircularLength()</span>,<span class="nctnt ncbi-func">TestForOverlap()</span>).</p></dd><dt>2</dt><dd><p class="no_top_margin">Introduced new class<span class="nctnt ncbi-class">CAnnot_CI</span>which allows to iterate over  seq-annot container objects rather than over individual annotations.</p></dd><dt>3</dt><dd><p class="no_top_margin">Implemented caching id1 reader (uses BDB library to keep BLOBs and  id resolution information on local disk).</p></dd><dt>4</dt><dd><p class="no_top_margin">Integrated local caching in Genome Workbench</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.alnmgr_12_08_20"><h4><span class="title">Alignment Manager (objects/alnmgr)</span></h4><p>Added support for translated (nucleotide to amino acid) alignments.</p></div><div class="sec" id="release_notes_12_08_2003.seq_align"><h4><span class="title">Sequence Alignment (objects/seqalign)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Extended<span class="nctnt ncbi-class">CDense_seg</span>with<span class="nctnt ncbi-var">m_Widths</span>to support translated alignments.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added a few methods for Seq-aligns (mostly for the Dense-seg type):<ul><li><p>Alignment transformers: Reverse, SwapRows for<span class="nctnt ncbi-class">CSeq_algin</span>and<span class="nctnt ncbi-class">CDense_seg</span>, and<span class="nctnt ncbi-func">CDense_seg::RemapToLoc</span>
									</p></li><li><p>Validators:<span class="nctnt ncbi-class">CheckNumRows</span>,<span class="nctnt ncbi-class">CheckNumSegs</span>and Validate with optional   full segments check.</p></li><li><p>Methods for range calculation:<span class="nctnt ncbi-func">GetSeqRange</span>,<span class="nctnt ncbi-func">GetSeqStart</span>,<span class="nctnt ncbi-func">GetSeqStop</span>.</p></li></ul>
						</p></dd><dt>3</dt><dd><p class="no_top_margin">Added a format converter:<span class="nctnt ncbi-func">CSeq_align::CreateDensegFromStdseg</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added validators.</p></dd><dt>5</dt><dd><p class="no_top_margin">Added methods for range calculation.</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.algo_align_12_0"><h4><span class="title">Generic Sequence Alignment Algorithms (algo/align)</span></h4><p>The following changes and updates have been made to the algo/align library.</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Formatting-related functionality has been moved from the root<span class="nctnt ncbi-class">CNWAligner</span>to a new<span class="nctnt ncbi-class">CNWFormatter</span>class. Code using old aligner's formatting members will work fine with minimal adjustment. For details please see the updated demo applications.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CNWAlignerMrna2Dna</span>was replaced with three classes to facilitate quality/performance trade-off in case of distorted sequences:<ul><li><p>
										<span class="nctnt ncbi-class">CSplicedAligner</span>.  Abstract base for spliced aligners;</p></li><li><p>
										<span class="nctnt ncbi-class">CSplicedAligner16</span>. Accounts for three conventional splices and a generic splice; uses 2 bytes per backtrace matrix cell</p></li><li><p>
										<span class="nctnt ncbi-class">CSplicedAligner32</span>. Accounts for three conventionales and splices that could be produced by damaging one or more bps of any conventional; uses 4 bytes per backtrace matrix cell.</p></li></ul>
						</p></dd></dl><p>Use the 16-bit version of the above classes for low-error sequences, and the 32-bit version for sequences with higher error rates.</p><p>A demo new application called Splign was introduced. Splign is built on top of the<span class="nctnt ncbi-class">CSplicedAligner</span>classes and is a real-world tool for doing spliced sequence alignments.</p></div><div class="sec" id="release_notes_12_08_2003.misc_12_08_2003"><h4><span class="title">Miscellaneous Libs (additions and improvements)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">`xutil' (util): [regexp]  Added wrapper for the Perl-compatible regular expression (PCRE) library and<span class="nctnt ncbi-class">CRegexp</span>class --<span class="nctnt ncbi-class">CRegexpUtil</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">`xhtml' (html) : Updated<span class="nctnt ncbi-class">CPopupMenu</span>class to use new version 2.3 of the Sergey  Kurdin's JavaScript PopupMenu2. Added support for Sergey Kurdin's slide menu.</p></dd><dt>3</dt><dd><p class="no_top_margin">New MSVC 6.0 project file converter to expand a single  configuration project file to multi-configuration project file  (<span class="nctnt ncbi-path">src\util\msvc\one2all.cpp</span>).</p></dd><dt>4</dt><dd><p class="no_top_margin">tables' (util/tables) -- NEW. C API for working with score matrices, along with (generated) hard-coded copies of standard matrices.</p></dd><dt>5</dt><dd><p class="no_top_margin">`xobjread' (objtools/readers).  [fasta] Now reads large FASTA sequences much more efficiently.</p></dd><dt>6</dt><dd><p class="no_top_margin">`xflat' (objtools/flat).  Added support for GFF/GTF [semi-experimental].</p></dd><dt>7</dt><dd><p class="no_top_margin">`general' (objects/general).   Added manipulators for CInt_fuzz.</p></dd><dt>8</dt><dd><p class="no_top_margin">`xobjutil' (objects/util). Improved relative-location support, which now handles fuzz and opposite-strand relative locations.</p></dd><dt>9</dt><dd><p class="no_top_margin">`seq' (objects/seq*).  Added support for several new prefixes to CSeq_id::IdentifyAccession.</p></dd><dt>10</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-path">util/file_obsolete.h</span>: Added<span class="nctnt ncbi-class">CFileObsolete</span>class - removes local files based on file mask  and creation/access time.</p></dd><dt>11</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-path">util/format_guess.hpp</span>:<span class="nctnt ncbi-class">CFormatGuess</span>- implemented new rules of format recognition with  much improved our work with binary ASN.1 files.</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.blast_12_08_200"><h4><span class="title">BLAST algorithms, C and C++ API (algo/blast)</span></h4><p>Integrated BLAST's core "C" (shared with the NCBI C Toolkit) and "C++" API. Actual development done by the BLAST group.</p></div></div><div class="sec" id="release_notes_12_08_2003.build_framework"><h3><span class="title">Build Framework</span></h3><div class="sec" id="release_notes_12_08_2003.config_unix_12_"><h4><span class="title">Configuration on Unix</span></h4><p>The Unix build framework adds make variables ({CC,CXX,LINK}_WRAPPER)  for specifying wrapper programs such as Purify or ccache.  As such,  `configure' now looks for ccache (http://ccache.samba.org/), and takes   advantage of it when present.  NB: ccache does not support Sun's  (WorkShop) C++ compiler, and requires users to set the environment   variable CCACHE_EXTENSION to "i" when using SGI's MIPSpro compiler.</p></div><div class="sec" id="release_notes_12_08_2003.build_macos_12_"><h4><span class="title">Building on the MacOS</span></h4><p>We now build all the libraries and most of the applications including the Genome Workbench (gbench).  We do not build any of the many test or demo apps (except testvalidator).  We also build the FLTK library's GUI editor, fluid.</p><p>All apps are built as application bundles except gbench_plugin_scan and datatool which are built as command line apps.  Any of the applications can be built as command line apps by tweaking the build scripts or the Codewarrior projects.</p><p>Build procedure is still the same: use an AppleScript editor to open and run the script files makeLibs.met and makeApps.met.  You must use a script editor capable of opening a script file larger than 32K, such as Apple's Script Editor v2.0, or Smile.  Script Editor v1.9 will not work since makeLibs.met just got too big.  The command line tool osascript also works.</p><p>Projects include targets to compile with BSD/Apple headers and libraries and with MSL headers and libraries, but plugins (for gbench) built with MSL do not link properly, and so, because of lack of interest to keep up with changes, some source code does not currently  compile with MSL.  Hopefully this will become easier to work with and get fixed with Codewarrior v.9.</p><p>Targets to be compiled can be controlled by including an empty file or folder in the compilers:mac_prj folder with the name 'Build' followed by the keywords of the targets you want built.  The keywords are: MSL, BSD, Debug and Final.  For example, to build only the BSD Debug targets use: "Build BSD Debug", to build both BSD debug and release (final) versions: "Build BSD".  The default is to build everything.</p><p>If you install the C++ toolkit under a different name than "ncbi_cxx" or in a different location than your home directory, you can edit the  script's properties, pRootFolderName and pRootFolderPath, to override these defaults.  Note: these paths, and those mentioned below, must be entered in mac format (e.g. disk:Users:username:) not Unix format (e.g.<span class="nctnt ncbi-path">/Users/username/</span>).  The disk name (and its following colon) may be omitted.</p><p>Certain third party libraries (see<a href="release_notes_12_08_2003.html#release_notes_12_08_2003.T1">Table 4</a>) are required to build the C++ toolkit. The scripts will try and find them if they are in your home directory, or you can specify where they were installed using properties at the beginning of the script.</p><div class="table" id="release_notes_12_08_2003.T1"><h4><span class="label">Table 4</span><span class="title">Third Party Libraries</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Library</th><th align="left" valign="top" rowspan="1" colspan="1">Property</th><th align="left" valign="top" rowspan="1" colspan="1">Example</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-lib">FLTK</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-var">pFLTKRootFolder</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">"usr:local:src:fltk-1.1.4:"</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-lib">DBD</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-var">pBdbRootFolder</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">"Users:myhome:mylibs:db-4.1.25"</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-lib">dlcompat</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-var">pDLRootFolder</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">"usr:"</span>
								</td></tr></tbody></table></div></div><p>The latest release of fltk, 1.1.4 should be used.</p><p>You do not have to build the FLTK or DBD libraries separately.  This is done by the scripts and Codewarrior along with the toolkit libraries.  Just unpack the source bundles in your home directory or where ever you have specified in the appropriate properties.  The root folders for FLTK and DBD do not have to have any particular names.  If there is more than one version the scripts will grab whichever is last alphabetically (e.g. fltk-1.1.4r2 will get used instead of fltk-1.1.3).</p><p>The scripts normally halt on any Codewarrior compilation errors.  If you want them to continue and save errors, set the next script property, pSaveContinueOnErrors, to true.  Compilation errors for a project will be saved in a file in the same folder as the project being built, with a name in the following format: projectName-targetNumber.errs (e.g. xncbi-2.errs).</p><p>The Genome Workbench's configuration file(s) is stored in the users Library:Application Support:gbench folder.</p><p>CFM builds are not supported.  OS 8 or 9 are not supported.  We know 10.2 works.  We think 10.1 still works.</p></div></div><div class="sec" id="release_notes_12_08_2003.documentation_1"><h3><span class="title">Documentation</span></h3><div class="sec" id="release_notes_12_08_2003.doc_loc_12_08_2"><h4><span class="title">Document Location</span></h4><p>The documentation is available  at</p><p>
					<!--ro--><a href="toc.html">http://www.ncbi.nlm.nih.gov/books/bv.fcgi?call=bv.View..ShowTOC&amp;rid=toolkit.TOC&amp;depth=2</a>
				</p><p>as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and  "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_12_08_2003.doc_changes_12_"><h4><span class="title">Document Content</span></h4><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit"  gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
					<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/">http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
				</p><p>The above link is also available under the "Source Code Browsers" that appears on each page.</p><p>A major effort is under way to document all source files, so that the DOXYGEN source browser can generate a detailed API reference for the classes/methods. A top-level meta module grouping exists under "Modules" on the Doxygen main page. This can be used to get a conceptual grouping of the different components.</p></div></div><div class="sec" id="release_notes_12_08_2003.platforms_12_08"><h3><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h3><div class="sec" id="release_notes_12_08_2003.platform_unix_1"><h4><span class="title">Unix</span></h4><p></p><div class="table" id="release_notes_12_08_2003.T2"><h4><span class="label">Table 5</span><span class="title">Unix OS's and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Platform</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4, 3.3.1, 3.3.2</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">ICC 7.1</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SPARC</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13  (64-bit, 32-bit)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SPARC</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 2.95.3, 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">IRIX64</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SGI-Mips</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MIPSpro 7.3.1.2m (64-bit, 32-bit)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">FreeBSD</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Tru64</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">ALPHA</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 2.95.3</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Tru64</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">ALPHA</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">Compaq C V6.3-029 / Compaq C++ V6.5-014</span>
								</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_12_08_2003.platform_window"><h4><span class="title">MS Windows</span></h4><p>MSVC++ 6.0  Service Pack 5.....</p><p>MSVC++ 7.0  (semi-experimental)</p></div><div class="sec" id="release_notes_12_08_2003.platform_mac_12"><h4><span class="title">Mac OS X</span></h4><p></p><div class="table" id="release_notes_12_08_2003.T3"><h4><span class="label">Table 6</span><span class="title">Mac OS, and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS 10.1</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.1 (patched, see "doc/config_darwin.html)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS 10.2</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.1 (patched, see "doc/config_darwin.html), GCC 3.3</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS 10.X</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">CodeWarrior 8.0 Update 8.3</span>
								</td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_12_08_2003.caveats_12_08_2"><h3><span class="title">Caveats and Hints</span></h3><div class="sec" id="release_notes_12_08_2003.macos_10x_12_08"><h4><span class="title">MacOS 10.X / CodeWarrior 9.1</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">None of the test or demo applications are built.</p></dd><dt>2</dt><dd><p class="no_top_margin">The source code for the latest release  of FLTK, 1.1.4 and berkeley database (4.1) should be present. See the release notes (or installation instructions) for details.</p></dd></dl></div><div class="sec" id="release_notes_12_08_2003.macos_10_2_12_0"><h4><span class="title">MacOS 10.2/GCC 3.3</span></h4><p>GCC 3.3 update for Dec. 2002 Developers Tools required from Apple. All C++ Toolkit configurations will build and run just fine.</p></div><div class="sec" id="release_notes_12_08_2003.gcc_3_0_4_12_08"><h4><span class="title">GCC 3.0.4</span></h4><p>Detected two bugs in GCC-3.0.4 compiler:<ul><li><p>Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body.</p><p>Fixed in GCC 3.3.</p></li><li><p>STL stream uses locale in thread unsafe way which may result to segmentation fault.</p><p>Fixed in GCC 3.3.</p></li></ul>
				</p></div><div class="sec" id="release_notes_12_08_2003.gcc_3_3_12_08_2"><h4><span class="title">GCC 3.3</span></h4><p>(Other than the feature described below, GCC 3.3.2 has been very good for us; it has a lot of very ugly bugs finally fixed.)</p><p>Painfully slow linking in debug mode on Linux with GCC-3.3 compiler:<ul><li><p>Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.</p><p>gcc starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking.</p><p>gcc 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4.</p><p>The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug.</p><p>We had to patch GCC 3.3 in-house with the fix described at "http://lists.boost.org/MailArchives/boost/msg53004.php".</p></li></ul>
				</p></div></div><div class="sec" id="release_notes_12_08_2003.last_updated_12"><h3><span class="title">Last Updated</span></h3><p>This section last updated on December 16, 2003</p></div></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (December 8, 2003)
                    . .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_12_08_2003</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (December 8, 2003)
                    . .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_12_08_2003</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
