diff -urN xerces-c-src_2_6_0.orig\src\xercesc\util\MsgLoaders\Icu\ICUMsgLoader.cpp xerces-c-src_2_6_0\src\xercesc\util\MsgLoaders\Icu\ICUMsgLoader.cpp
--- xerces-c-src_2_6_0.orig\src\xercesc\util\MsgLoaders\Icu\ICUMsgLoader.cpp	2004-09-30 14:52:59.000000000 -0400
+++ xerces-c-src_2_6_0\src\xercesc\util\MsgLoaders\Icu\ICUMsgLoader.cpp	2005-04-15 16:24:15.302016200 -0400
@@ -129,6 +129,7 @@
 #include "ICUMsgLoader.hpp"
 #include "unicode/uloc.h"
 #include "unicode/udata.h" 
+#include "unicode/putil.h" 
 
 #include "string.h"
 #include <stdio.h>
@@ -148,11 +149,7 @@
  *  will then  be able to fetch resources from the data.
  */
 
-#if defined(_WIN32) || defined(WIN32)
-extern "C" void U_IMPORT *XercesMessages2_6_0_dat;
-#else
 extern "C" void U_IMPORT *XercesMessages_dat;
-#endif
 
 /* 
  *  Tell ICU where our resource data is located in memory. The data lives in the XercesMessages dll, and we just
@@ -172,11 +169,7 @@
     {
         setAppDataDone = true;
         UErrorCode err = U_ZERO_ERROR;
-#if defined(_WIN32) || defined(WIN32)
-        udata_setAppData("XercesMessages", &XercesMessages2_6_0_dat, &err);
-#else
         udata_setAppData("XercesMessages", &XercesMessages_dat, &err);
-#endif        
         if (U_SUCCESS(err))
         {
     	    setAppDataOK = true;
