diff -urN xerces-c-src_2_6_0.orig/src/xercesc/util/Compilers/VCPPDefs.hpp  xerces-c-src_2_6_0/src/xercesc/util/Compilers/VCPPDefs.hpp
--- xerces-c-src_2_6_0.orig/src/xercesc/util/Compilers/VCPPDefs.hpp	2004-09-08 09:56:32.000000000 -0400
+++ xerces-c-src_2_6_0/src/xercesc/util/Compilers/VCPPDefs.hpp	2005-04-05 14:24:15.159891000 -0400
@@ -112,8 +112,13 @@
 //  These defines provide the platform specific keywords that they need
 //  to do this.
 // ---------------------------------------------------------------------------
+#if defined(XERCES_DLL)
 #define PLATFORM_EXPORT     __declspec(dllexport)
 #define PLATFORM_IMPORT     __declspec(dllimport)
+#else
+#define PLATFORM_EXPORT
+#define PLATFORM_IMPORT
+#endif
 
 // ---------------------------------------------------------------------------
 //  Indicate that we do not support native bools
