diff -urN xml-xalan.orig/c/src/xalanc/Include/VCPPDefinitions.hpp xml-xalan/c/src/xalanc/Include/VCPPDefinitions.hpp
--- xml-xalan.orig/c/src/xalanc/Include/VCPPDefinitions.hpp	2004-04-28 04:33:05.000000000 -0400
+++ xml-xalan/c/src/xalanc/Include/VCPPDefinitions.hpp	2006-06-19 14:51:32.882371000 -0400
@@ -26,8 +26,13 @@
 //  These defines provide the platform specific keywords that they need
 //  to do this.
 // ---------------------------------------------------------------------------
+#if defined(XALAN_DLL)
 #define XALAN_PLATFORM_EXPORT     __declspec(dllexport)
 #define XALAN_PLATFORM_IMPORT     __declspec(dllimport)
+#else
+#define XALAN_PLATFORM_EXPORT
+#define XALAN_PLATFORM_IMPORT
+#endif
 #define XALAN_PLATFORM_EXPORT_FUNCTION(T) XALAN_PLATFORM_EXPORT T
 #define XALAN_PLATFORM_IMPORT_FUNCTION(T) XALAN_PLATFORM_IMPORT T
 
@@ -48,4 +53,8 @@
 
 #define XALAN_UNALIGNED
 
+#if _MSC_VER >= 1300
+#define XALAN_USE_NATIVE_WCHAR_T
+#endif
+
 #endif	// VCPPDEFINITIONS_HEADER_GUARD_1357924680

diff -urN xml-xalan.orig/c/src/xalanc/XPathCAPI/XPathCAPI.h xml-xalan/c/src/xalanc/XPathCAPI/XPathCAPI.h
--- xml-xalan.orig/c/src/xalanc/XPathCAPI/XPathCAPI.h	2004-04-28 04:33:14.000000000 -0400
+++ xml-xalan/c/src/xalanc/XPathCAPI/XPathCAPI.h	2005-04-06 09:23:27.578659900 -0400
@@ -20,6 +20,7 @@
 
 #if defined(_MSC_VER)
 
+#if defined(XALAN_DLL)
 #if defined(XALAN_XPATHCAPI_BUILD_DLL)
 
 #define XALAN_XPATHCAPI_EXPORT __declspec(dllexport)
@@ -31,6 +32,10 @@
 #define XALAN_XPATHCAPI_EXPORT_FUNCTION(T) T __declspec(dllimport)
 
 #endif
+#else
+#define XALAN_XPATHCAPI_EXPORT 
+#define XALAN_XPATHCAPI_EXPORT_FUNCTION(T) T 
+#endif
 
 #else
 
