diff -urN fltk-1.1.5.orig/FL/win32.H fltk-1.1.5/FL/win32.H
--- fltk-1.1.5.orig/FL/win32.H	2004-04-11 00:38:55.000000000 -0400
+++ fltk-1.1.5/FL/win32.H	2004-11-15 13:39:55.923041600 -0500
@@ -88,7 +88,7 @@
 };
 extern FL_EXPORT HCURSOR fl_default_cursor;
 extern FL_EXPORT UINT fl_wake_msg;
-inline Window fl_xid(const Fl_Window*w) {return Fl_X::i(w)->xid;}
+inline Window fl_xid(const Fl_Window*w) {return Fl_X::i(w) ? Fl_X::i(w)->xid : NULL;}
 FL_EXPORT Fl_Window* fl_find(Window xid);
 extern FL_EXPORT char fl_override_redirect; // hack into Fl_Window::make_xid()
 extern FL_EXPORT int fl_background_pixel;  // hack into Fl_Window::make_xid()
