diff -urN fltk-1.1.6.orig/FL/Fl.H fltk-1.1.6/FL/Fl.H
--- fltk-1.1.6.orig/FL/Fl.H	2004-11-23 14:47:50.000000000 -0500
+++ fltk-1.1.6/FL/Fl.H	2004-11-24 12:08:36.076880200 -0500
@@ -261,8 +261,23 @@
   // Widget deletion:
   static void delete_widget(Fl_Widget *w);
   static void do_widget_deletion();
+
+  // patch for focus handling
+public:
+    typedef void (TFocusCallback_F)(Fl_Widget* new_focus, void* data);
+
+    static void         set_focus_callback(TFocusCallback_F* callback, void* data);
+    TFocusCallback_F*   get_focus_callback() const;
+private:
+    static TFocusCallback_F*   focus_callback_;
+    static void*    focus_callback_data_;
+  // end of patch for focus handling
 };
 
+// patch for focus handling
+#define FLTK_SET_FOCUS_PATCH
+// end of patch for focus handling
+
 #endif // !Fl_H
 
 //
diff -urN fltk-1.1.6.orig/src/Fl.cxx fltk-1.1.6/src/Fl.cxx
--- fltk-1.1.6.orig/src/Fl.cxx	2004-11-23 14:50:58.000000000 -0500
+++ fltk-1.1.6/src/Fl.cxx	2004-11-24 12:06:11.731733900 -0500
@@ -437,12 +437,35 @@
 
 ////////////////////////////////////////////////////////////////
 
+// patch for focus handling
+Fl::TFocusCallback_F*   Fl::focus_callback_ = NULL;
+void*   Fl::focus_callback_data_ = NULL;
+
+void  Fl::set_focus_callback(TFocusCallback_F* callback, void* data)
+{
+    focus_callback_ = callback;
+    focus_callback_data_ = data; 
+}
+
+Fl::TFocusCallback_F*    Fl::get_focus_callback() const
+{
+    return focus_callback_;
+}
+// end of patch
+
 Fl_Widget* fl_oldfocus; // kludge for Fl_Group...
 
 void Fl::focus(Fl_Widget *o) {
   if (o && !o->visible_focus()) return;
   if (grab()) return; // don't do anything while grab is on
   Fl_Widget *p = focus_;
+
+  // patch for focus handling
+  if(focus_callback_)  {
+    (*focus_callback_)(o, focus_callback_data_);
+  }
+  // end of patch
+
   if (o != p) {
     Fl::compose_reset();
     focus_ = o;
