diff -urN fltk-1.1.6.orig/FL/Fl_Group.H fltk-1.1.6/FL/Fl_Group.H
--- fltk-1.1.6.orig/FL/Fl_Group.H	2004-04-11 00:38:54.000000000 -0400
+++ fltk-1.1.6/FL/Fl_Group.H	2004-12-22 09:21:51.608526500 -0500
@@ -41,6 +41,10 @@
   int navigation(int);
   static Fl_Group *current_;
 
+  // unimplemented copy ctor and assignment operator
+  Fl_Group(const Fl_Group&);
+  Fl_Group& operator=(const Fl_Group&);
+
 protected:
 
   void draw();
diff -urN fltk-1.1.6.orig/FL/Fl_Widget.H fltk-1.1.6/FL/Fl_Widget.H
--- fltk-1.1.6.orig/FL/Fl_Widget.H	2004-11-23 14:47:50.000000000 -0500
+++ fltk-1.1.6/FL/Fl_Widget.H	2004-12-22 09:21:51.561651500 -0500
@@ -69,13 +69,9 @@
 
   const char *tooltip_;
 
-#  if !defined(WIN32) || !defined(FL_DLL)
-  // "de-implement" the copy constructors, EXCEPT for when we are using the
-  // WIN32 DLL interface, in which case we can't hide them because Microsoft
-  // requires the copy constructors to implement subclassing...
-  Fl_Widget & operator=(const Fl_Widget &) { return *this; }
-  Fl_Widget(const Fl_Widget &) {}
-#  endif // !WIN32 || !FL_DLL
+  // unimplemented copy ctor and assignment operator
+  Fl_Widget(const Fl_Widget &);
+  Fl_Widget& operator=(const Fl_Widget &);
 
 protected:
 
diff -urN fltk-1.1.6.orig/FL/Fl_Window.H fltk-1.1.6/FL/Fl_Window.H
--- fltk-1.1.6.orig/FL/Fl_Window.H	2004-11-19 22:19:57.000000000 -0500
+++ fltk-1.1.6/FL/Fl_Window.H	2004-12-22 09:21:51.671026500 -0500
@@ -58,6 +58,10 @@
   };
   void _Fl_Window(); // constructor innards
 
+  // unimplemented copy ctor and assignment operator
+  Fl_Window(const Fl_Window&);
+  Fl_Window& operator=(const Fl_Window&);
+
 protected:
 
   static Fl_Window *current_;
