diff -urN wxMSW-2.8.6.orig/include/wx/msw/setup.h wxMSW-2.8.6/include/wx/msw/setup.h
--- wxMSW-2.8.6.orig/include/wx/msw/setup.h	2007-09-25 08:26:44.000000000 -0400
+++ wxMSW-2.8.6/include/wx/msw/setup.h	2007-10-02 12:10:52.380214800 -0400
@@ -244,7 +244,7 @@
 // Default is 1
 //
 // Recommended setting: 0 unless you do plan to develop MT applications
-#define wxUSE_THREADS 1
+#define wxUSE_THREADS 0
 
 // If enabled, compiles wxWidgets streams classes
 //
@@ -416,7 +416,7 @@
 // Default is 1.
 //
 // Recommended setting: 1
-#define wxUSE_DIALUP_MANAGER   1
+#define wxUSE_DIALUP_MANAGER   0
 
 // Compile in classes for run-time DLL loading and function calling.
 // Required by wxUSE_DIALUP_MANAGER.
@@ -444,7 +444,7 @@
 #define wxUSE_FS_ARCHIVE    1
 
 // Set to 1 to enable virtual Internet filesystem (requires wxUSE_FILESYSTEM)
-#define wxUSE_FS_INET       1
+#define wxUSE_FS_INET       0
 
 // wxArchive classes for accessing archives such as zip and tar
 #define wxUSE_ARCHIVE_STREAMS     1
@@ -991,7 +991,7 @@
 // Default is 0.
 //
 // Recommended setting: 1 if you intend to use OpenGL, 0 otherwise
-#define wxUSE_GLCANVAS       0
+#define wxUSE_GLCANVAS       1
 
 // wxRichTextCtrl allows editing of styled text.
 //
@@ -1134,7 +1134,7 @@
 // Default is 1.
 //
 // Recommended setting: whatever your compiler likes more
-#define wxUSE_IOSTREAMH     1
+#define wxUSE_IOSTREAMH     0
 
 // ----------------------------------------------------------------------------
 // image format support

diff -urN wxMSW-2.8.6.orig/include/msvc/wx/setup.h wxMSW-2.8.6/include/msvc/wx/setup.h
--- wxMSW-2.8.6.orig/include/msvc/wx/setup.h	2007-09-25 08:48:44.000000000 -0400
+++ wxMSW-2.8.6/include/msvc/wx/setup.h	2007-10-02 12:12:38.630894800 -0400
@@ -14,236 +14,7 @@
 // explicitly!
 
 #ifdef _MSC_VER
-    #ifdef _UNICODE
-        #ifdef WXUSINGDLL
-            #ifdef _DEBUG
-                #include "../../../lib/vc_dll/mswud/wx/setup.h"
-            #else
-                #include "../../../lib/vc_dll/mswu/wx/setup.h"
-            #endif
-        #else
-            #ifdef _DEBUG
-                #include "../../../lib/vc_lib/mswud/wx/setup.h"
-            #else
-                #include "../../../lib/vc_lib/mswu/wx/setup.h"
-            #endif
-        #endif
-
-        #ifdef _DEBUG
-            #pragma comment(lib,"wxbase28ud")
-            #pragma comment(lib,"wxbase28ud_net")
-            #pragma comment(lib,"wxbase28ud_xml")
-            #if wxUSE_REGEX
-                #pragma comment(lib,"wxregexud")
-            #endif
-
-            #if wxUSE_GUI
-                #if wxUSE_XML
-                    #pragma comment(lib,"wxexpatd")
-                #endif
-                #if wxUSE_LIBJPEG
-                    #pragma comment(lib,"wxjpegd")
-                #endif
-                #if wxUSE_LIBPNG
-                    #pragma comment(lib,"wxpngd")
-                #endif
-                #if wxUSE_LIBTIFF
-                    #pragma comment(lib,"wxtiffd")
-                #endif
-                #if wxUSE_ZLIB
-                    #pragma comment(lib,"wxzlibd")
-                #endif
-                #pragma comment(lib,"wxmsw28ud_adv")
-                #pragma comment(lib,"wxmsw28ud_core")
-                #pragma comment(lib,"wxmsw28ud_html")
-                #if wxUSE_GLCANVAS
-                    #pragma comment(lib,"wxmsw28ud_gl")
-                #endif
-                #if wxUSE_DEBUGREPORT
-                    #pragma comment(lib,"wxmsw28ud_qa")
-                #endif
-                #if wxUSE_XRC
-                    #pragma comment(lib,"wxmsw28ud_xrc")
-                #endif
-                #if wxUSE_AUI
-                    #pragma comment(lib,"wxmsw28ud_aui")
-                #endif
-                #if wxUSE_RICHTEXT
-                    #pragma comment(lib,"wxmsw28ud_richtext")
-                #endif
-                #if wxUSE_MEDIACTRL
-                    #pragma comment(lib,"wxmsw28ud_media")
-                #endif
-                #if wxUSE_ODBC
-                    #pragma comment(lib,"wxbase28ud_odbc")
-                #endif
-            #endif // wxUSE_GUI
-        #else // release
-            #pragma comment(lib,"wxbase28u")
-            #pragma comment(lib,"wxbase28u_net")
-            #pragma comment(lib,"wxbase28u_xml")
-            #if wxUSE_REGEX
-                #pragma comment(lib,"wxregexu")
-            #endif
-
-            #if wxUSE_GUI
-                #if wxUSE_XML
-                    #pragma comment(lib,"wxexpat")
-                #endif
-                #if wxUSE_LIBJPEG
-                    #pragma comment(lib,"wxjpeg")
-                #endif
-                #if wxUSE_LIBPNG
-                    #pragma comment(lib,"wxpng")
-                #endif
-                #if wxUSE_LIBTIFF
-                    #pragma comment(lib,"wxtiff")
-                #endif
-                #if wxUSE_ZLIB
-                    #pragma comment(lib,"wxzlib")
-                #endif
-                #pragma comment(lib,"wxmsw28u_adv")
-                #pragma comment(lib,"wxmsw28u_core")
-                #pragma comment(lib,"wxmsw28u_html")
-                #if wxUSE_GLCANVAS
-                    #pragma comment(lib,"wxmsw28u_gl")
-                #endif
-                #if wxUSE_DEBUGREPORT
-                    #pragma comment(lib,"wxmsw28u_qa")
-                #endif
-                #if wxUSE_XRC
-                    #pragma comment(lib,"wxmsw28u_xrc")
-                #endif
-                #if wxUSE_AUI
-                    #pragma comment(lib,"wxmsw28u_aui")
-                #endif
-                #if wxUSE_RICHTEXT
-                    #pragma comment(lib,"wxmsw28u_richtext")
-                #endif
-                #if wxUSE_MEDIACTRL
-                    #pragma comment(lib,"wxmsw28u_media")
-                #endif
-                #if wxUSE_ODBC
-                    #pragma comment(lib,"wxbase28u_odbc")
-                #endif
-            #endif // wxUSE_GUI
-        #endif // debug/release
-    #else // !_UNICODE
-        #ifdef WXUSINGDLL
-            #ifdef _DEBUG
-                #include "../../../lib/vc_dll/mswd/wx/setup.h"
-            #else
-                #include "../../../lib/vc_dll/msw/wx/setup.h"
-            #endif
-        #else // static lib
-            #ifdef _DEBUG
-                #include "../../../lib/vc_lib/mswd/wx/setup.h"
-            #else
-                #include "../../../lib/vc_lib/msw/wx/setup.h"
-            #endif
-        #endif // shared/static
-
-        #ifdef _DEBUG
-            #pragma comment(lib,"wxbase28d")
-            #pragma comment(lib,"wxbase28d_net")
-            #pragma comment(lib,"wxbase28d_xml")
-            #if wxUSE_REGEX
-                #pragma comment(lib,"wxregexd")
-            #endif
-
-            #if wxUSE_GUI
-                #if wxUSE_XML
-                    #pragma comment(lib,"wxexpatd")
-                #endif
-                #if wxUSE_LIBJPEG
-                    #pragma comment(lib,"wxjpegd")
-                #endif
-                #if wxUSE_LIBPNG
-                    #pragma comment(lib,"wxpngd")
-                #endif
-                #if wxUSE_LIBTIFF
-                    #pragma comment(lib,"wxtiffd")
-                #endif
-                #if wxUSE_ZLIB
-                    #pragma comment(lib,"wxzlibd")
-                #endif
-                #pragma comment(lib,"wxmsw28d_adv")
-                #pragma comment(lib,"wxmsw28d_core")
-                #pragma comment(lib,"wxmsw28d_html")
-                #if wxUSE_GLCANVAS
-                    #pragma comment(lib,"wxmsw28d_gl")
-                #endif
-                #if wxUSE_DEBUGREPORT
-                    #pragma comment(lib,"wxmsw28d_qa")
-                #endif
-                #if wxUSE_XRC
-                    #pragma comment(lib,"wxmsw28d_xrc")
-                #endif
-                #if wxUSE_AUI
-                    #pragma comment(lib,"wxmsw28d_aui")
-                #endif
-                #if wxUSE_RICHTEXT
-                    #pragma comment(lib,"wxmsw28d_richtext")
-                #endif
-                #if wxUSE_MEDIACTRL
-                    #pragma comment(lib,"wxmsw28d_media")
-                #endif
-                #if wxUSE_ODBC
-                    #pragma comment(lib,"wxbase28d_odbc")
-                #endif
-            #endif // wxUSE_GUI
-        #else // release
-            #pragma comment(lib,"wxbase28")
-            #pragma comment(lib,"wxbase28_net")
-            #pragma comment(lib,"wxbase28_xml")
-            #if wxUSE_REGEX
-                #pragma comment(lib,"wxregex")
-            #endif
-
-            #if wxUSE_GUI
-                #if wxUSE_XML
-                    #pragma comment(lib,"wxexpat")
-                #endif
-                #if wxUSE_LIBJPEG
-                    #pragma comment(lib,"wxjpeg")
-                #endif
-                #if wxUSE_LIBPNG
-                    #pragma comment(lib,"wxpng")
-                #endif
-                #if wxUSE_LIBTIFF
-                    #pragma comment(lib,"wxtiff")
-                #endif
-                #if wxUSE_ZLIB
-                    #pragma comment(lib,"wxzlib")
-                #endif
-                #pragma comment(lib,"wxmsw28_adv")
-                #pragma comment(lib,"wxmsw28_core")
-                #pragma comment(lib,"wxmsw28_html")
-                #if wxUSE_GLCANVAS
-                    #pragma comment(lib,"wxmsw28_gl")
-                #endif
-                #if wxUSE_DEBUGREPORT
-                    #pragma comment(lib,"wxmsw28_qa")
-                #endif
-                #if wxUSE_XRC
-                    #pragma comment(lib,"wxmsw28_xrc")
-                #endif
-                #if wxUSE_AUI
-                    #pragma comment(lib,"wxmsw28_aui")
-                #endif
-                #if wxUSE_RICHTEXT
-                    #pragma comment(lib,"wxmsw28_richtext")
-                #endif
-                #if wxUSE_MEDIACTRL
-                    #pragma comment(lib,"wxmsw28_media")
-                #endif
-                #if wxUSE_ODBC
-                    #pragma comment(lib,"wxbase28_odbc")
-                #endif
-            #endif // wxUSE_GUI
-        #endif // debug/release
-    #endif // _UNICODE/!_UNICODE
+    #include "../../wx/msw/setup.h"
 #else
     #error "This file should only be included when using Microsoft Visual C++"
 #endif
-
