diff -urN icu.orig/source/common/unicode/pwin32.h icu/source/common/unicode/pwin32.h 
--- icu.orig/source/common/unicode/pwin32.h	2004-11-01 19:27:44.000000000 -0500
+++ icu/source/common/unicode/pwin32.h	2005-04-18 16:01:58.456785800 -0400
@@ -227,11 +227,12 @@
 
 #ifdef U_STATIC_IMPLEMENTATION
 #define U_EXPORT
+#define U_IMPORT
 #else
 #define U_EXPORT __declspec(dllexport)
+#define U_IMPORT __declspec(dllimport)
 #endif
 #define U_EXPORT2 __cdecl
-#define U_IMPORT __declspec(dllimport)
 
 /*===========================================================================*/
 /* Code alignment and C function inlining                                    */
