<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>BLAST API - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="ch_blast" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><script type="text/javascript" src="js/jig/1.7/js/jig.min.js"></script><script type="text/javascript" src="js/toolkitSymbolSearch.js"></script><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><a href="#__nav_cntnt" class="jig-ncbipopper">Contents ▾</a><div style="display:none;width:700px;height:500px;overflow:auto;" id="__nav_cntnt"><a class="right" href="toc.html">Table of Contents Page</a><ul class="simple-list"><li class="half_rhythm"><a href="toolkit.fm.html">Book Information</a></li><li class="half_rhythm"><a href="part1.html">Part 1. Overview</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_intro.html">1. Introduction to the C++ Toolkit</a></li><li class="half_rhythm"><a href="ch_start.html">2. Getting Started</a></li></ul></li><li class="half_rhythm"><a href="part2.html">Part 2. Development Framework</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_getcode_svn.html">3. Retrieve the Source Code (FTP and Subversion)</a></li><li class="half_rhythm"><a href="ch_config.html">4. Configure, Build, and Use the Toolkit</a></li><li class="half_rhythm"><a href="ch_build.html">5. Working with Makefiles</a></li><li class="half_rhythm"><a href="ch_proj.html">6. Project Creation and Management</a></li><li class="half_rhythm"><a href="ch_style.html">7. Programming Policies and Guidelines</a></li></ul></li><li class="half_rhythm"><a href="part3.html">Part 3. C++ Toolkit Library Reference</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_core.html">8. Portability, Core Functionality and Application Framework</a></li><li class="half_rhythm"><a href="ch_conn.html">9. Networking and IPC</a></li><li class="half_rhythm"><a href="ch_dbapi.html">10. Database Access - SQL, Berkley DB</a></li><li class="half_rhythm"><a href="ch_cgi.html">11. CGI and Fast-CGI </a></li><li class="half_rhythm"><a href="ch_html.html">12. HTML</a></li><li class="half_rhythm"><a href="ch_ser.html">13. Data Serialization (ASN.1, XML)</a></li><li class="half_rhythm"><a href="ch_datamod.html">14. Biological Sequence Data Model</a></li><li class="half_rhythm"><a href="ch_objmgr.html">15. Biological Object Manager</a></li><li class="half_rhythm"><a href="ch_blast.html">16. BLAST API</a></li><li class="half_rhythm"><a href="ch_dataaccess.html">17. Access to NCBI data</a></li><li class="half_rhythm"><a href="ch_algoalign.html">18. Biological Sequence Alignment</a></li><li class="half_rhythm"><a href="ch_gui.html">19. GUI and Graphics</a></li><li class="half_rhythm"><a href="ch_boost.html">20. Using the Boost Unit Test Framework</a></li></ul></li><li class="half_rhythm"><a href="part4.html">Part 4. Wrappers for 3rd-Party Packages</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_xmlwrapp.html">21. XmlWrapp (XML parsing and handling, XSLT, XPath)</a></li></ul></li><li class="half_rhythm"><a href="part5.html">Part 5. Software</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_debug.html">22. Debugging, Exceptions, and Error Handling</a></li><li class="half_rhythm"><a href="ch_grid.html">23. Distributed Computing</a></li><li class="half_rhythm"><a href="ch_app.html">24. Applications</a></li><li class="half_rhythm"><a href="ch_demo.html">25. Examples and Demos</a></li><li class="half_rhythm"><a href="ch_res.html">26. C Toolkit Resources for C++ Toolkit Users</a></li></ul></li><li class="half_rhythm"><a href="part6.html">Part 6. Help and Support</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_browse.html">27. NCBI C++ Toolkit Source Browser</a></li><li class="half_rhythm"><a href="ch_devtools.html">28. Software Development Tools</a></li><li class="half_rhythm"><a href="ch_xmlauthor.html">29. XML Authoring using Word</a></li><li class="half_rhythm"><a href="ch_faq.html">30. FAQs, Useful Documentation Links, and Mailing Lists</a></li></ul></li><li class="half_rhythm"><a href="part7.html">Part 7. Library and Applications Configuration</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_libconfig.html">31. Library Configuration</a></li></ul></li><li class="half_rhythm"><a href="part8.html">Release Notes</a><ul class="simple-list"><li class="half_rhythm"><a href="release_notes.html">Release Notes (Version 9, May 2012)</a></li><li class="half_rhythm"><a href="release_notes_7-05_2011.html">Release Notes (Version 7, May 2011)</a></li><li class="half_rhythm"><a href="release_notes_06_29_2010.html">Release Notes (June, 2010)</a></li><li class="half_rhythm"><a href="release_notes_05_15_2009.html">Release Notes (May, 2009)</a></li><li class="half_rhythm"><a href="release_notes_12_24_2008.html">Release Notes (December, 2008)</a></li><li class="half_rhythm"><a href="release_notes_12_31_2008.html">Release Notes (March, 2008)</a></li><li class="half_rhythm"><a href="release_notes_08_27_2007.html">Release Notes (August, 2007)</a></li><li class="half_rhythm"><a href="release_notes_03_12_2007.html">Release Notes (March, 2007)</a></li><li class="half_rhythm"><a href="release_notes_08_14_2006.html">Release Notes (August, 2006)</a></li><li class="half_rhythm"><a href="release_notes_04_30_2006.html">Release Notes (April 30, 2006)</a></li><li class="half_rhythm"><a href="release_notes_12_31_2005.html">Release Notes (December 31, 2005)</a></li><li class="half_rhythm"><a href="release_notes_10_03_2005.html">Release Notes (August, 2005)</a></li><li class="half_rhythm"><a href="release_notes_05_05_2005.html">Release Notes (April, 2005)</a></li><li class="half_rhythm"><a href="release_notes_03_09_2005.html">Release Notes (February, 2005)</a></li><li class="half_rhythm"><a href="release_notes_11_22_2004.html">Release Notes (November 22, 2004)</a></li><li class="half_rhythm"><a href="release_notes_10_2_2004.html">Release Notes (October 2, 2004)</a></li><li class="half_rhythm"><a href="release_notes_7_8_2004.html">Release Notes (July 8, 2004)</a></li><li class="half_rhythm"><a href="release_notes_april_16_2004.html">Release Notes (April 16, 2004)</a></li><li class="half_rhythm"><a href="release_notes_12_08_2003.html">Release Notes (December 8, 2003)</a></li><li class="half_rhythm"><a href="release_notes_08_01_2003.html">Release Notes (August 1, 2003)</a></li></ul></li><li class="half_rhythm"><a href="app1.appendix1.html">Appendix - Books and Styles</a></li></ul></div><div class="pagination"><a class="pagelink active prev" title="Previous page" href="ch_objmgr.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="ch_dataaccess.html">Next &gt;</a></div></div><div class="meta_content"><h1><span class="label">16</span><span class="title">BLAST API</span></h1><p class="small">Created: August 22, 2006; Last Update: April 13, 2010.</p></div><div class="body_content"><div class="sec"><h2><span class="title">Overview</span></h2><p>The overview for this chapter consists of the following topics:</p><ul><li><p><a href="#ch_blast.intro">Introduction</a></p></li><li><p><a href="#ch_blast.outline">Chapter Outline</a></p></li></ul><div class="sec"><h3><span class="title">Introduction</span></h3><p>BLAST (Basic Local Alignment Search Tool) is used to perform sequence similarity searches. Most often this means that BLAST is used to search a sequence (either DNA or protein) against a database of other sequences (either all nucleotide or all protein) in order to identify similar sequences. BLAST has many different flavors and can not only search DNA against DNA or protein against protein but also can translate a nucleotide query and search it against a protein database as well as the other way around. It can also compute a “profile” for the query sequence and use that for further searches as well as search the query against a database of profiles. BLAST is available as a web service at the NCBI, as a stand-alone binary, and is built into other tools. It is an extremely versatile program and probably the most heavily used similarity search program in the world. BLAST runs on a multitude of different platforms that include Windows, MacOS, LINUX, and many flavors of UNIX. It is also under continuing development with new algorithmic innovations. Multiple references to BLAST can be found at <a href="http://www.ncbi.nlm.nih.gov/BLAST/blast_references.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blast_references.shtml</a>.</p><p>The version of BLAST in the NCBI C++ Toolkit was rewritten from scratch based upon the version in the C Toolkit that was originally introduced in 1997. A decision was made to break the code for the new version of BLAST into two different categories. There is the “core” code of BLAST that is written in vanilla C and does not use any part of the NCBI C or C++ Toolkits. There is also the “API” code that is written in C++ and takes full advantage of the tools provided by the NCBI C++ Toolkit. The reason to write the core part of the code in vanilla C was so that the same code could be used in the C Toolkit (to replace the 1997 version) as well as to make it possible for researchers interested in algorithmic development to work with the core of BLAST independently of any Toolkit. Even though the core part was written without the benefit of the C++ or C Toolkits an effort was made to conform to the <a href="ch_style.html">Programming Policies and Guidelines</a> chapter of this book. Doxygen-style comments are used to allow API documentation to be automatically generated (see the BLAST Doxygen link at <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/group__AlgoBlast.html">http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/group__AlgoBlast.html</a>). Both the core and API parts of BLAST can be found under <span class="nctnt ncbi-path">algo/blast</span> in the C++ Toolkit.</p><p>An attempt was made to isolate the user of the BLAST API (as exposed in <span class="nctnt ncbi-path">algo/blast/api</span>) from the core of BLAST, so that algorithmic enhancements or refactoring of that code would be transparent to the API programmer as far as that is possible. Since BLAST is continually under development and many of the developments involve new features it is not always possible or desirable to isolate the API programmer from these changes. This chapter will focus on the API for the C++ Toolkit. A few different search classes will be discussed. These include the <span class="nctnt ncbi-class">CLocalBlast</span> class, typically used for searching a query (or queries) against a BLAST database; <span class="nctnt ncbi-class">CRemoteBlast</span>, used for sending searches to the NCBI servers; as well as <span class="nctnt ncbi-class">CBl2Seq</span>, useful for searching target sequences that have not been formatted as a BLAST database.</p></div><div class="sec"><h3><span class="title">Chapter Outline</span></h3><p><a href="ch_blast.html#ch_blast.CLocalBlast">CLocalBlast</a></p><ul><li><p><a href="ch_blast.html#ch_blast._Query_Sequence">Query Sequence</a></p></li><li><p><a href="ch_blast.html#ch_blast._Options">Options</a></p></li><li><p><a href="ch_blast.html#ch_blast._Target_Sequences">Target Sequences</a></p></li><li><p><a href="ch_blast.html#ch_blast._Results">Results</a></p></li></ul><p><a href="ch_blast.html#ch_blast.CRemoteBlast">CRemoteBlast</a></p><ul><li><p><a href="ch_blast.html#ch_blast._Query_Sequence_1">Query Sequence</a></p></li><li><p><a href="ch_blast.html#ch_blast._Options_1">Options</a></p></li><li><p><a href="ch_blast.html#ch_blast._Target_Sequences_1">Target Sequences</a></p></li><li><p><a href="ch_blast.html#ch_blast._Results_1">Results</a></p></li></ul><p><a href="ch_blast.html#ch_blast.The_Uniform_Interfac">The Uniform Interface</a></p><p><a href="ch_blast.html#ch_blast.CBl2Seq">CBl2Seq</a></p><ul><li><p><a href="ch_blast.html#ch_blast._Query_Sequence_2">Query Sequence</a></p></li><li><p><a href="ch_blast.html#ch_blast.Options_and_Program_">Options and Program Type</a></p></li><li><p><a href="ch_blast.html#ch_blast._Target_Sequences_2">Target Sequences</a></p></li><li><p><a href="ch_blast.html#ch_blast._Results_2">Results</a></p></li></ul><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/blast_opts_cookbook.html">C++ BLAST Options Cookbook</a></p><p><a href="ch_blast.html#ch_blast.Sample_Applications">Sample Applications</a></p></div></div><div class="sec" id="ch_blast.CLocalBlast"><h2><span class="title">CLocalBlast</span></h2><p>The class <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CLocalBlast&amp;d=C">CLocalBlast</a> can be used for searches that run locally on a machine (as opposed to sending the request over the network to use the CPU of another machine) and search a query (or queries) against a preformatted BLAST database, which holds the target sequence data in a format optimal for BLAST searches. The demonstration program <span class="nctnt ncbi-path">blast_demo.cpp</span> illustrates the use of <span class="nctnt ncbi-class">CLocalBlast</span>. There are a few different <span class="nctnt ncbi-class">CLocalBlast</span> constructors, but they always take three arguments reflecting the need for a query sequence, a set of BLAST options, and a set of target sequences (e.g., BLAST database). First we discuss how to construct these arguments and then we discuss how to access the results.</p><div class="sec" id="ch_blast._Query_Sequence"><h3><span class="title">Query Sequence</span></h3><p>The classes that perform BLAST searches expect to be given query sequences in one of a few formats. Each is a container for one or more query sequences expressed as <span class="nctnt ncbi-class">CSeq_loc</span> objects, along with ancillary information. In this document we will only discuss classes that take either a <span class="nctnt ncbi-type">SSeqLoc</span> or a <span class="nctnt ncbi-class">TSeqLocVector</span>, which is just a collection of <span class="nctnt ncbi-type">SSeqLoc</span>’s.</p><p><span class="nctnt ncbi-class">CBlastInput</span> is a class that converts an abstract source of sequence data into a format suitable for use by the BLAST search classes. This class may produce either a <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=TSeqLocVector&amp;d=T">TSeqLocVector</a> container or a <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CBlastQueryVector&amp;d=C">CBlastQueryVector</a> container to represent query sequences. As mentioned above we limit our discussion to the <span class="nctnt ncbi-class">TSeqLocVector</span> class here.</p><p><span class="nctnt ncbi-class">CBlastInput</span> can produce a single container that includes all the query sequences, or can output a batch of sequences at a time (the combined length of the sequences within each batch can be specified) until all of the sequences within the data source have been consumed.</p><p>Sources of sequence data are represented by a <span class="nctnt ncbi-class">CBlastInputSource</span>, or a class derived from it. <span class="nctnt ncbi-class">CBlastInput</span> uses these classes to read one sequence at a time from the data source and convert to a container suitable for use by the BLAST search classes.</p><p>An example use of <span class="nctnt ncbi-class">CBlastInputSource</span> is <span class="nctnt ncbi-class">CBlastFastaInputSource</span>, which represents a stream containing fasta-formatted biological sequences. Usually this class represents a collection of sequences residing in a text file. One sequence at a time is read from the file and converted into a BLAST input container.</p><p><span class="nctnt ncbi-class">CBlastFastaInputSource</span> uses <span class="nctnt ncbi-class">CBlastInputConfig</span> to provide more control over the file reading process. For example, the read process can be limited to a range of each sequence, or sequence letters that appear in lowercase can be scheduled for masking by BLAST. <span class="nctnt ncbi-class">CBlastInputConfig</span> can be used by other classes to provide the same kind of control, although not all class members will be appropriate for every data source.</p></div><div class="sec" id="ch_blast._Options"><h3><span class="title">Options</span></h3><p>The BLAST options classes were designed to allow a programmer to easily set the options to values appropriate to common tasks, but then modify individual options as needed. <a href="ch_blast.html#ch_blast.T18.3">Table 1</a> lists the supported tasks.</p><div class="table" id="ch_blast.T18.3"><div class="caption"><p>Table 1: List of tasks supported by the CBlastOptionsHandle. “Translated nucleotide” means that the input was nucleotide, but the comparison is based upon the protein. PSSM is a “position-specific scoring matrix”. The “EProgram” can be used as an argument to CBlastOptionsFactory::Create</p></div><table><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1"><b>EProgram (enum)</b></td><td align="left" valign="top" rowspan="1" colspan="1"><b>Default Word-size</b></td><td align="left" valign="top" rowspan="1" colspan="1"><b>Query type</b></td><td align="left" valign="top" rowspan="1" colspan="1"><b>Target type</b></td><td align="left" valign="top" rowspan="1" colspan="1"><b>Notes</b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eBlastN</span></td><td align="left" valign="top" rowspan="1" colspan="1">11</td><td align="left" valign="top" rowspan="1" colspan="1">Nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eMegablast</span></td><td align="left" valign="top" rowspan="1" colspan="1">28</td><td align="left" valign="top" rowspan="1" colspan="1">Nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Optimized for speed and closely related sequences</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eDiscMegablast</span></td><td align="left" valign="top" rowspan="1" colspan="1">11</td><td align="left" valign="top" rowspan="1" colspan="1">Nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Optimized for cross-species matches</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eBlastp</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eBlastx</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Translated nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eTblastn</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1">Translated nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eTblastx</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Translated nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">Translated nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eRPSBlast</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1">PSSM</td><td align="left" valign="top" rowspan="1" colspan="1">Can very quickly identify domains</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">eRPSTblastn</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Translated nucleotide</td><td align="left" valign="top" rowspan="1" colspan="1">PSSM</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">ePSIBlast</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">PSSM</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1">Extremely sensitive method to find distant homologies</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-type">ePHIBlastp</span></td><td align="left" valign="top" rowspan="1" colspan="1">3</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1">Protein</td><td align="left" valign="top" rowspan="1" colspan="1">Uses pattern in query to start alignments</td></tr></tbody></table></div><p>The <span class="nctnt ncbi-class">CBlastOptionsFactory</span> class offers a single static method to create <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptionsHandle.html">CBlastOptionsHandle</a> subclasses so that options applicable to all variants of BLAST can be inspected or modified. The actual type of the <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptionsHandle.html">CBlastOptionsHandle</a> returned by the <span class="nctnt ncbi-func">Create()</span> method is determined by its <span class="nctnt ncbi-monospace">EProgram</span> argument (see <a href="ch_blast.html#ch_blast.T18.3">Table 1</a>). The return value of this function is guaranteed to have reasonable defaults set for the selected task.</p><p>The <span class="nctnt ncbi-class">CBlastOptionsHandle</span> class encapsulates options that are common to all variants of BLAST, from which more specific tasks can inherit the common options. The subclasses of <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptionsHandle.html">CBlastOptionsHandle</a> should present an interface that is more specific, i.e.: only contain options relevant to the task at hand, although it might not be an exhaustive interface for all options available for the task. Please note that the initialization of this class' data members follows the template method design pattern, and this should be followed by subclasses also. Below is an example use of the <span class="nctnt ncbi-class">CBlastOptionsHandle</span> to create a set of options appropriate to “blastn” and then to set the expect value to non-default values:</p><pre>using ncbi::blast;<br /><br /><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCRef.html">CRef&lt;CBlastOptionsHandle&gt;</a><br />    opts_handle(<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/group__AlgoBlast.html#g59eed1e74b8c89243ae">CBlastOptionsFactory::Create</a>(<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/blast__types_8hpp.html#e1c0ad647974cd781398">eBlastn</a>));<br />opts_handle-&gt;SetEvalueThreshold(1e-10);<br />blast(<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/struct__dp__demo_8cpp.html#I8bb76a03a06523f3046">query</a>, opts_handle, db);</pre><p>The <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptionsHandle.html">CBlastOptionsHandle</a> classes offers a <span class="nctnt ncbi-func">Validate()</span> method in its interface which is called by the BLAST search classes prior to performing the actual search, but users of the C++ BLAST options APIs might also want to invoke this method to ensure that any exceptions thrown by the BLAST search classes do not originate from an incorrect setting of BLAST options. Please note that the <span class="nctnt ncbi-func">Validate()</span> method throws a <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastException.html">CBlastException</a> in case of failure.</p><p>If the same type of search (e.g., nucleotide query vs. nucleotide database) will always be performed, then it may be preferable to create an instance of the derived classes of the <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptionsHandle.html">CBlastOptionsHandle</a>. These classes expose an interface that is relevant to the task at hand, but the popular options can be modified as necessary:</p><pre>using ncbi::blast;<br /><br />CRef&lt;CBlastNucleotideOptionsHandle&gt;    nucl_handle(new CBlastNucleotideOptionsHandle);<br />...<br />nucl_handle-&gt;<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/group__AlgoBlast.html#g6b3c0e438d4c6cbc400">SetTraditionalBlastnDefaults</a>();<br />nucl_handle-&gt;<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastNucleotideOptionsHandle.html#f8b6653245a785b49df5">SetStrandOption</a>(<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/group__dataspec__NCBI__Seqloc.html#gg2d4d48fde1d3f62563">objects::eNa_strand_plus</a>);<br />...<br />CRef&lt;CBlastOptionsHandle&gt;    opts = CRef&lt;CBlastOptionsHandle&gt; (&amp;*nucl_handle);<br />CLocalBlast blast(query_factory, opts, db);</pre><p>The <span class="nctnt ncbi-class">CBlastOptionsHandle</span> design arranges the BLAST options in a hierarchy. For example all searches that involve protein-protein comparisons (including proteins translated from a nucleotide sequence) are handled by <span class="nctnt ncbi-class">CBlastProteinOptionsHandle</span> or a subclass (e.g., <span class="nctnt ncbi-class">CBlastxOptionsHandle</span>). A limitation of this design is that the introduction of new algorithms or new options that only apply to some programs may violate the class hierarchy. To allow advanced users to overcome this limitation the <span class="nctnt ncbi-func">GetOptions()</span> and <span class="nctnt ncbi-func">SetOptions()</span> methods of the <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptionsHandle.html">CBlastOptionsHandle</a> hierarchy allow access to the <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/classCBlastOptions.html">CBlastOptions</a> class, the lowest level class in the C++ BLAST options API which contains all options available to all variants of the BLAST algorithm. No guarantees about the validity of the options are made if this interface is used, therefore invoking <span class="nctnt ncbi-func">Validate()</span> is <i>strongly</i> recommended.</p></div><div class="sec" id="ch_blast._Target_Sequences"><h3><span class="title">Target Sequences</span></h3><p>One may specify a BLAST database to search with the <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CSearchDatabase&amp;d=C">CSearchDatabase</a> class. Normally it is only necessary to provide a string for the database name and state whether it is a nucleotide or protein database. It is also possible to specify an entrez query or a vector of GI’s that will be used to limit the search.</p></div><div class="sec" id="ch_blast._Results"><h3><span class="title">Results</span></h3><p>The <span class="nctnt ncbi-func">Run()</span> method of <span class="nctnt ncbi-class">CLocalBlast</span> returns a <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CSearchResultSet&amp;d=C">CSearchResultSet</a> that may be used to obtain results of the search. The <span class="nctnt ncbi-class">CSearchResultSet</span> class is a random access container of <span class="nctnt ncbi-class">CSearchResults</span> objects, one for each query submitted in the search. The <span class="nctnt ncbi-class">CSearchResult</span> class provides access to alignment (as a <span class="nctnt ncbi-class">CSeq_align_set</span>), the query <span class="nctnt ncbi-var">Cseq_id</span>, warning or error messages that were generated during the run, as well as the filtered query regions (assuming query filtering was set).</p></div></div><div class="sec" id="ch_blast.CRemoteBlast"><h2><span class="title">CRemoteBlast</span></h2><p>The <span class="nctnt ncbi-class">CRemoteBlast</span> class sends a BLAST request to the SPLITD system at the NCBI. This can be advantageous in many situations. There is no need to download the (possibly) large BLAST databases to the user’s machine; the search may be spread across many machines by the SPLITD system at the NCBI, making it very fast; and the results will be kept on the NCBI server for 36 hours in case the users wishes to retrieve them again the next day. On the other hand the user must select one of the BLAST databases maintained by the NCBI since it is not possible to upload a custom database for searching. Here we discuss a <span class="nctnt ncbi-class">CRemoteBlast</span> constructor that takes three arguments, reflecting the need for a query sequence(s), a set of BLAST options, and a BLAST database. Readers are advised to read the <span class="nctnt ncbi-class">CLocalBlast</span> section before they read this section.</p><div class="sec" id="ch_blast._Query_Sequence_1"><h3><span class="title">Query Sequence</span></h3><p>A <span class="nctnt ncbi-class">TSeqLocVector</span> should be used as input to <span class="nctnt ncbi-class">CRemoteBlast</span>. Please see the section on <a href="ch_blast.html#ch_blast.CLocalBlast">CLocalBlast</a> for details.</p></div><div class="sec" id="ch_blast._Options_1"><h3><span class="title">Options</span></h3><p><span class="nctnt ncbi-func">CBlastOptionsFactory::Create()</span> can again be used to create options for <span class="nctnt ncbi-class">CRemoteBlast</span>. In this case though it is necessary to set the second (default) argument of <span class="nctnt ncbi-func">Create()</span> to <span class="nctnt ncbi-var">CBlastOptions::eRemote</span>.</p></div><div class="sec" id="ch_blast._Target_Sequences_1"><h3><span class="title">Target Sequences</span></h3><p>One may use the <span class="nctnt ncbi-class">CSearchDatabase</span> class to specify a BLAST database, similar to the method outlined in the <a href="ch_blast.html#ch_blast.CLocalBlast">CLocalBlast</a> section. In this case it is important to remember though that the user must select from the BLAST databases available on the NCBI Web site and not one built locally.</p></div><div class="sec" id="ch_blast._Results_1"><h3><span class="title">Results</span></h3><p>After construction of the <span class="nctnt ncbi-class">CRemoteBlast</span> object the user should call one of the <span class="nctnt ncbi-func">SubmitSync()</span> methods. After this returns the method <span class="nctnt ncbi-func">GetResultSet()</span> will return a <span class="nctnt ncbi-class">CSearchResultSet</span> which the user can interrogate using the same methods as in <span class="nctnt ncbi-class">CLocalBlast</span>. Additionally the user may obtain the request identifier (RID) issued by the SPLITD system with the method <span class="nctnt ncbi-func">GetRID()</span>.</p><p>Finally <span class="nctnt ncbi-class">CRemoteBlast</span> provides a constructor that takes a string, which it expects to be an RID issued by the SPLITD system. This RID might have been obtained by an earlier run of <span class="nctnt ncbi-class">CRemoteBlast</span> or it could be one that was obtained from the NCBI SPLITD system via the web page. Note that the SPLITD system will keep results on it’s server for 36 hours, so the RID cannot be older than that.</p></div></div><div class="sec" id="ch_blast.The_Uniform_Interfac"><h2><span class="title">The Uniform Interface</span></h2><p>The <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=ISeqSearch">ISeqSearch</a> class is an abstract interface class. Concrete subclasses can run either local (<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CLocalSeqSearch">CLocalSeqSearch</a>) or remote searches (<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CRemoteSeqSearch">CRemoteSeqSearch</a>). The concrete classes will only perform an intersection of the tasks that <span class="nctnt ncbi-class">CLocalBlast</span> and <span class="nctnt ncbi-class">CRemoteBlast</span> can perform. As an example, there is no method to retrieve a Request identifier (RID) from subclasses of <span class="nctnt ncbi-class">ISeqSearch</span> as this is supported only for remote searches but not for local searches. The methods supported by the concrete subclasses and the return values are similar to those of <span class="nctnt ncbi-class">CLocalBlast</span> and <span class="nctnt ncbi-class">CRemoteBlast</span>.</p></div><div class="sec" id="ch_blast.CBl2Seq"><h2><span class="title">CBl2Seq</span></h2><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CBl2Seq&amp;d=C">CBl2Seq</a> is a class useful for searching a query (or queries) against one or more target sequences that have not been formatted as a BLAST database. These sequences may, for example, come from a user who pasted them into a web page or be fetched from the Entrez or ID1 services at the NCBI. The <span class="nctnt ncbi-class">CBl2Seq</span> constructors all take three arguments, reflecting the need for a set of query sequences, a set of target sequences, and some information about the BLAST options or program type to use. In this section it is assumed the reader has already read the previous section on <span class="nctnt ncbi-class">CLocalBlast</span>.</p><p>The BLAST database holds the target sequence data in a format optimal for BLAST searches, so that if a target sequence is to be searched more than a few times it is best to convert it to a BLAST database and use <span class="nctnt ncbi-class">CLocalBlast</span>.</p><div class="sec" id="ch_blast._Query_Sequence_2"><h3><span class="title">Query Sequence</span></h3><p>The query sequence (or sequences) is represented either as a <span class="nctnt ncbi-type">SSeqLoc</span> (for a single query sequence) or as a <span class="nctnt ncbi-class">TSeqLocVector</span> (in the case of multiple query sequences). The <span class="nctnt ncbi-class">CBlastInput</span> class, described in the <a href="ch_blast.html#ch_blast.CLocalBlast">CLocalBlast</a> section, can be used to produce a <span class="nctnt ncbi-class">TSeqLocVector</span>.</p></div><div class="sec" id="ch_blast.Options_and_Program_"><h3><span class="title">Options and Program Type</span></h3><p>The <span class="nctnt ncbi-class">CBl2Seq</span> constructor takes either an <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=EProgram">EProgram</a> enum (see <a href="ch_blast.html#ch_blast.T18.3">Table 1</a>) or <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident?i=CBlastOptionsHandle">CBlastOptionsHandle</a> (see relevant section under <a href="ch_blast.html#ch_blast.CLocalBlast">CLocalBlast</a>). In the former case the default set of options for the given <span class="nctnt ncbi-type">EProgram</span> are used. In the latter case it is possible for the user to set options to non-default values.</p></div><div class="sec" id="ch_blast._Target_Sequences_2"><h3><span class="title">Target Sequences</span></h3><p>The target sequence(s) is represented either as a <span class="nctnt ncbi-type">SSeqLoc</span> or <span class="nctnt ncbi-class">TSeqLocVector</span>.</p></div><div class="sec" id="ch_blast._Results_2"><h3><span class="title">Results</span></h3><p>The <span class="nctnt ncbi-func">Run()</span> method of the <span class="nctnt ncbi-class">CBl2Seq</span> class returns a collection of <span class="nctnt ncbi-class">CSeq_align_set</span>’s. The method <span class="nctnt ncbi-func">GetMessages()</span> may be used to obtain any error or warning messages generated during the search.</p></div></div><div class="sec" id="ch_blast.Sample_Applications"><h2><span class="title">Sample Applications</span></h2><p>The following are sample applications that demonstrate the usage of the CBl2Seq and CLocalBlast classes respectively:</p><ul><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/sample/app/blast/blast_sample.cpp">blast_sample.cpp</a></p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/sample/app/blast/blast_demo.cpp">blast_demo.cpp</a></p></li></ul></div></div></div><div class="col four_col last"><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#ch_blast.CLocalBlast">CLocalBlast</a></li><li><a href="#ch_blast.CRemoteBlast">CRemoteBlast</a></li><li><a href="#ch_blast.The_Uniform_Interfac">The Uniform Interface</a></li><li><a href="#ch_blast.CBl2Seq">CBl2Seq</a></li><li><a href="#ch_blast.Sample_Applications">Sample Applications</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Download</h3></div></div><div class="portlet_content"><ul><li><a href="pdf/ch_blast.pdf">PDF version of this page</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                          middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />   
					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                          middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li><li class="internal-only">LXR: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">public</a> / <a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident">internal</a></li><li class="internal-only">Doxygen: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml">public</a> / <a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml">internal</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li><a href="http://www.ncbi.nlm.nih.gov/bookshelf/br.fcgi?book=toolkit&amp;part=ch_getcode_svn#ch_getcode_svn.code_retrieval">Server</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++">Web
                            (dev)</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++">Web
                            (extra)</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components">Web (prod)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit">Wiki</a></li><li class="internal-only"><a href="http://jira/browse/CXX">JIRA</a></li><li class="internal-only"><a href="http://intranet/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
